

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Covariate data preprocessing &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/covariate_preprocessing';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Factor analysis using Bi-Cross validation" href="covariate/BiCV_factor.html" />
    <link rel="prev" title="Phenotype data formatting" href="phenotype/phenotype_formatting.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">Covariate data preprocessing</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/covariate_preprocessing.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/covariate_preprocessing.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/data_preprocessing/covariate_preprocessing.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Covariate data preprocessing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-covariate">Merge covariate</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-removal-of-outliers-not-part-of-the-protocol-data-analysis">Optional: Removal of outliers (Not part of the protocol data analysis)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#residual-expression">Residual Expression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#factor-analysis">Factor Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merged-factor-covariates">Merged factor &amp; covariates</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="covariate-data-preprocessing">
<h1>Covariate data preprocessing<a class="headerlink" href="#covariate-data-preprocessing" title="Permalink to this headline">#</a></h1>
<p>This section documents output from the factor analysis section (colored in orange) of command generator MWE and explained the purpose for each of the command. The file used in this page can be found at <a class="reference external" href="https://drive.google.com/drive/folders/16ZUsciZHqCeeEWwZQR46Hvh5OtS8lFtA?usp=sharing">here</a>.</p>
<p><strong>Each commands in the factor analysis tutorials will be generated once per theme. The MWE is considered a one theme analysis</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">../</span><span class="n">images</span><span class="o">/</span><span class="n">eqtl_command</span><span class="o">.</span><span class="n">png</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ../images/eqtl_command.png (254.4 KiB):</div></div><img alt="../../_images/2a1007985802c4ec80878bfc482f95d9ef647170c5e60a3f2b38b9afd8480f24.png" src="../../_images/2a1007985802c4ec80878bfc482f95d9ef647170c5e60a3f2b38b9afd8480f24.png" />
</div>
</div>
<section id="merge-covariate">
<h2>Merge covariate<a class="headerlink" href="#merge-covariate" title="Permalink to this headline">#</a></h2>
<p>This step generate a concatanated principle component + covaraiate (pc-cov) matrixs so that we can use it to generate a residual phenotype, as outlined in the <span class="xref myst">phenotype processing</span> page</p>
<p>The tolerance for sample-wide NA rate of any covariates/pc are specified by <code class="docutils literal notranslate"><span class="pre">tol_cov</span></code>.  -1 means quit, otherwise for covariate with missing rate larger than tol_cov will be removed, with missing rate smaller than tol_cov will be mean_imputed.</p>
<p>The first k PCs that satisfied &gt;70% PVE will be merged with the covariates for downstream analysis and for estimating the residual expression in the <code class="docutils literal notranslate"><span class="pre">phenotype_preprocessing</span></code> section.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>
 sos run pipeline/covariate_formatting.ipynb merge_pca_covariate \
        --cwd output/data_preprocessing/protocol-eqtl/covariates \
        --pcaFile output/data_preprocessing/protocol-eqtl/pca/xqtl_protocol_data.merged.protocol-eqtl.related.filtered.extracted.pca.projected.rds \
        --covFile  ptotocol_data.covariate.tsv \
        --tol_cov 0.3  \
        --k `awk &#39;$3 &lt; 0.7&#39; output/data_preprocessing/protocol-eqtl/pca/xqtl_protocol_data.merged.protocol-eqtl.related.filtered.extracted.pca.projected.scree.txt | tail -1 | cut -f 1 ` \
        --container containers/bioinfo.sif  --name ptotocol_data.covariate.pca --outliersFile output/data_preprocessing/protocol-eqtl/pca/xqtl_protocol_data.merged.protocol-eqtl.related.filtered.extracted.pca.projected.outliers




 sos run pipeline/covariate_formatting.ipynb merge_pca_covariate \
        --cwd output/data_preprocessing/protocol-sqtl/covariates \
        --pcaFile output/data_preprocessing/protocol-sqtl/pca/xqtl_protocol_data.merged.protocol-sqtl.related.filtered.extracted.pca.projected.rds \
        --covFile  ptotocol_data.covariate.tsv \
        --tol_cov 0.3  \
        --k `awk &#39;$3 &lt; 0.7&#39; output/data_preprocessing/protocol-sqtl/pca/xqtl_protocol_data.merged.protocol-sqtl.related.filtered.extracted.pca.projected.scree.txt | tail -1 | cut -f 1 ` \
        --container containers/bioinfo.sif  --name ptotocol_data.covariate.pca --outliersFile output/data_preprocessing/protocol-sqtl/pca/xqtl_protocol_data.merged.protocol-sqtl.related.filtered.extracted.pca.projected.outliers




 sos run pipeline/covariate_formatting.ipynb merge_pca_covariate \
        --cwd output/data_preprocessing/protocol-mqtl/covariates \
        --pcaFile output/data_preprocessing/protocol-mqtl/pca/xqtl_protocol_data.merged.protocol-mqtl.related.filtered.extracted.pca.projected.rds \
        --covFile  ptotocol_data.covariate.tsv \
        --tol_cov 0.3  \
        --k `awk &#39;$3 &lt; 0.7&#39; output/data_preprocessing/protocol-mqtl/pca/xqtl_protocol_data.merged.protocol-mqtl.related.filtered.extracted.pca.projected.scree.txt | tail -1 | cut -f 1 ` \
        --container containers/bioinfo.sif  --name ptotocol_data.covariate.pca --outliersFile output/data_preprocessing/protocol-mqtl/pca/xqtl_protocol_data.merged.protocol-mqtl.related.filtered.extracted.pca.projected.outliers
</pre></div>
</div>
</div>
</div>
<section id="optional-removal-of-outliers-not-part-of-the-protocol-data-analysis">
<h3>Optional: Removal of outliers (Not part of the protocol data analysis)<a class="headerlink" href="#optional-removal-of-outliers-not-part-of-the-protocol-data-analysis" title="Permalink to this headline">#</a></h3>
<p>At this step, based on the output plot of PCA step, the user can decide if outliers should be removed. By default, the <code class="docutils literal notranslate"><span class="pre">outlierFile</span></code> will be filled, but the user have to manually add the <code class="docutils literal notranslate"><span class="pre">--remove-outliers</span></code> flag which will remove all the samples ID in the second column of the <code class="docutils literal notranslate"><span class="pre">.outlier</span></code> file from the header of pc-cov matrix, effectively remove them from all down stream analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/covariate_formatting.ipynb merge_pca_covariate \
        --cwd output/data_preprocessing/MWE/covariates \
        --pcaFile data_preprocessing/MWE/pca/MWE.MWE.related.filtered.extracted.pca.projected.rds \
        --covFile  MWE.covariate.cov.gz \
        --tol_cov 0.3  \
        --k `awk &#39;$3 &lt; 0.7&#39; output/data_preprocessing/MWE/pca/MWE.MWE.related.filtered.extracted.pca.projected.scree.txt | tail -1 | cut -f 1 ` \
        --container containers/bioinfo.sif  --outliersFile data_preprocessing/MWE/pca/MWE.MWE.related.filtered.extracted.pca.projected.outliers --remove-outliers
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="residual-expression">
<h2>Residual Expression<a class="headerlink" href="#residual-expression" title="Permalink to this headline">#</a></h2>
<p>The residual expression will be computed using the <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/blob/main/code/data_preprocessing/covariate/covariate_formatting.ipynb">covariate-PCs matrixs</a>. Residual expression is needed because  <a class="reference external" href="https://biofam.github.io/MOFA2/faq.html">PEER(MOFA2) handle covariates poorly</a>. However, the residual expression is only used for factor analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>  <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">covariate_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">compute_residual</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">resid_phenotype</span> \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">rnaseqc</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covFile</span>  <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ptotocol_data</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>  <span class="o">--</span><span class="kp">name</span> <span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">rnaseqc</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">cov_pca</span>




  <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">covariate_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">compute_residual</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">resid_phenotype</span> \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">leaf_cutter</span><span class="o">/</span><span class="n">xqtl_protocol_data_bam_list_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz_raw_data</span><span class="o">.</span><span class="n">qqnorm</span><span class="o">.</span><span class="n">formated</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covFile</span>  <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ptotocol_data</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>  <span class="o">--</span><span class="kp">name</span> <span class="n">xqtl_protocol_data_bam_list_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz_raw_data</span><span class="o">.</span><span class="n">qqnorm</span><span class="o">.</span><span class="n">formated</span><span class="o">.</span><span class="n">cov_pca</span>




  <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">covariate_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">compute_residual</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">resid_phenotype</span> \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">methylation</span><span class="o">/</span><span class="n">xqtl_protocol_data_arrayMethylation_covariates</span><span class="o">.</span><span class="n">sesame</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">filter_na</span><span class="o">.</span><span class="n">soft</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covFile</span>  <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ptotocol_data</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>  <span class="o">--</span><span class="kp">name</span> <span class="n">xqtl_protocol_data_arrayMethylation_covariates</span><span class="o">.</span><span class="n">sesame</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">filter_na</span><span class="o">.</span><span class="n">soft</span><span class="o">.</span><span class="n">cov_pca</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="factor-analysis">
<h2>Factor Analysis<a class="headerlink" href="#factor-analysis" title="Permalink to this headline">#</a></h2>
<p>The <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/blob/main/code/data_preprocessing/phenotype_preprocessing.ipynb">residual expression</a> will be used to conduct the factor analysis using either <a class="reference external" href="https://corbinq.github.io/apex/">BiCV (APEX)</a> or <a class="reference external" href="https://biofam.github.io/MOFA2/">PEER (MOFA2)</a>. The purpose of factor analysis is to uncovered un-measured factor embedded in the phenotype data, potentiall factor could be experiment batch effct, or unmeasure morbidity of the sampels .etc.</p>
<p>The primary factor analysis method we used in our analysis is PEER.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span> <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">PEER_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">PEER</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">covariates</span> \
   <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">resid_phenotype</span><span class="o">/</span><span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">rnaseqc</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>  \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">PEER</span><span class="o">.</span><span class="n">sif</span>  \
   <span class="o">--</span><span class="kp">walltime</span> <span class="mi">24</span><span class="n">h</span> \
   <span class="o">--</span><span class="n">numThreads</span> <span class="mi">8</span> \
   <span class="o">--</span><span class="n">iteration</span> <span class="mi">1000</span> \
   <span class="o">--</span><span class="n">N</span> <span class="mi">0</span>  




 <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">PEER_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">PEER</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">covariates</span> \
   <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">resid_phenotype</span><span class="o">/</span><span class="n">xqtl_protocol_data_bam_list_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz_raw_data</span><span class="o">.</span><span class="n">qqnorm</span><span class="o">.</span><span class="n">formated</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>  \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">PEER</span><span class="o">.</span><span class="n">sif</span>  \
   <span class="o">--</span><span class="kp">walltime</span> <span class="mi">24</span><span class="n">h</span> \
   <span class="o">--</span><span class="n">numThreads</span> <span class="mi">8</span> \
   <span class="o">--</span><span class="n">iteration</span> <span class="mi">1000</span> \
   <span class="o">--</span><span class="n">N</span> <span class="mi">0</span>  




 <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">PEER_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">PEER</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">covariates</span> \
   <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">resid_phenotype</span><span class="o">/</span><span class="n">xqtl_protocol_data_arrayMethylation_covariates</span><span class="o">.</span><span class="n">sesame</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">filter_na</span><span class="o">.</span><span class="n">soft</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>  \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">PEER</span><span class="o">.</span><span class="n">sif</span>  \
   <span class="o">--</span><span class="kp">walltime</span> <span class="mi">24</span><span class="n">h</span> \
   <span class="o">--</span><span class="n">numThreads</span> <span class="mi">8</span> \
   <span class="o">--</span><span class="n">iteration</span> <span class="mi">1000</span> \
   <span class="o">--</span><span class="n">N</span> <span class="mi">0</span>  
</pre></div>
</div>
</div>
</div>
</section>
<section id="merged-factor-covariates">
<h2>Merged factor &amp; covariates<a class="headerlink" href="#merged-factor-covariates" title="Permalink to this headline">#</a></h2>
<p>The factors estimated above will be concatanated with pc-cov matrix for downstream analysis as a complete covariate matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span> <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">covariate_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">merge_factor_covariate</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">covariates</span> \
        <span class="o">--</span><span class="n">factorFile</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">rnaseqc</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covFile</span>  <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">eqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ptotocol_data</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>  <span class="o">--</span><span class="kp">name</span> <span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">rnaseqc</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span>




 <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">covariate_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">merge_factor_covariate</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">covariates</span> \
        <span class="o">--</span><span class="n">factorFile</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">xqtl_protocol_data_bam_list_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz_raw_data</span><span class="o">.</span><span class="n">qqnorm</span><span class="o">.</span><span class="n">formated</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covFile</span>  <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">sqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ptotocol_data</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>  <span class="o">--</span><span class="kp">name</span> <span class="n">xqtl_protocol_data_bam_list_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz_raw_data</span><span class="o">.</span><span class="n">qqnorm</span><span class="o">.</span><span class="n">formated</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span>




 <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">covariate_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">merge_factor_covariate</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">covariates</span> \
        <span class="o">--</span><span class="n">factorFile</span> <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">xqtl_protocol_data_arrayMethylation_covariates</span><span class="o">.</span><span class="n">sesame</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">filter_na</span><span class="o">.</span><span class="n">soft</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covFile</span>  <span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">protocol</span><span class="o">-</span><span class="n">mqtl</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ptotocol_data</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>  <span class="o">--</span><span class="kp">name</span> <span class="n">xqtl_protocol_data_arrayMethylation_covariates</span><span class="o">.</span><span class="n">sesame</span><span class="o">.</span><span class="n">beta</span><span class="o">.</span><span class="n">filter_na</span><span class="o">.</span><span class="n">soft</span><span class="o">.</span><span class="n">cov_pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span>
</pre></div>
</div>
</div>
</div>
</section>
<div class="toctree-wrapper compound">
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="phenotype/phenotype_formatting.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Phenotype data formatting</p>
      </div>
    </a>
    <a class="right-next"
       href="covariate/BiCV_factor.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Factor analysis using Bi-Cross validation</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-covariate">Merge covariate</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-removal-of-outliers-not-part-of-the-protocol-data-analysis">Optional: Removal of outliers (Not part of the protocol data analysis)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#residual-expression">Residual Expression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#factor-analysis">Factor Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merged-factor-covariates">Merged factor &amp; covariates</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>