

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Reference data standardization &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/reference_data';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="RNA-seq expression" href="../molecular_phenotypes/bulk_expression.html" />
    <link rel="prev" title="Univariate xQTL Discovery" href="../commands_generator/eQTL_analysis_commands.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/reference_data.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/reference_data.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/data_preprocessing/reference_data.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Reference data standardization</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-files-for-rna-seq-expression-quantification">Reference files for RNA-seq expression quantification</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-files-for-methylation">Reference files for Methylation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-files-for-alternative-splicing">Reference files for alternative splicing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#details">Details</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#process-reference-fasta-file">Process reference fasta file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-transcript-feature-file-gff3-to-gtf">Convert transcript feature file gff3 to gtf</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Convert transcript feature file gff3 to gtf</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#collapse-transcript-features-into-genes">Collapse transcript features into genes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-star-index-based-on-gtf-and-reference-fasta">Generate STAR index based on gtf and reference fasta</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-rsem-index-based-on-gtf-and-reference-fasta">Generate RSEM index based on gtf and reference fasta</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#list-of-commands-to-prepare-reference-data">List of commands to prepare reference data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-download">Data download</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gff3-to-gtf-formatting">GFF3 to GTF formatting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#hg-reference-file-preprocessing">HG reference file preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#transcript-and-gene-model-reference-processing">Transcript and gene model reference processing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-star">Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-output">Step Output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-rsem">Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-refflat-file">Generation of RefFlat file</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-suppa-annotation-for-psichomics">Generation of SUPPA annotation for psichomics.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-psichomics-hg38-splicing-annotation">Generate psichomics Hg38 splicing annotation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-tad-associated-index">Generate TAD associated index</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="reference-data-standardization">
<h1>Reference data standardization<a class="headerlink" href="#reference-data-standardization" title="Permalink to this headline">#</a></h1>
<p>This module provides reference data download, indexing and preprocessing (if necessary), in preparation for use throughout the pipeline.</p>
<p>We have included the PDF document compiled by Data Standardization Working Group in the <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36416587">on Synapse</a> as well as on <a class="reference external" href="https://www.niagads.org/adsp/content/adspgcadgenomeresources-v2pdf">ADSP Dashboard</a>. It contains the reference data to use for the project.</p>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">#</a></h2>
<p>This module is based on the <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md">TOPMed workflow from Broad</a>. The reference data after we process it (details see Methods section and the rest of the analysis) can be found <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36416587">in this folder on Synapse</a>.</p>
<section id="reference-files-for-rna-seq-expression-quantification">
<h3>Reference files for RNA-seq expression quantification<a class="headerlink" href="#reference-files-for-rna-seq-expression-quantification" title="Permalink to this headline">#</a></h3>
<p><strong>We have decided to use these preprocessed reference files for RNA-seq expression and splicing quantification.</strong> Specifically:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.{dict,fasta,fasta.fai}</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf</span></code> for stranded protocol, and <code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.103.chr.reformatted.gene.ERCC.gtf</span></code> for unstranded protocol.</p></li>
<li><p>Everything under <code class="docutils literal notranslate"><span class="pre">STAR_Index</span></code> folder</p></li>
<li><p>Everything under <code class="docutils literal notranslate"><span class="pre">RSEM_Index</span></code> folder</p></li>
<li><p>Optionally, for quality control, <code class="docutils literal notranslate"><span class="pre">gtf_ref.flat</span></code></p></li>
</ol>
</section>
<section id="reference-files-for-methylation">
<h3>Reference files for Methylation<a class="headerlink" href="#reference-files-for-methylation" title="Permalink to this headline">#</a></h3>
<p>FIXME &#64;hao</p>
</section>
<section id="reference-files-for-alternative-splicing">
<h3>Reference files for alternative splicing<a class="headerlink" href="#reference-files-for-alternative-splicing" title="Permalink to this headline">#</a></h3>
<p>FIXME &#64;xuanhe</p>
</section>
</section>
<section id="details">
<h2>Details<a class="headerlink" href="#details" title="Permalink to this headline">#</a></h2>
<p>Workflows implemented include:</p>
<section id="process-reference-fasta-file">
<h3>Process reference fasta file<a class="headerlink" href="#process-reference-fasta-file" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Input:  a reference fasta file.</p></li>
<li><p>Output: a reference fasta file without HLA, ALT, Decoy but with ERCC.</p></li>
</ul>
</section>
<section id="convert-transcript-feature-file-gff3-to-gtf">
<h3>Convert transcript feature file gff3 to gtf<a class="headerlink" href="#convert-transcript-feature-file-gff3-to-gtf" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Input: an uncompressed gff3 file.(i.e. can be view via cat)</p></li>
<li><p>Output: a gtf file.</p></li>
</ul>
</section>
<section id="id1">
<h3>Convert transcript feature file gff3 to gtf<a class="headerlink" href="#id1" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Input: an uncompressed gff3 file.(i.e. can be view via cat)</p></li>
<li><p>Output: a gtf file.</p></li>
</ul>
</section>
<section id="collapse-transcript-features-into-genes">
<h3>Collapse transcript features into genes<a class="headerlink" href="#collapse-transcript-features-into-genes" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Input: a gtf file.</p></li>
<li><p>Output: a gtf file with collapesed gene model.</p></li>
</ul>
</section>
<section id="generate-star-index-based-on-gtf-and-reference-fasta">
<h3>Generate STAR index based on gtf and reference fasta<a class="headerlink" href="#generate-star-index-based-on-gtf-and-reference-fasta" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Input: a gtf file and an acompanying fasta file.</p></li>
<li><p>Output: A folder of STAR index.</p></li>
</ul>
</section>
<section id="generate-rsem-index-based-on-gtf-and-reference-fasta">
<h3>Generate RSEM index based on gtf and reference fasta<a class="headerlink" href="#generate-rsem-index-based-on-gtf-and-reference-fasta" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Input: a gtf file and an acompanying fasta file.</p></li>
<li><p>Output: A folder of RSEM index.</p></li>
</ul>
</section>
</section>
<section id="list-of-commands-to-prepare-reference-data">
<h2>List of commands to prepare reference data<a class="headerlink" href="#list-of-commands-to-prepare-reference-data" title="Permalink to this headline">#</a></h2>
<p>To download reference data,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_hg_reference</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_gene_annotation</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_ercc_reference</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_dbsnp</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: an hour, depending on the network.</p></li>
</ul>
<p>To format reference data,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">hg_reference</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">ercc</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">ERCC92</span><span class="o">.</span><span class="n">fa</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">fa</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">hg_gtf</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">stranded</span>
</pre></div>
</div>
</div>
</div>
<p>To format gene feature data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">gene_annotation</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">ercc</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">ERCC92</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">stranded</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: 10min</p></li>
<li><p>Memory: 16GB</p></li>
</ul>
<p><strong>Notice that for un-stranded RNA-seq protocol please use switch <code class="docutils literal notranslate"><span class="pre">--no-stranded</span></code> to the command above instead of <code class="docutils literal notranslate"><span class="pre">--stranded</span></code>. More details can be found later in the document.</strong></p>
<p>Generating STAR index without the GTF annotation file allow customize read lenght lateron in STAR alignment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">STAR_index</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: 30 min</p></li>
<li><p>Memory: 40 GB</p></li>
</ul>
<p>To generate RSEM index with the gtf file <strong>prior</strong> to the gene collapsing step ( <strong>without</strong> the gene tag in its file name.)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">RSEM_index</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>  
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: 1 min</p></li>
</ul>
<p>To generate RefFlat annotation for Picard QC,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">RefFlat_generation</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<p>To generate the SUPPA annotation for psichomics to detect RNA alternative splicing events,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">SUPPA_annotation</span> \
    <span class="o">--</span><span class="n">hg_gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">psichomics</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<p>Approximate time: ? min</p>
<p>This step extract the rsID for the known varriants, so that we can distinguish them from the novel variants we had in our study. The procedure/rationale is <a class="reference external" href="https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionVI/lessons/03_annotation-snpeff.html">explained in this post</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">VCF_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">dbsnp_annotate</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">reference_data</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">50</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">add_chr</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run reference_data.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  download_hg_reference
  download_gene_annotation
  download_ercc_reference
  download_dbsnp
  gff3_to_gtf
  hg_reference
  hg_gtf
  ercc_gtf
  gene_annotation
  STAR_index
  RSEM_index
  RefFlat_generation
  SUPPA_annotation
  psi_hg38_annotation_modification

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option

Sections
  download_hg_reference:
  download_gene_annotation:
  download_ercc_reference:
  download_dbsnp:
  gff3_to_gtf:
    Workflow Options:
      --gff3-file VAL (as path, required)
  hg_reference_1:
    Workflow Options:
      --hg-reference VAL (as path, required)
                        Path to HG reference file
  hg_reference_2:
    Workflow Options:
      --ercc-reference VAL (as path, required)
  hg_reference_3:
  hg_gtf_1:
    Workflow Options:
      --hg-reference VAL (as path, required)
      --hg-gtf VAL (as path, required)
  hg_gtf_2:
    Workflow Options:
      --[no-]stranded (required)
  ercc_gtf:
    Workflow Options:
      --ercc-gtf VAL (as path, required)
  gene_annotation:
  STAR_index:
    Workflow Options:
      --hg-reference VAL (as path, required)
  RSEM_index:
    Workflow Options:
      --hg-gtf VAL (as path, required)
      --hg-reference VAL (as path, required)
  RefFlat_generation:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  SUPPA_annotation_1:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  SUPPA_annotation_2:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  psi_hg38_annotation_modification_1:
    Workflow Options:
      --hg-gtf VAL (as path, required)
      --hgrc-db VAL (as path, required)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># The output directory for generated files.</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-download">
<h2>Data download<a class="headerlink" href="#data-download" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_hg_reference</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/GRCh38_full_analysis_set_plus_decoy_hla.fa&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="mf">.1000</span><span class="n">genomes</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">ftp</span><span class="o">/</span><span class="n">technical</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">GRCh38_reference_genome</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_gene_annotation</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/Homo_sapiens.GRCh38.103.chr.gtf&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">decompress</span><span class="o">=</span><span class="kc">True</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ensembl</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">release</span><span class="o">-</span><span class="mi">103</span><span class="o">/</span><span class="n">gtf</span><span class="o">/</span><span class="n">homo_sapiens</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_ercc_reference</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/ERCC92.gtf&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/ERCC92.fa&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">decompress</span><span class="o">=</span><span class="kc">True</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">tools</span><span class="o">.</span><span class="n">thermofisher</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">content</span><span class="o">/</span><span class="n">sfs</span><span class="o">/</span><span class="n">manuals</span><span class="o">/</span><span class="n">ERCC92</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_dbsnp</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/00-All.vcf.gz&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/00-All.vcf.gz.tbi&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">snp</span><span class="o">/</span><span class="n">organisms</span><span class="o">/</span><span class="n">human_9606</span><span class="o">/</span><span class="n">VCF</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">snp</span><span class="o">/</span><span class="n">organisms</span><span class="o">/</span><span class="n">human_9606</span><span class="o">/</span><span class="n">VCF</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">tbi</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="gff3-to-gtf-formatting">
<h2>GFF3 to GTF formatting<a class="headerlink" href="#gff3-to-gtf-formatting" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[gff3_to_gtf]
parameter: gff3_file = path
input: gff3_file
output: f&#39;{cwd}/{_input:bn}.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;
        gffread ${_input} -T -o ${_output}
</pre></div>
</div>
</div>
</div>
</section>
<section id="hg-reference-file-preprocessing">
<h2>HG reference file preprocessing<a class="headerlink" href="#hg-reference-file-preprocessing" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p>Remove the HLA/ALT/Decoy record from the fasta – because none of the downstreams RNA-seq calling pipeline component can handle them properly.</p></li>
<li><p>Adding in ERCC information to the fasta file – even if ERCC is not included in the RNA-seq library it does not harm to add them.</p></li>
<li><p>Generating index for the fasta file</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hg_reference_1</span> <span class="p">(</span><span class="n">HLA</span> <span class="n">ALT</span> <span class="n">Decoy</span> <span class="n">removal</span><span class="p">)]</span>
<span class="c1"># Path to HG reference file</span>
<span class="kn">parameter:</span> <span class="n">hg_reference</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_reference</span>
<span class="kn">output:</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.noALT_noHLA_noDecoy.fasta&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta</span><span class="p">:</span>
        <span class="n">contigs</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        <span class="n">contigs</span> <span class="o">=</span> <span class="n">contigs</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
        <span class="n">contig_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">contigs</span><span class="p">]</span>

        <span class="c1"># exclude ALT, HLA and decoy contigs</span>
        <span class="n">filtered_fasta</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">contig_ids</span><span class="p">,</span> <span class="n">contigs</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="o">-</span><span class="mi">4</span><span class="p">:]</span><span class="o">==</span><span class="s1">&#39;_alt&#39;</span> <span class="ow">or</span> <span class="n">i</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;HLA&#39;</span> <span class="ow">or</span> <span class="n">i</span><span class="p">[</span><span class="o">-</span><span class="mi">6</span><span class="p">:]</span><span class="o">==</span><span class="s1">&#39;_decoy&#39;</span><span class="p">)])</span>
    
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta</span><span class="p">:</span>
        <span class="n">fasta</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">filtered_fasta</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_reference_2 (merge with ERCC reference)]
parameter: ercc_reference = path
output: f&#39;{cwd}/{_input:bn}_ERCC.fasta&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container
    sed &#39;s/ERCC-/ERCC_/g&#39; ${ercc_reference} &gt; ${ercc_reference:n}.patched.fa
    cat ${_input} ${ercc_reference:n}.patched.fa &gt; ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_reference_3 (index the fasta file)]
output: f&#39;{cwd}/{_input:bn}.dict&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container
    samtools faidx ${_input}
    java -jar /opt/picard-tools/picard.jar \
        CreateSequenceDictionary \
        R=${_input} \
        O=${_output}
</pre></div>
</div>
</div>
</div>
</section>
<section id="transcript-and-gene-model-reference-processing">
<h2>Transcript and gene model reference processing<a class="headerlink" href="#transcript-and-gene-model-reference-processing" title="Permalink to this headline">#</a></h2>
<p>This step modify the <code class="docutils literal notranslate"><span class="pre">gtf</span></code> file for following reasons:</p>
<ol class="arabic simple">
<li><p>RSEM require GTF input to have the same chromosome name format (with <code class="docutils literal notranslate"><span class="pre">chr</span></code> prefix) as the fasta file. <strong>Although for STAR, this problem can be solved by the <code class="docutils literal notranslate"><span class="pre">--sjdbGTFchrPrefix</span> <span class="pre">&quot;chr&quot;</span></code> option, we have to add <code class="docutils literal notranslate"><span class="pre">chr</span></code> to it for use with RSEM anyways.</strong> That is why we would like to do it here.</p></li>
<li><p>Gene model collapsing script <code class="docutils literal notranslate"><span class="pre">collapse_annotation.py</span></code> from GTEx require the gtf have <code class="docutils literal notranslate"><span class="pre">transcript_type</span></code> instead <code class="docutils literal notranslate"><span class="pre">transcript_biotype</span></code> in its annotation. We rename it here, although <strong>this problem can also be solved by modifying the collapse_annotation.py while building the docker, since we are already customizing the reference file for 1 above, we add this in as well, as another customization made.</strong></p></li>
<li><p>Adding in ERCC information to the <code class="docutils literal notranslate"><span class="pre">gtf</span></code> reference.</p></li>
</ol>
<p>We may reimplement 1 and 2 with the alternative approaches discussed above if the problem with RSEM is solved, or when RSEM is no longer needed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_gtf_1 (add chr prefix to gtf file)]
parameter: hg_reference = path
parameter: hg_gtf = path
input: hg_reference, hg_gtf
output: f&#39;{cwd}/{_input[1]:bn}.reformatted.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;dplyr&quot;)
    options(scipen = 999)
    con &lt;- file(&quot;${_input[0]}&quot;,&quot;r&quot;)
    fasta &lt;- readLines(con,n=1)
    close(con)
    gtf = read_delim(&quot;${_input[1]}&quot;, &quot;\t&quot;,  col_names  = F, comment = &quot;#&quot;, col_types=&quot;ccccccccc&quot;)
    if(!str_detect(fasta,&quot;&gt;chr&quot;)) {
        gtf_mod = gtf%&gt;%mutate(X1 = str_remove_all(X1,&quot;chr&quot;))
    } else if (!any(str_detect(gtf$X1[1],&quot;chr&quot;))) {
        gtf_mod = gtf%&gt;%mutate(X1 = paste0(&quot;chr&quot;,X1))
    } else (gtf_mod = gtf)
    if(any(str_detect(gtf_mod$X9, &quot;transcript_biotype&quot;))) {
      gtf_mod = gtf_mod%&gt;%mutate(X9 = str_replace_all(X9,&quot;transcript_biotype&quot;,&quot;transcript_type&quot;))
    }
    gtf_mod%&gt;%write.table(&quot;${_output}&quot;,sep = &quot;\t&quot;,quote = FALSE,col.names = F,row.names = F)
</pre></div>
</div>
</div>
</div>
<p><strong>Text below is taken from <a class="github reference external" href="https://github.com/broadinstitute/gtex-pipeline/tree/master/gene_model">broadinstitute/gtex-pipeline</a></strong></p>
<p>Gene-level expression and eQTLs from the GTEx project are calculated based on a collapsed gene model (i.e., combining all isoforms of a gene into a single transcript), according to the following rules:</p>
<ol class="arabic simple">
<li><p>Transcripts annotated as “retained_intron” or “read_through” are excluded. Additionally, transcripts that overlap with annotated read-through transcripts may be blacklisted (blacklists for GENCODE v19, 24 &amp; 25 are provided in this repository; no transcripts were blacklisted for v26).</p></li>
<li><p>The union of all exon intervals of each gene is calculated.</p></li>
<li><p>Overlapping intervals between genes are excluded from all genes.</p></li>
</ol>
<p>The purpose of step 3 is primarily to exclude overlapping regions from genes annotated on both strands, which can’t be unambiguously quantified from unstranded RNA-seq (GTEx samples were sequenced using an unstranded protocol). For stranded protocols, this step can be skipped by adding the <code class="docutils literal notranslate"><span class="pre">--collapse_only</span></code> flag.</p>
<p>Further documentation is available on the <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">GTEx Portal</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_gtf_2 (collapsed gene model)]
parameter: stranded = bool
output: f&#39;{_input:n}{&quot;.collapse_only&quot; if stranded else &quot;&quot;}.gene.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container
    collapse_annotation.py ${&quot;--collapse_only&quot; if stranded else &quot;&quot;} ${_input} ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ercc_gtf</span> <span class="p">(</span><span class="n">Preprocess</span> <span class="n">ERCC</span> <span class="n">gtf</span> <span class="n">file</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">ercc_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">ercc_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.genes.patched.gtf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">exon_gtf</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gene_gtf</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">exon_gtf</span><span class="p">:</span>
            <span class="n">f</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>  <span class="c1"># required for RNA-SeQC/GATK (no &#39;-&#39; in contig name)</span>
        
            <span class="n">attr</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">attr</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;;&#39;</span><span class="p">:</span>
                <span class="n">attr</span> <span class="o">=</span> <span class="n">attr</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">attr</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">attr</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;; &#39;</span><span class="p">)])</span>
            <span class="c1"># add gene_name, gene_type</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ercc_control&#39;</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;KNOWN&#39;</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;level&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;status&#39;</span><span class="p">]:</span>
                <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;transcript_&#39;</span><span class="o">+</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_&#39;</span><span class="o">+</span><span class="n">k</span><span class="p">]</span>
        
            <span class="n">attr_str</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_type&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_status&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span>
                <span class="s1">&#39;transcript_type&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript_status&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript_name&#39;</span><span class="p">]:</span>
                <span class="n">attr_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0:s}</span><span class="s1"> &quot;</span><span class="si">{1:s}</span><span class="s1">&quot;;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">attr</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>
            <span class="n">attr_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0:s}</span><span class="s1"> </span><span class="si">{1:d}</span><span class="s1">;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;level&#39;</span><span class="p">,</span> <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;level&#39;</span><span class="p">]))</span>
            <span class="n">f</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">attr_str</span><span class="p">)</span>
        
            <span class="c1"># write gene, transcript, exon</span>
            <span class="n">gene_gtf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">3</span><span class="p">:])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">gene_gtf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;transcript&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">3</span><span class="p">:])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">f</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">attr_str</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">gene_gtf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;exon&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">3</span><span class="p">:])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[gene_annotation]
input: output_from(&quot;hg_gtf_1&quot;), output_from(&quot;hg_gtf_2&quot;), output_from(&quot;ercc_gtf&quot;)
output: f&#39;{cwd}/{_input[0]:bn}.ERCC.gtf&#39;, f&#39;{cwd}/{_input[1]:bn}.ERCC.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container
    cat ${_input[0]} ${_input[2]} &gt; ${_output[0]}
    cat ${_input[1]} ${_input[2]} &gt; ${_output[1]}
</pre></div>
</div>
</div>
</div>
</section>
<section id="generating-index-file-for-star">
<h2>Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code><a class="headerlink" href="#generating-index-file-for-star" title="Permalink to this headline">#</a></h2>
<p>This step generate the index file for STAR alignment. This file just need to generate once and can be re-used.</p>
<p><strong>At least 40GB of memory is needed</strong>.</p>
<section id="step-inputs">
<h3>Step Inputs<a class="headerlink" href="#step-inputs" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code> and <code class="docutils literal notranslate"><span class="pre">fasta</span></code>: path to reference sequence. Both of them needs to be unzipped. <code class="docutils literal notranslate"><span class="pre">gtf</span></code> should be the one prior to collapse by gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sjdbOverhang</span></code>: specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads. We use 100 here as recommended by the TOPMed pipeline. See here for <a class="reference external" href="https://groups.google.com/g/rna-star/c/h9oh10UlvhI/m/BfSPGivUHmsJ">some additional discussions</a>.</p></li>
</ul>
</section>
<section id="step-output">
<h3>Step Output<a class="headerlink" href="#step-output" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Indexing file stored in <code class="docutils literal notranslate"><span class="pre">{cwd}/STAR_index</span></code>, which will be used by <code class="docutils literal notranslate"><span class="pre">STAR</span></code></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[STAR_index]
parameter: hg_reference = path
# Specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads.
# Default choice follows from TOPMed pipeline recommendation.
if expand_size(mem) &lt; expand_size(&#39;40G&#39;):
    print(&quot;Insufficent memory for STAR, changing to 40G&quot;)
    star_mem = &#39;40G&#39;
else:
    star_mem = mem
input: hg_reference
output: f&quot;{cwd}/STAR_Index/chrName.txt&quot;, 
        f&quot;{cwd}/STAR_Index/SAindex&quot;, f&quot;{cwd}/STAR_Index/SA&quot;, f&quot;{cwd}/STAR_Index/genomeParameters.txt&quot;, 
        f&quot;{cwd}/STAR_Index/chrStart.txt&quot;,
        f&quot;{cwd}/STAR_Index/chrLength.txt&quot;, 
        f&quot;{cwd}/STAR_Index/Genome&quot;, f&quot;{cwd}/STAR_Index/chrNameLength.txt&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bd}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[1]:n}.stderr&#39;, stdout = f&#39;{_output[1]:n}.stdout&#39;
    STAR --runMode genomeGenerate \
         --genomeDir ${_output[0]:d} \
         --genomeFastaFiles ${_input[0]} \
         --runThreadN ${numThreads}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="generating-index-file-for-rsem">
<h2>Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#generating-index-file-for-rsem" title="Permalink to this headline">#</a></h2>
<p>This step generate the indexing file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>. This file just need to generate once.</p>
<section id="id2">
<h3>Step Inputs<a class="headerlink" href="#id2" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code> and <code class="docutils literal notranslate"><span class="pre">fasta</span></code>: path to reference sequence. <code class="docutils literal notranslate"><span class="pre">gtf</span></code> should be the one prior to collapse by gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sjdbOverhang</span></code>: specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads.</p></li>
</ul>
</section>
<section id="step-outputs">
<h3>Step Outputs<a class="headerlink" href="#step-outputs" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Indexing file stored in <code class="docutils literal notranslate"><span class="pre">RSEM_index_dir</span></code>, which will be used by <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[RSEM_index]
parameter: hg_gtf = path
parameter: hg_reference = path
input: hg_reference, hg_gtf
output: f&quot;{cwd}/RSEM_Index/rsem_reference.n2g.idx.fa&quot;, f&quot;{cwd}/RSEM_Index/rsem_reference.grp&quot;, 
        f&quot;{cwd}/RSEM_Index/rsem_reference.idx.fa&quot;, f&quot;{cwd}/RSEM_Index/rsem_reference.ti&quot;, 
        f&quot;{cwd}/RSEM_Index/rsem_reference.chrlist&quot;, f&quot;{cwd}/RSEM_Index/rsem_reference.seq&quot;, 
        f&quot;{cwd}/RSEM_Index/rsem_reference.transcripts.fa&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bd}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[1]:n}.stderr&#39;, stdout = f&#39;{_output[1]:n}.stdout&#39;
    rsem-prepare-reference \
            ${_input[0]} \
            ${_output[1]:n} \
            --gtf ${_input[1]} \
            --num-threads ${numThreads}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="generation-of-refflat-file">
<h2>Generation of RefFlat file<a class="headerlink" href="#generation-of-refflat-file" title="Permalink to this headline">#</a></h2>
<p>This file is needed for picard CollectRnaSeqMetrics module, which in turn</p>
<blockquote>
<div><p>produces metrics describing the distribution of the bases within the transcripts. It calculates the total numbers and the fractions of nucleotides within specific genomic regions including untranslated regions (UTRs), introns, intergenic sequences (between discrete genes), and peptide-coding sequences (exons). This tool also determines the numbers of bases that pass quality filters that are specific to Illumina data (PF_BASES).</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[RefFlat_generation]
parameter: hg_gtf = path
input: hg_gtf
output: f&#39;{_input:n}.ref.flat&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bd}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
    gtfToGenePred ${_input}  ${_output}.tmp -genePredExt -geneNameAsName2
    awk -F&#39;\t&#39; -v OFS=&quot;\t&quot; &#39;{$1=$12 OFS $1;}7&#39; ${_output}.tmp | cut -f 1-11 &gt; ${_output}
    rm ${_output}.tmp
</pre></div>
</div>
</div>
</div>
</section>
<section id="generation-of-suppa-annotation-for-psichomics">
<h2>Generation of SUPPA annotation for psichomics.<a class="headerlink" href="#generation-of-suppa-annotation-for-psichomics" title="Permalink to this headline">#</a></h2>
<p>The generation of custom alternative splicing annotation is based on <a class="reference external" href="https://rpubs.com/nuno-agostinho/preparing-AS-annotation">this tutorial</a>. The way to generate the local alternative splicing suppa output is documented on <a class="reference external" href="https://github.com/comprna/SUPPA">SUPPA github page</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">SUPPA_annotation</span> \
    <span class="o">--</span><span class="n">hg_gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">psochimics</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[SUPPA_annotation_1]
parameter: hg_gtf = path
input: hg_gtf
output: f&#39;{cwd}/hg38.{_input:bn}_SE_strict.ioe&#39; # The stderr file must not shared the same start with the output file
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{cwd}/{_input:bn}.stderr&#39;, stdout = f&#39;{cwd}/{_input:bn}.stdout&#39;
    python /opt/SUPPA/suppa.py generateEvents -i ${_input} -o ${cwd}/hg38.${_input:bn} -f ioe -e SE SS MX RI FL
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[SUPPA_annotation_2]
parameter: hg_gtf = path
output: f&#39;{cwd}/{hg_gtf:bn}.SUPPA_annotation.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
    library(&quot;psichomics&quot;)
    suppa &lt;- parseSuppaAnnotation(&quot;${_input:d}&quot;, genome=&quot;hg38&quot;) 
    annot &lt;- prepareAnnotationFromEvents(suppa)
    saveRDS(annot, file=${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
<section id="generate-psichomics-hg38-splicing-annotation">
<h2>Generate psichomics Hg38 splicing annotation<a class="headerlink" href="#generate-psichomics-hg38-splicing-annotation" title="Permalink to this headline">#</a></h2>
<p>Since the original annotation provided by psichomics package is using gene symbols, we modified it to use Ensembl IDs. The modified annotation will be used for <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/molecular_phenotypes/calling/splicing_calling.html">detecting RNA alternative splicing using psichomics</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">psi_hg38_annotation</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformated</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hgrc</span><span class="o">-</span><span class="n">db</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">hgnc_database</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">psichomics</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[psi_hg38_annotation]
parameter: hg_gtf = path
# FIXME: Please document what this file is and where do we get it @xuanhe.
parameter: hgrc_db = path
input: hg_gtf, hgrc_db
output: f&#39;{cwd}/psichomics_hg38_annotation.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
    library(&quot;psichomics&quot;)
    library(&quot;purrr&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;data.table&quot;)
  
    # load psicomics default annotation, option of hg38 from listSplicingAnnotations()
    annotation &lt;- loadAnnotation(&quot;AH63657&quot;)

    
    # reduce the demension of annotation file
    annotation &lt;- 
      map(annotation, ~.x%&gt;%
                       tidyr::unnest(cols = `Gene`))
  
    # Create empty colomns for each event for easier mapping
    annotation[[&quot;Tandem UTR&quot;]][[&quot;SUPPA.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Tandem UTR&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Alternative first exon&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Alternative last exon&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Mutually exclusive exon&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
  
    # extract Ensembl ID substring from original SUPPA.ID and VASTTOOL.ID
    annotation &lt;- 
      map(annotation, ~.x%&gt;%
                       mutate(ENSG.SUPPA = substr(`SUPPA.Event.ID`, 1, 15))%&gt;%
                       mutate(ENSG.VAST = substr(`VAST-TOOLS.Event.ID`, 1, 15)))
  
    # Load gtf file
    gtf_sample &lt;- read.table(&#39;${_input[0]}&#39;,header = FALSE, sep = &#39;\t&#39;)
  
    # from the gtf file, seperate gene names and corresponding Ensembl ID
    gtf_sample &lt;- separate(gtf_sample, V9, sep = &quot;;&quot;,into = c(&quot;gene_id&quot;, &quot;transcript_id&quot;, &quot;exon_number&quot;, &quot;gene_name&quot;))
    gtf_sample &lt;- separate(gtf_sample, gene_id, sep = &quot; &quot;,into = c(&quot;gene_id&quot;, &quot;gene_id_val&quot;))
    gtf_sample &lt;- separate(gtf_sample, gene_name, sep = &quot;e &quot;,into = c(&quot;gene_name&quot;, &quot;gene_name_val&quot;))
  
    gtf_name_id_match &lt;- gtf_sample[,c(&quot;gene_id_val&quot;,&quot;gene_name_val&quot;)]
    gtf_name_id_match &lt;- gtf_name_id_match[!duplicated(gtf_name_id_match), ]
  
    # For any matched approved id in the psi hg38 annotation and gtf file, record the corresponding Ensembl ID
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                       mutate(`ENSG.GTF` = gtf_name_id_match$gene_id_val[match(`Gene`, gtf_name_id_match$gene_name_val)]))
  
    # load hgnc database
    hgnc_db &lt;- fread(&#39;${_input[1]}&#39;, fill = TRUE, header = TRUE, sep = &#39;\t&#39;, quote=&quot;&quot;)
  
    # Combine the `Ensembl.ID.supplied.by.Ensembl.` and `Ensembl.gene.ID` column, if there are any conflict use the former
    # For conflict ones (15 total) both the former and latter records are poiting to the same gene name in Ensembl website so the order should not matter
    hgnc_db &lt;- hgnc_db %&gt;%
    mutate(ENSG.ID = ifelse(`Ensembl ID(supplied by Ensembl)` == &quot;&quot;, `Ensembl gene ID`, `Ensembl ID(supplied by Ensembl)`))
  
    # Create a one to one reference list for approved names, previous names and aliases
    # There is no duplicate symbol and Ensembl id info for approved symbol so no need for chromosome verification
    hgnc_name_id_match &lt;- hgnc_db[,c(&quot;Approved symbol&quot;,&quot;ENSG.ID&quot;)]
    hgnc_name_prev_check &lt;- hgnc_db[,c(&quot;Previous symbols&quot;,&quot;Chromosome&quot;,&quot;ENSG.ID&quot;)]
    hgnc_name_alias_check &lt;- hgnc_db[,c(&quot;Alias symbols&quot;,&quot;Chromosome&quot;,&quot;ENSG.ID&quot;)]
  
    # Remove NAs
    hgnc_name_prev_check &lt;- hgnc_name_prev_check[hgnc_name_prev_check$ENSG.ID != &quot;&quot;,]
    hgnc_name_alias_check &lt;- hgnc_name_alias_check[hgnc_name_alias_check$ENSG.ID != &quot;&quot;,]

    hgnc_name_prev_check &lt;- hgnc_name_prev_check[hgnc_name_prev_check$&quot;Previous symbols&quot; != &quot;&quot;,] 
    hgnc_name_alias_check &lt;- hgnc_name_alias_check[hgnc_name_alias_check$&quot;Alias symbols&quot; != &quot;&quot;,]

    # Seperate symbol column values from list of sybols to individual rows with one each
    hgnc_name_prev_check &lt;- separate_rows(hgnc_name_prev_check, &quot;Previous symbols&quot;, convert = FALSE)
    hgnc_name_alias_check &lt;- separate_rows(hgnc_name_alias_check, &quot;Alias symbols&quot;, convert = FALSE)
  
    # Convert chomosome info in hgnc database to number for matching with other database
    hgnc_name_prev_check &lt;- separate(hgnc_name_prev_check, &quot;Chromosome&quot;, sep = &#39;p&#39;, into = &quot;Chrp&quot;, remove = FALSE)
    hgnc_name_prev_check &lt;- separate(hgnc_name_prev_check, &quot;Chromosome&quot;, sep = &#39;q&#39;, into = &quot;Chrq&quot;, remove = FALSE)
    hgnc_name_prev_check &lt;- hgnc_name_prev_check%&gt;%
                                mutate(Chr = ifelse(nchar(hgnc_name_prev_check$Chrp) &lt;= 2, Chrp, Chrq))
    
    hgnc_name_alias_check &lt;- separate(hgnc_name_alias_check, &quot;Chromosome&quot;, sep = &#39;p&#39;, into = &quot;Chrp&quot;, remove = FALSE)
    hgnc_name_alias_check &lt;- separate(hgnc_name_alias_check, &quot;Chromosome&quot;, sep = &#39;q&#39;, into = &quot;Chrq&quot;, remove = FALSE)
    hgnc_name_alias_check &lt;- hgnc_name_alias_check%&gt;%
                                mutate(Chr = ifelse(nchar(hgnc_name_alias_check$Chrp) &lt;= 2, Chrp, Chrq))
  
    # For any matched approved id in the psi hg38 annotation and hgnc database, record the corresponding Ensembl ID
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                       mutate(`ENSG.HGNC` = hgnc_name_id_match$`ENSG.ID`[match(`Gene`, hgnc_name_id_match$&quot;Approved symbol&quot;)]))
  
    # Drop hypothetical genes
    annotation&lt;-
      map(annotation, ~.x%&gt;%
            subset(`Gene` != &#39;Hypothetical&#39;))
  
    # IN remaining NAs, for any matched alias/previous names and chromosome in the psi hg38 annotation and hgnc database, record the corresponding Ensembl ID
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                      mutate(ENSG.HGNC = ifelse(is.na(`ENSG.HGNC`) | `ENSG.HGNC` == &quot;&quot;,
                                                hgnc_name_alias_check$`ENSG.ID`[match(`Gene`, hgnc_name_alias_check$&quot;Alias symbols&quot;) &amp; match(`Chromosome`, hgnc_name_alias_check$Chr)],
                                                `ENSG.HGNC`))%&gt;%
                      mutate(ENSG.HGNC = ifelse(is.na(`ENSG.HGNC`) | `ENSG.HGNC` == &quot;&quot;,
                                                hgnc_name_prev_check$`ENSG.ID`[match(`Gene`, hgnc_name_prev_check$&quot;Previous symbols&quot;) &amp; match(`Chromosome`, hgnc_name_prev_check$Chr)],
                                                `ENSG.HGNC`)))
  
    # Build the final Ensembl id column base on the gtf file first, then for remaining NAs check the HGNC database record, SUPPA and VASTTOOL record,
    # Drop special cases that Ensembl ID is not recorded in VASTTOOLs and SUPPA
    # finnally in the remaining NA Ensmbl ID if the gene name in original annotation is NCBI IDs just use it
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                       mutate(`ENSG.ID` = `ENSG.GTF`)%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`),
                                               `ENSG.HGNC`,
                                                `ENSG.ID`))%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`),
                                               `ENSG.VAST`,
                                                `ENSG.ID`))%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`),
                                               `ENSG.SUPPA`,
                                                `ENSG.ID`))%&gt;%
                       mutate(`ENSG.ID` = ifelse(substr(`ENSG.ID`, 1, 4) == &#39;ENSG&#39;,
                                                 `ENSG.ID`,
                                                 NA))%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`) &amp; substr(`Gene`, 1, 3) == &#39;LOC&#39;,
                                               `Gene`,
                                                `ENSG.ID`)))

    # Use the Ensembl IDs to replace gene names, drop remaining NAs
    annotation &lt;-
    map(annotation, ~.x%&gt;%
                       mutate(`Gene` = `ENSG.ID`)%&gt;%
            drop_na(`Gene`)
          )

    # save modified annotation
    saveRDS(annotation, file = &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="generate-tad-associated-index">
<h2>Generate TAD associated index<a class="headerlink" href="#generate-tad-associated-index" title="Permalink to this headline">#</a></h2>
<p>We use <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7895846/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7895846/</a> to define the genotype cis region for fine mapping with different phenotypes. The actual TAD dataframe for each cell types and tissues are download from the hg38 TAD link from  <a class="reference external" href="http://3dgenome.fsm.northwestern.edu/publications.html">http://3dgenome.fsm.northwestern.edu/publications.html</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Tad_annotateion</span><span class="p">]</span>
<span class="c1">## The tad file downloads from </span>
<span class="kn">parameter:</span> <span class="n">TAD_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">TAD_genotype</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">TAD_list</span><span class="p">,</span> <span class="n">region_list</span>
<span class="kn">output:</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">region_list</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.annotation&#39;</span> 
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="n">tad</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input[0]}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">tad</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
    <span class="n">tad</span><span class="p">[</span><span class="s2">&quot;tad_index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;tad</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">tad</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
    <span class="n">region_list</span> <span class="o">=</span>  <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input[1]}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">tad</span> <span class="o">=</span> <span class="n">tad</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s2">&quot;#chr&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;start_y &gt; start_x &amp;  end_y &lt; end_x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;start_y&quot;</span><span class="p">,</span><span class="s2">&quot;end_y&quot;</span><span class="p">]</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;start_x&quot;</span><span class="p">:</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end_x&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span> <span class="p">})</span>
    <span class="n">tad</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> 
    <span class="n">tad_geno_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_TAD_genotype}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">tad_gene_id</span> <span class="o">=</span> <span class="n">tad</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">tad_geno_list</span><span class="p">,</span><span class="n">left_on</span> <span class="o">=</span> <span class="s2">&quot;tad_index&quot;</span><span class="p">,</span><span class="n">right_on</span> <span class="o">=</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">4</span><span class="p">,</span><span class="mi">6</span><span class="p">]]</span>
    <span class="n">tad_gene_id</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="s2">&quot;dir&quot;</span><span class="p">]</span>
    <span class="n">tad_gene_id</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">,genotype_files_list.tsv&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/hs3163/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument &#39;filepath_or_buffer&#39; will be keyword-only.
  exec(code_obj, self.user_global_ns, self.user_ns)
/home/hs3163/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument &#39;filepath_or_buffer&#39; will be keyword-only.
  exec(code_obj, self.user_global_ns, self.user_ns)
/home/hs3163/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument &#39;filepath_or_buffer&#39; will be keyword-only.
  exec(code_obj, self.user_global_ns, self.user_ns)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../commands_generator/eQTL_analysis_commands.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Univariate xQTL Discovery</p>
      </div>
    </a>
    <a class="right-next"
       href="../molecular_phenotypes/bulk_expression.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">RNA-seq expression</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-files-for-rna-seq-expression-quantification">Reference files for RNA-seq expression quantification</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-files-for-methylation">Reference files for Methylation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-files-for-alternative-splicing">Reference files for alternative splicing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#details">Details</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#process-reference-fasta-file">Process reference fasta file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-transcript-feature-file-gff3-to-gtf">Convert transcript feature file gff3 to gtf</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Convert transcript feature file gff3 to gtf</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#collapse-transcript-features-into-genes">Collapse transcript features into genes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-star-index-based-on-gtf-and-reference-fasta">Generate STAR index based on gtf and reference fasta</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-rsem-index-based-on-gtf-and-reference-fasta">Generate RSEM index based on gtf and reference fasta</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#list-of-commands-to-prepare-reference-data">List of commands to prepare reference data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-download">Data download</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gff3-to-gtf-formatting">GFF3 to GTF formatting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#hg-reference-file-preprocessing">HG reference file preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#transcript-and-gene-model-reference-processing">Transcript and gene model reference processing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-star">Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-output">Step Output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-rsem">Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-refflat-file">Generation of RefFlat file</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-suppa-annotation-for-psichomics">Generation of SUPPA annotation for psichomics.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-psichomics-hg38-splicing-annotation">Generate psichomics Hg38 splicing annotation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-tad-associated-index">Generate TAD associated index</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>