

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Univariate xQTL Discovery &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/commands_generator/eQTL_analysis_commands';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Reference data standardization" href="../data_preprocessing/reference_data.html" />
    <link rel="prev" title="RNA-seq calling and QC" href="bulk_expression_commands.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/commands_generator/eQTL_analysis_commands.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/commands_generator/eQTL_analysis_commands.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/commands_generator/eQTL_analysis_commands.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Univariate xQTL Discovery</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype">Genotype</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#per-tissue-analysis">Per tissue analysis</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#protocol-data">Protocol data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-for-running-the-workflow">Example for running the workflow</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preprocessing">Data Preprocessing</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-preprocessing-once-for-all-tissues">Genotype Preprocessing (Once for all tissues)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#molecular-phenotype-processing">Molecular Phenotype Processing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-processing">Genotype Processing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#factor-analysis">Factor Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#association-scan">Association Scan</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trans-association-scan">Trans Association Scan</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sumstat-merger">Sumstat Merger</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="univariate-xqtl-discovery">
<h1>Univariate xQTL Discovery<a class="headerlink" href="#univariate-xqtl-discovery" title="Permalink to this headline">#</a></h1>
<p>This notebook provide a command generator on the XQTL workflow so it can automate the work for data preprocessing and association testing on multiple data collection as proposed. This version of command generator is suitable for bed file with sQTL and eQTL, the user should specify the input and the output in a recipe file, as documented below</p>
<p>This master control notebook is mainly to serve the 8 tissues snuc_bulk_expression analysis, but should be functional on all analysis where expression data are are a tsv table in a bed.gz like format.</p>
<p>Input:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>A recipe file,each row is a data collection and with the following column:

    Theme
        name of dataset, must be different, each uni_study analysis will be performed in a folder named after each, meta analysis will be performed in a folder named as {study1}_{study2}
        
        The column name must contain the # and be the first column

    genotype_file
        {Path to a whole genome genotype file}
    
    molecular_pheno
        {Path to file}
        
    covariate_file
        {Path to file}
    
    ### note: Only data collection from the same Populations and conditions will me merged to perform Fix effect meta analysis
    
A genotype list, with two column, `#chr` and  `path`
    
    This can be generated by the genotype session of this command generator.
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>1 set of association_scan result for each tissue (each row in the recipe)
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Theme&quot;</span><span class="p">:</span><span class="s2">&quot;MWE&quot;</span><span class="p">,</span><span class="s2">&quot;molecular_pheno&quot;</span><span class="p">:</span><span class="s2">&quot;MWE.log2cpm.tsv&quot;</span><span class="p">,</span><span class="s2">&quot;genotype_file&quot;</span><span class="p">:</span><span class="s2">&quot;MWE.bed&quot;</span><span class="p">,</span><span class="s2">&quot;covariate_file&quot;</span><span class="p">:</span><span class="s2">&quot;MWE.covariate.cov.gz&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;/mnt/vast/hpc/csg/snuc_pseudo_bulk/eight_tissue_analysis/MWE/command_generator&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Theme</p></th>
<th class="head"><p>molecular_pheno</p></th>
<th class="head"><p>genotype_file</p></th>
<th class="head"><p>covariate_file</p></th>
<th class="head"><p>phenotype_group</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>eQTL</p></td>
<td><p>MWE.log2cpm.bed.gz</p></td>
<td><p>/data/genotype_data/GRCh38_liftedover_sorted_all.add_chr.leftnorm.filtered.bed</p></td>
<td><p>MWE.covariate.cov.gz</p></td>
<td><p>/data/genotype_data/GRCh38_liftedover_sorted_all.add_chr.leftnorm.filtered.phenotype_group</p></td>
</tr>
</tbody>
</table>
<section id="minimal-working-example">
<h2>Minimal Working Example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">#</a></h2>
<section id="genotype">
<h3>Genotype<a class="headerlink" href="#genotype" title="Permalink to this headline">#</a></h3>
<p>The MWE for the genotype session can be ran with the following commands, please be noted that a <a class="reference external" href="https://drive.google.com/file/d/1zaacRlZ63Nf_oEUv2nIiqekpQmt2EDch/view?usp=sharing">seperated MWE genoFile</a> was needed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">plink_per_chrom</span> \
    <span class="o">--</span><span class="n">ref_fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>    \
    <span class="o">--</span><span class="n">genoFile</span>   <span class="n">mwe_genotype</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">dbSNP_vcf</span>  <span class="n">reference_data</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">n</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="per-tissue-analysis">
<h3>Per tissue analysis<a class="headerlink" href="#per-tissue-analysis" title="Permalink to this headline">#</a></h3>
<p>A MWE for the core per tissue analysis can be ran with the following commands, a complete collection of input file as well as intermediate output of the analysis can be found at <a class="reference external" href="https://drive.google.com/drive/folders/16ZUsciZHqCeeEWwZQR46Hvh5OtS8lFtA?usp=sharing">here</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_standardizer</span> \
    <span class="o">--</span><span class="n">recipe</span> <span class="n">MWE</span><span class="o">.</span><span class="n">recipe</span>   \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>      \
    <span class="o">--</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>     \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;TensorQTL&quot;</span>  <span class="o">-</span><span class="n">n</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span> <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_standardizer</span>    \
    <span class="o">--</span><span class="n">recipe</span> <span class="n">MWE</span><span class="o">.</span><span class="n">recipe</span>    \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>      \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span>    \
    <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;TensorQTL&quot;</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">output2</span> <span class="o">--</span><span class="n">run</span>  <span class="o">--</span><span class="n">yml</span> <span class="s2">&quot;&quot;</span> <span class="o">--</span><span class="n">N</span> <span class="mi">3</span>   
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_merge</span> \
    <span class="o">--</span><span class="n">recipe</span>  <span class="n">MWE</span><span class="o">.</span><span class="n">recipe</span> \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>      \
    <span class="o">--</span><span class="n">annotation_gtf</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">genes</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">gtf</span>     \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;APEX&quot;</span>   <span class="o">-</span><span class="n">n</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">factor</span> \
    <span class="o">--</span><span class="n">recipe</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">methyl</span><span class="o">/</span><span class="n">recipe_PCC_methylation</span> \
    <span class="o">--</span><span class="n">annotation_gtf</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">genes</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">gtf</span>     \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">methyl</span><span class="o">/</span><span class="n">Methylation_sample_list</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">n</span> <span class="o">&gt;</span> <span class="n">Methylation_mqtl_script</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="protocol-data">
<h2>Protocol data<a class="headerlink" href="#protocol-data" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">plink_per_chrom</span> \
    <span class="o">--</span><span class="n">ref_fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>    \
    <span class="o">--</span><span class="n">genoFile</span>   <span class="n">input_data</span><span class="o">/</span><span class="n">Genotype</span><span class="o">/</span><span class="n">DEJ_11898_B01_GRM_WGS_2017</span><span class="o">-</span><span class="mi">05</span><span class="o">-</span><span class="mi">15_2</span><span class="o">*.</span><span class="n">recalibrated_variants</span><span class="o">.</span><span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">add_chr</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">dbSNP</span>  <span class="n">reference_data</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">add_chr</span><span class="o">.</span><span class="n">variants</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">rnaseq</span> <span class="o">--</span><span class="n">run</span> <span class="o">--</span><span class="n">monitor</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">output_rerun</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span> <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_standardizer</span>    \
    <span class="o">--</span><span class="n">recipe</span> <span class="n">protocol_data</span><span class="o">.</span><span class="n">recipe</span>    \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="n">output_rerun</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>      \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">rnaseq</span>    \
    <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;TensorQTL&quot;</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">output3</span> <span class="o">--</span><span class="n">run</span>  <span class="o">--</span><span class="n">N</span> <span class="mi">3</span>    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_rerun</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span> <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_standardizer</span>    \
    <span class="o">--</span><span class="n">recipe</span> <span class="n">protocol_data</span><span class="o">.</span><span class="n">recipe</span>    \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="n">output_rerun</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">xqtl_protocol_data</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>      \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">rnaseq</span>    \
    <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;TensorQTL&quot;</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">output3</span> <span class="o">--</span><span class="n">run</span>  <span class="o">--</span><span class="n">N</span> <span class="mi">3</span>    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_hidden_factor</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="example-for-running-the-workflow">
<h2>Example for running the workflow<a class="headerlink" href="#example-for-running-the-workflow" title="Permalink to this headline">#</a></h2>
<p>This will run the workflow from via several submission</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_merge</span> \
    <span class="o">--</span><span class="n">recipe</span>  <span class="n">data</span><span class="o">/</span><span class="n">recipe_8tissue_new</span> \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="n">data</span><span class="o">/</span><span class="n">genotype_qced</span><span class="o">/</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>      \
    <span class="o">--</span><span class="n">annotation_gtf</span> <span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">genes</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">gtf</span>     \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;TensorQTL&quot;</span>   <span class="o">-</span><span class="n">n</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">eQTL_analysis_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">sumstat_merge</span>  \
    <span class="o">--</span><span class="n">recipe</span>  <span class="o">&lt;</span><span class="p">(</span><span class="n">cat</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">//</span><span class="n">data</span><span class="o">/</span><span class="n">recipe_8tissue_new</span>  <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">2</span><span class="p">)</span>    \
    <span class="o">--</span><span class="n">genotype_list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">genotype_qced</span><span class="o">/</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>      \
    <span class="o">--</span><span class="n">annotation_gtf</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">genes</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">gtf</span>     \
    <span class="o">--</span><span class="n">sample_participant_lookup</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">factor_option</span> <span class="s2">&quot;PEER&quot;</span> <span class="o">--</span><span class="n">Association_option</span> <span class="s2">&quot;TensorQTL&quot;</span>   <span class="o">-</span><span class="n">n</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1">## The aforementioned input recipe</span>
<span class="kn">parameter:</span> <span class="n">recipe</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>  <span class="c1"># Added option to run genotype part without the recipe input, which was not used.</span>
<span class="c1">## Overall wd, the file structure of analysis is wd/[steps]/[sub_dir for each steps]</span>
<span class="kn">parameter:</span>  <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1">## Diretory to the excutable</span>
<span class="kn">parameter:</span> <span class="n">exe_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;~/GIT/xqtl-pipeline/&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">container_base_bioinfo</span> <span class="o">=</span> <span class="s1">&#39;containers/bioinfo.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_apex</span> <span class="o">=</span> <span class="s1">&#39;containers/apex.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_PEER</span> <span class="o">=</span> <span class="s1">&#39;containers/PEER.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_PCAtools</span> <span class="o">=</span> <span class="s1">&#39;containers/PCAtools.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_TensorQTL</span> <span class="o">=</span> <span class="s1">&#39;containers//TensorQTL.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_rnaquant</span> <span class="o">=</span> <span class="s1">&#39;containers/rna_quantification.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_flashpca</span> <span class="o">=</span> <span class="s1">&#39;containers/flashpcaR.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_susie</span> <span class="o">=</span> <span class="s1">&#39;containers/stephenslab.sif&#39;</span>
<span class="kn">parameter:</span> <span class="n">phenotype_id_type</span> <span class="o">=</span> <span class="s2">&quot;gene_name&quot;</span> 
<span class="kn">parameter:</span> <span class="n">yml</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;csg.yml&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">run</span> <span class="o">=</span> <span class="kc">False</span>
<span class="kn">parameter:</span> <span class="n">monitor</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">interpreter</span> <span class="o">=</span> <span class="s1">&#39;cat&#39;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">run</span> <span class="k">else</span> <span class="s1">&#39;bash&#39;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="k">if</span> <span class="n">recipe</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
    <span class="n">input_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">recipe</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;records&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">parameter:</span> <span class="n">jobs</span> <span class="o">=</span> <span class="mi">50</span> <span class="c1"># Number of jobs that are submitted to the cluster</span>
<span class="kn">parameter:</span> <span class="kp">queue</span> <span class="o">=</span> <span class="s2">&quot;csg2&quot;</span> <span class="c1"># The queue that jobs are submitted to</span>
<span class="n">submission</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;-J </span><span class="si">{</span><span class="n">jobs</span><span class="si">}</span><span class="s1"> -c </span><span class="si">{</span><span class="n">yml</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="kp">queue</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1">## Control of the workflow</span>
<span class="kn">parameter:</span> <span class="n">factor_option</span> <span class="o">=</span> <span class="s2">&quot;Marchenko_pc&quot;</span>
<span class="c1">### Association scan option (APEX vs TensorQTL)</span>
<span class="kn">parameter:</span> <span class="n">Association_option</span> <span class="o">=</span> <span class="s2">&quot;TensorQTL&quot;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-preprocessing">
<h2>Data Preprocessing<a class="headerlink" href="#data-preprocessing" title="Permalink to this headline">#</a></h2>
<section id="genotype-preprocessing-once-for-all-tissues">
<h3>Genotype Preprocessing (Once for all tissues)<a class="headerlink" href="#genotype-preprocessing-once-for-all-tissues" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[VCF_QC]
parameter: genoFile = paths
parameter: ref_fasta = path
parameter: add_chr = True
parameter: dbSNP = path
input: genoFile, group_by = 1
output: f&#39;{cwd}/data_preprocessing/genotype/{_input[0]:bnn}.{&quot;add_chr.&quot; if  add_chr else False}leftnorm.filtered.bed&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
       $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]//pipeline/VCF_QC.ipynb qc \
            --genoFile $[_input[0]] \
            --dbsnp-variants $[dbSNP] \
            --reference-genome $[ref_fasta] \
            --cwd $[_output:d] \
            --container $[container_base_bioinfo] \
            --walltime &quot;24h&quot; \
            $[submission if yml.is_file() else &quot;&quot; ] $[&quot;--add_chr&quot; if add_chr else &quot;--no-add_chr&quot;  ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[merge_plink]
parameter: genoFile = paths
skip_if(len(genoFile) == 1)
input: output_from(&quot;VCF_QC&quot;), group_by = &quot;all&quot;
parameter: name = &quot;xqtl_protocol_data.merged&quot;
output: f&quot;{cwd}/data_preprocessing/genotype//{name}.bed&quot;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
 $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]//pipeline/genotype_formatting.ipynb merge_plink \
            --genoFile $[_input:r] \
            --cwd $[_output:d] --name $[name] \
            --container $[container_base_bioinfo]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[plink_QC]
# minimum MAF filter to use. 0 means do not apply this filter.
parameter: maf_filter = 0
# minimum MAC filter to use. 0 means do not apply this filter.
parameter: mac_filter = 5
# maximum MAF filter to use. 0 means do not apply this filter.
parameter: maf_max_filter = 0.0
# Maximum missingess per-variant
parameter: geno_filter = 0.1
# Maximum missingness per-sample
parameter: mind_filter = 0.1
# HWE filter 
parameter: hwe_filter = 1e-06
# MAC filter
parameter: mac_filter = 5
parameter: genoFile = paths

input: (output_from(&quot;VCF_QC&quot;) if len(genoFile) == 1 else output_from(&quot;merge_plink&quot;))
output: f&#39;{_input:n}.filtered.bed&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
        $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]//pipeline/GWAS_QC.ipynb qc_no_prune \
            --cwd $[_output:d] \
            --genoFile $[_input] \
            --maf-filter $[maf_filter] \
            --mac_filter $[0 if maf_filter else mac_filter] \
            --geno-filter $[geno_filter] \
            --mind-filter $[mind_filter] \
            --hwe-filter $[hwe_filter]   \
            --mem  $[mem]  \
            --container $[container_base_bioinfo]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[plink_per_chrom]
input: output_from(&quot;plink_QC&quot;)
output: f&#39;{cwd}/data_preprocessing//genotype/{_input:bn}.plink_files_list.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
        $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]//pipeline/genotype_formatting.ipynb plink_by_chrom \
            --genoFile $[_input] \
            --cwd $[_output:d] \
            --chrom  `cut -f 1  $[_input:n].bim  | uniq | sed &quot;s/chr//g&quot;` \
            --container $[container_base_bioinfo]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[plink_to_vcf]
parameter: genotype_list = path
input: genotype_list
import pandas as pd
parameter: genotype_file_name = pd.read_csv(_input,&quot;\t&quot;,nrows = 1).values.tolist()[0][1]
output: f&#39;{cwd}/data_preprocessing//genotype/{path(genotype_file_name):bnn}.vcf_files_list.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
        $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]//pipeline/genotype_formatting.ipynb plink_to_vcf \
            --genoFile $[_input] \
            --cwd $[_output:d] \
            --container $[container_base_bioinfo]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[plink_per_gene]
# The plink genotype file
parameter: genoFile = path
input: output_from(&quot;region_list_concat&quot;),genoFile
output: f&#39;{cwd}/data_preprocessing/genotype/{_input[1]:bn}.plink_files_list.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
        $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline//genotype_formatting.ipynb plink_by_gene \
            --genoFile $[_input[1]] \
            --cwd $[_output:d] \
            --region_list  $[_input[0]] \
            --container $[container_base_bioinfo]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
<section id="molecular-phenotype-processing">
<h3>Molecular Phenotype Processing<a class="headerlink" href="#molecular-phenotype-processing" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[phenotype_partition_by_chrom]
parameter: region_list = path() 
input: for_each = &quot;input_inv&quot;
output: per_chrom_pheno_list = f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/phenotype_data/{path(_input_inv[&quot;molecular_pheno&quot;]):bn}.per_chrom.recipe&#39;        
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/phenotype_formatting.ipynb partition_by_chrom \
        --cwd $[_output:d]  \
        --phenoFile $[path(_input_inv[&quot;molecular_pheno&quot;]):a] \
        --region-list  $[region_list if region_list.is_file() else f&#39;&lt;(zcat {path(_input_inv[&quot;molecular_pheno&quot;]):a}  | cut -f 1,2,3,4)&#39;]  \
        --container $[container_rnaquant] \
        $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]       $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
<section id="genotype-processing">
<h3>Genotype Processing<a class="headerlink" href="#genotype-processing" title="Permalink to this headline">#</a></h3>
<p>Since genotype is shared among the eight tissue, the QC of whole genome file is not needed. Only pca needed to be run again.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sample_match]
parameter: sample_participant_lookup = path
input: for_each = &quot;input_inv&quot;
output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/{sample_participant_lookup:bn}.filtered.txt&#39;,
        geno = f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/{sample_participant_lookup:bn}.filtered_geno.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/sample_matcher.ipynb filtered_sample_list \
        --cwd $[_output[0]:d]  \
        --phenoFile $[_input_inv[&quot;molecular_pheno&quot;]]  \
        --genoFile $[path(_input_inv[&quot;genotype_file&quot;]):n].fam  \
        --sample-participant-lookup $[sample_participant_lookup] \
        --container $[container_base_bioinfo] \
        --translated_phenoFile $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[king]
parameter: sample_participant_lookup = path
parameter: maximize_unrelated = False
input:output_from(&quot;sample_match&quot;)[&quot;geno&quot;], group_with = &quot;input_inv&quot;
output: related = f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/genotype_data/{path(_input_inv[&quot;genotype_file&quot;]):bn}.{_input_inv[&quot;Theme&quot;]}.related.bed&#39;,
        unrelated = f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/genotype_data/{path(_input_inv[&quot;genotype_file&quot;]):bn}.{_input_inv[&quot;Theme&quot;]}.unrelated.bed&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
 $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/GWAS_QC.ipynb king \
    --cwd $[_output[0]:d] \
    --genoFile $[_input_inv[&quot;genotype_file&quot;]] \
    --name $[_input_inv[&quot;Theme&quot;]] \
    --keep-samples $[_input] \
    --container $[container_base_bioinfo] \
    --walltime 48h  $[submission if yml.is_file() else &quot;&quot; ] $[&quot;--maximize_unrelated&quot; if maximize_unrelated else &quot;--no-maximize_unrelated&quot;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[unrelated_QC]
input: output_from(&quot;king&quot;)[&quot;unrelated&quot;]
output: unrelated_bed = f&#39;{_input:n}.filtered.prune.bed&#39;, 
        prune = f&#39;{_input:n}.filtered.prune.in&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
 $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/GWAS_QC.ipynb qc \
    --cwd $[_output[0]:d] \
    --genoFile $[_input] \
    --exclude-variants /mnt/vast/hpc/csg/snuc_pseudo_bulk/Ast/genotype/dupe_snp_to_exclude \
    --mac-filter 5 \
    --container $[container_base_bioinfo] \
    $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]   $[submission if yml.is_file() else &quot;&quot; ] 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[related_QC]
input: output_from(&quot;king&quot;)[&quot;related&quot;],output_from(&quot;unrelated_QC&quot;)[&quot;prune&quot;]
output: f&#39;{_input[0]:n}.filtered.extracted.bed&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
 $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/GWAS_QC.ipynb qc_no_prune \
    --cwd $[_output[0]:d] \
    --genoFile $[_input[0]] \
    --maf-filter 0 \
    --geno-filter 0 \
    --mind-filter 0.1 \
    --hwe-filter 0 \
    --keep-variants $[_input[1]]  \
    --container $[container_base_bioinfo]  \
    $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]    $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="factor-analysis">
<h2>Factor Analysis<a class="headerlink" href="#factor-analysis" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[pca]
input: output_from(&quot;unrelated_QC&quot;)[&quot;unrelated_bed&quot;],group_with = &quot;input_inv&quot;
output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/pca/{_input:bn}.pca.rds&#39;,
        f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/pca/{_input:bn}.pca.scree.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
     $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/PCA.ipynb flashpca \
        --cwd $[_output:d] \
        --genoFile $[_input] \
        --container $[container_flashpca]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[projected_sample]
# The percentage of PVE explained
parameter: PVE_treshold = 0.7
input: output_from(&quot;related_QC&quot;),output_from(&quot;pca&quot;), group_with = &quot;input_inv&quot;
output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/pca/{_input[0]:bn}.pca.projected.rds&#39;,
        f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/pca/{_input[0]:bn}.pca.projected.scree.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/PCA.ipynb project_samples \
            --cwd $[_output:d] \
            --genoFile $[_input[0]] \
            --pca-model  $[_input[1]] \
            --maha-k `awk &#39;$3 &lt; $[PVE_treshold]&#39; $[_input[2]] | tail -1 | cut -f 1  ` \
            --container $[container_flashpca]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[merge_pca_covariate]
# The percentage of PVE explained
parameter: PVE_treshold = 0.7
input: output_from(&quot;projected_sample&quot;),group_with = &quot;input_inv&quot;
output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/covariates/{path(_input_inv[&quot;covariate_file&quot;]):bn}.pca.gz&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/covariate_formatting.ipynb merge_pca_covariate \
            --cwd $[_output:d] \
            --pcaFile $[_input[0]:a] \
            --covFile  $[path(_input_inv[&quot;covariate_file&quot;])] \
            --tol_cov 0.3  \
            --k `awk &#39;$3 &lt; $[PVE_treshold]&#39; $[_input[1]] | tail -1 | cut -f 1 ` \
            --container $[container_base_bioinfo] $[submission if yml.is_file() else &quot;&quot; ] --name $[_output:bn] --outliersFile $[_input[0]:an].outliers
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[resid_exp]
input: output_from(&quot;merge_pca_covariate&quot;),group_with = &quot;input_inv&quot;
output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/resid_phenotype/{path(_input_inv[&quot;molecular_pheno&quot;]):bnn}.cov_pca.resid.bed.gz&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;]  sos run $[exe_dir]/pipeline/covariate_formatting.ipynb compute_residual \
            --cwd $[_output:d] \
            --phenoFile $[path(_input_inv[&quot;molecular_pheno&quot;]):a] \
            --covFile  $[_input[0]:a] \
            --container $[container_base_bioinfo] $[submission if yml.is_file() else &quot;&quot; ] --name $[_output:bnnn]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[factor]
parameter: N = 0
input: output_from(&quot;resid_exp&quot;),output_from(&quot;merge_pca_covariate&quot;),group_with = &quot;input_inv&quot;
output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/covariates/{_input[0]:bnn}.{factor_option}.gz&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
         $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/Hidden_factor.ipynb $[factor_option] \
            --cwd $[_output:d] \
            --phenoFile $[_input[0]:a]  \
            --container $[container_PCAtools if factor_option == &quot;Marchenko_pc&quot; else container_PEER]  \
            --covFile  $[_input[1]:a] \
            --walltime 24h \
            --numThreads 8 \
            --iteration 1000 \
            --N $[N]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#[merge_factor_covariate]</span>
<span class="c1">## The percentage of PVE explained</span>
<span class="c1">#parameter: PVE_treshold = 0.7</span>
<span class="c1">#input: output_from(&quot;factor&quot;),output_from(&quot;merge_pca_covariate&quot;),group_with = &quot;input_inv&quot;</span>
<span class="c1">#output: f&#39;{cwd}/data_preprocessing/{_input_inv[&quot;Theme&quot;]}/covariates/{_input[0]:bn}.cov.gz&#39;</span>
<span class="c1">#script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;</span>
<span class="c1">#    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/covariate_formatting.ipynb merge_factor_covariate \</span>
<span class="c1">#            --cwd $[_output:d] \</span>
<span class="c1">#            --factorFile $[_input[0]:a] \</span>
<span class="c1">#            --covFile  $[_input[1]:a] \</span>
<span class="c1">#            --container $[container_base_bioinfo] $[submission if yml.is_file() else &quot;&quot; ] --name $[_output:bn]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="association-scan">
<h2>Association Scan<a class="headerlink" href="#association-scan" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[TensorQTL]
# The number of minor allele count as treshold for the analysis
parameter: MAC = 0
# The minor allele frequency as treshold for the analysis, overwrite MAC
parameter: maf_threshold = 0
parameter: genotype_list = path
parameter: cis_window = 1000000
parameter: mem = &quot;50G&quot;
input: genotype_list, output_from(&quot;phenotype_partition_by_chrom&quot;),output_from(&quot;factor&quot;),group_with = &quot;input_inv&quot;
output: f&#39;{cwd:a}/association_scan/{_input_inv[&quot;Theme&quot;]}/TensorQTL/TensorQTL.cis._recipe.tsv&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/TensorQTL.ipynb cis \
        --genotype-file $[_input[0]]   \
        --phenotype-file  $[_input[1]] \
        --region-list  &lt;(cat $[_input[1]] | grep -v  &quot;\.1\.&quot; | grep -v chrX | grep -v chrY | grep -v chrM | cut -f 1,2,3,4 )  \
        --covariate-file $[_input[2]]    \
        --cwd $[_output:d]  \
        --window $[cis_window] \
         $[submission if yml.is_file() else &quot;&quot; ] $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ] \
         $[f&#39;--MAC {MAC}&#39; if MAC else &quot;&quot;]  \
         $[f&#39;--maf_threshold {maf_threshold}&#39; if maf_threshold else &quot;&quot;] \
         $[f&#39;--phenotype_group {_input_inv[&quot;phenotype_group&quot;]}&#39; if _input_inv[&quot;phenotype_group&quot;] else &quot;&quot; ] \
        --container $[container_TensorQTL]  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[APEX]
parameter: genotype_list = path
input: output_from(&quot;plink_to_vcf&quot;), output_from(&quot;phenotype_partition_by_chrom&quot;),output_from(&quot;merge_factor_covariate&quot;),group_with = &quot;input_inv&quot;
output: f&#39;{cwd:a}/association_scan/{_input_inv[&quot;Theme&quot;]}/APEX/APEX_QTL_recipe.tsv&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/APEX.ipynb cis \
        --genotype-list $[_input[0]]   \
        --phenotype-list $[_input[1]] \
        --covariate-file $[_input[2]]    \
        --cwd $[_output:d]  \
        --container $[container_apex] $[submission if yml.is_file() else &quot;&quot; ] --name $[_input[1]:bnn]
</pre></div>
</div>
</div>
</div>
</section>
<section id="trans-association-scan">
<h2>Trans Association Scan<a class="headerlink" href="#trans-association-scan" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[TensorQTL_Trans]
parameter: MAC = 0
# The minor allele frequency as treshold for the analysis, overwrite MAC
parameter: maf_threshold = 0
parameter: genotype_list = path
parameter: region_list = path
input: genotype_list, output_from(&quot;phenotype_partition_by_chrom&quot;),output_from(&quot;merge_factor_covariate&quot;),group_with = &quot;input_inv&quot;
output: f&#39;{cwd:a}/association_scan/{_input_inv[&quot;Theme&quot;]}/Trans/TensorQTL.trans._recipe.tsv&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/TensorQTL.ipynb trans \
        --genotype-list $[_input[0]]   \
        --phenotype-list $[_input[1]] \
        --covariate-file $[_input[2]]    \
        --cwd $[_output:d]  \
        --region_list $[region_list] \
        --container $[container_TensorQTL] $[submission if yml.is_file() else &quot;&quot; ] $[f&#39;--MAC {MAC}&#39; if MAC else &quot;&quot;] $[f&#39;--maf_threshold {maf_threshold}&#39; if maf_threshold else &quot;&quot;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="sumstat-merger">
<h2>Sumstat Merger<a class="headerlink" href="#sumstat-merger" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[yml_generation]
parameter: TARGET_list = path(&quot;./&quot;)
input: output_from(Association_option), group_by = &quot;all&quot;
output: f&#39;{cwd:a}/data_intergration/{Association_option}/qced_sumstat_list.txt&#39;,f&#39;{cwd:a}/data_intergration/{Association_option}/yml_list.txt&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.step_{step_name}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
        $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run $[exe_dir]/pipeline/yml_generator.ipynb yml_list \
            --sumstat-list $[_input] \
            --cwd  $[_output[1]:d] --name $[&quot; &quot;.join([str(x).split(&quot;/&quot;)[-3] for x in _input])]  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[target_generation]
parameter: fasta = &quot;reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta&quot;
parameter: mem = &quot;50G&quot;
input: output_from(&quot;yml_generation&quot;)
output: f&#39;{_input[0]:d}/TARGET.ref.list&#39;
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{step_name}.step_{step_name}.stderr&#39;, stdout = f&#39;{step_name}.stdout&#39;
      $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run  $[exe_dir]/pipeline/summary_stats_standardizer.ipynb TARGET_generation  \
            --fasta $[fasta] \
            --sumstat-list $[_input[0]]    \
            --yml-list $[_input[1]]    \
            --cwd $[_input[0]:d]  $[submission if yml.is_file() else &quot;&quot; ] $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]  --walltime 48h --container $[container_base_bioinfo]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sumstat_standardizer]
parameter: mem = &quot;50G&quot;
input: output_from(&quot;yml_generation&quot;), output_from(&quot;target_generation&quot;)
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{step_name}.step_{step_name}.stderr&#39;, stdout = f&#39;{step_name}.stdout&#39;
      $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run  $[exe_dir]/pipeline/summary_stats_standardizer.ipynb sumstat_standardization  \
            --sumstat-list $[_input[0]]    \
            --yml-list $[_input[1]]    \
            --TARGET_list $[_input[2]] \
            --cwd $[_input[0]:d]  $[submission if yml.is_file() else &quot;&quot; ] $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]  --walltime 48h --container $[container_base_bioinfo]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sumstat_to_vcf]
parameter: mem = &quot;50G&quot;
input: output_from(&quot;yml_generation&quot;)
depends: sos_step(&#39;sumstat_standardizer&#39;)
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{step_name}.step_{step_name}.stderr&#39;, stdout = f&#39;{step_name}.stdout&#39;
      $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run  $[exe_dir]/pipeline/summary_stats_standardizer.ipynb sumstat_to_vcf  \
            --sumstat-list $[_input[0]]    \
            --cwd $[_input[0]:d]  $[submission if yml.is_file() else &quot;&quot; ] $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]  --walltime 48h --container $[container_base_bioinfo]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[assoc_result_proc]
input: output_from(&quot;yml_generation&quot;)
depends: sos_step(&#39;sumstat_to_vcf&#39;)
script: interpreter = interpreter, expand = &quot;$[ ]&quot;, stderr = f&#39;{step_name}.step_{step_name}.stderr&#39;, stdout = f&#39;{step_name}.stdout&#39;
    $[&quot;monitor.py&quot; if monitor else &quot;&quot;] sos run  $[exe_dir]/pipeline/assoc_result_processing.ipynb genome  \
        --vcf `ls $[_input[0]:d]/*/*cis_long_table.vcf` \
        --padjust-method &quot;bonferroni&quot;      \
        --container $[container_base_bioinfo] \
        $[f&#39;--mem {mem}&#39; if yml.is_file() else &quot;&quot; ]   $[submission if yml.is_file() else &quot;&quot; ] 
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/commands_generator"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="bulk_expression_commands.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">RNA-seq calling and QC</p>
      </div>
    </a>
    <a class="right-next"
       href="../data_preprocessing/reference_data.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Reference data standardization</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype">Genotype</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#per-tissue-analysis">Per tissue analysis</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#protocol-data">Protocol data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-for-running-the-workflow">Example for running the workflow</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preprocessing">Data Preprocessing</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-preprocessing-once-for-all-tissues">Genotype Preprocessing (Once for all tissues)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#molecular-phenotype-processing">Molecular Phenotype Processing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-processing">Genotype Processing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#factor-analysis">Factor Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#association-scan">Association Scan</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trans-association-scan">Trans Association Scan</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sumstat-merger">Sumstat Merger</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>