

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Sample level RNA-seq quality control &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/molecular_phenotypes/QC/bulk_expression_QC';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Bulk RNA-seq counts normalization" href="bulk_expression_normalization.html" />
    <link rel="prev" title="Quantifying expression from RNA-seq data" href="../calling/RNA_calling.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../bulk_expression.html">RNA-seq expression</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/molecular_phenotypes/QC/bulk_expression_QC.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/molecular_phenotypes/QC/bulk_expression_QC.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/molecular_phenotypes/QC/bulk_expression_QC.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Sample level RNA-seq quality control</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gtex-v8-qc-procedures">GTEx V8 QC procedures</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#our-implementation">Our implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#three-statistics-to-identify-outliers">Three statistics to identify outliers</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rle">RLE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hierarchical-clustering">Hierarchical clustering</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#d-statistic">D-statistic</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="sample-level-rna-seq-quality-control">
<h1>Sample level RNA-seq quality control<a class="headerlink" href="#sample-level-rna-seq-quality-control" title="Permalink to this headline">#</a></h1>
<p>This pipeline implements RNA-seq QC that identifies and removes genes and samples from the TPM matrix.</p>
<section id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">#</a></h2>
<section id="gtex-v8-qc-procedures">
<h3>GTEx V8 QC procedures<a class="headerlink" href="#gtex-v8-qc-procedures" title="Permalink to this headline">#</a></h3>
<p>From <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods</a> :</p>
<ol class="arabic simple">
<li><p>RNA-seq expression outliers were identified and excluded using a multidimensional extension of the statistic described in (Wright et al., Nat. Genet. 2014 ). Briefly, for each tissue, read counts from each sample were normalized using size factors calculated with DESeq2 and log-transformed with an offset of 1; genes with a log-transformed value &gt;1 in &gt;10% of samples were selected, and the resulting read counts were centered and unit-normalized. The resulting matrix was then hierarchically clustered (based on average and cosine distance), and a chi2 p-value was calculated based on Mahalanobis distance. Clusters with ≥60% samples with Bonferroni-corrected p-values &lt;0.05 were marked as outliers, and their samples were excluded.</p></li>
<li><p>Samples with &lt;10 million mapped reads were removed.</p></li>
<li><p>Genes were selected based on expression thresholds of &gt;0.1 TPM in at least 20% of samples and ≥6 reads in at least 20% of samples.</p></li>
</ol>
</section>
<section id="our-implementation">
<h3>Our implementation<a class="headerlink" href="#our-implementation" title="Permalink to this headline">#</a></h3>
<p>For #1 and #2 above because the GTEx source code is not available, we adopted <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04307-0">some R scripts developed in this paper</a> to perform gene and sample QC using TPM data, which we implements GTEx recommendations. #3 above is implemented in the gene expression normalization module.</p>
</section>
</section>
<section id="three-statistics-to-identify-outliers">
<h2>Three statistics to identify outliers<a class="headerlink" href="#three-statistics-to-identify-outliers" title="Permalink to this headline">#</a></h2>
<section id="rle">
<h3>RLE<a class="headerlink" href="#rle" title="Permalink to this headline">#</a></h3>
<p>RLE(Relative Log Expression)is assume that most expressions in a sample should be near a mean value, only a few genes have differential expression, which means higher or lower than other genes. To calculate it, for each gene <em>i</em> , calculate it median in <em>N</em> sample as <em>Medi</em> , for each sample <em>j</em> , and expression value <em>eij</em>, count the difference between <em>eij</em> and <em>Medi</em> : <em>deij = eij-Medi</em> , than boxplot each sample base on <em>deij</em> and sort by IQR(interquartile range), the sample with lager IQR is more likely to be an outlier.</p>
</section>
<section id="hierarchical-clustering">
<h3>Hierarchical clustering<a class="headerlink" href="#hierarchical-clustering" title="Permalink to this headline">#</a></h3>
<p>We use(1 - spearman’s correlation) as distances between samples and assume samples should be homogeneous, so all samples should have short distances between others and clustered homogeneously. So samples far away others may be outliers. So we first use distances which is 1-spearmen correlation to hierarchically clustered, than use top 100 gene sort by variance to calculate Mahalanobis Distance, then a chi2 p-value will be calculated based on mahalanobis distance. Then clusters with <span class="math notranslate nohighlight">\( \geq 60\%\)</span> samples with Bonferroni-corrected p-values &lt; 0.05 were marked as outliers.</p>
</section>
<section id="d-statistic">
<h3>D-statistic<a class="headerlink" href="#d-statistic" title="Permalink to this headline">#</a></h3>
<p>For each sample, D-statistic represent the average correlation between its expression and other samples. A sample with lower D-statistic is more likely to be an outlier.</p>
</section>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">#</a></h2>
<p>The MWE data can be found on <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk">Google Drive</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">bulk_expression_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">tpm</span><span class="o">-</span><span class="n">gct</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">TPM</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">counts</span><span class="o">-</span><span class="n">gct</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">Counts</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">log</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;Gene expression profiles loaded successfully!&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;19 genes and 1118 samples are loaded from data/mwe.TPM.gct&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;0 genes are filtered, because &gt; 20 </span><span class="si">% s</span><span class="s2">amples have expression values &lt; 0.1&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;19 genes left, saving output.&quot;</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">log</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">ls</span> <span class="n">output</span><span class="o">/*.</span><span class="n">pdf</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">cluster</span><span class="o">.</span><span class="n">pdf</span>  <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">D_stat_hist</span><span class="o">.</span><span class="n">pdf</span>  <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">preQC_cluster</span><span class="o">.</span><span class="n">pdf</span>  <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">RLEplot</span><span class="o">.</span><span class="n">pdf</span>
</pre></div>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">bulk_expression_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run bulk_expression_QC.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  qc

Global Workflow Options:
  --tpm-gct VAL (as path, required)
                        Required input is TPM file
  --counts-gct . (as path)
                        Raw counts file is optional and if available, it will be
                        filtered to match with the TPM file sample and genes
  --cwd output (as path)
  --container &#39;&#39;

Sections
  qc_1:
    Workflow Options:
      --low-expr-TPM 0.1 (as float)
      --low-expr-TPM-percent 0.2 (as float)
  qc_2:
    Workflow Options:
      --RLEFilterPercent 0.05 (as float)
      --DSFilterPercent 0.05 (as float)
      --topk-genes 100 (as int)
      --cluster-percent 0.6 (as float)
      --pvalue-cutoff 0.05 (as float)
      --cluster-level 5 (as int)
  qc_3:
</pre></div>
</div>
</div>
</div>
</section>
<section id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Required input is TPM file</span>
<span class="kn">parameter:</span> <span class="n">tpm_gct</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Raw counts file is optional and if available, it will be filtered to match with the TPM file sample and genes</span>
<span class="kn">parameter:</span> <span class="n">counts_gct</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>

<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[qc_1 (basic check and low expression filtering)]
parameter: low_expr_TPM = 0.1
parameter: low_expr_TPM_percent = 0.2
input: tpm_gct
output: f&#39;{cwd}/{_input:bnnn}.low_expression_filtered.tpm.gct.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
R: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output:nnn}.stderr&#39;, stdout = f&#39;{_output:nnn}.log&#39;,container = container 
    # Load packages
    library(dplyr)
    library(readr)
    library(tibble)
    library(purrr)
    ## Setting inputs
    low_expr_TPM &lt;- $[low_expr_TPM]
    low_expr_TPM_percent &lt;- $[low_expr_TPM_percent]
    ## QC
    ### Data integrity check. The input file should has sample ID as column name; the first column is gene ID
    TPM_data &lt;- read_tsv($[_input:ar], col_names = TRUE, comment = &quot;#&quot;)%&gt;%as.data.frame()
    names(TPM_data)[1] &lt;- &quot;feature&quot;
    if( sum(duplicated(TPM_data$feature)) &gt; 0){
      message(&quot;feature (e.g. gene names) should be in the first column. Please remove duplicated feature IDs, Exit!&quot;)
      quit(save = &quot;no&quot;, status = 1, runLast = FALSE)
    }else{
      rownames(TPM_data) &lt;- TPM_data$feature
      TPM_data = TPM_data[,-1]
      loaded_sample_count &lt;- ncol(TPM_data)
    }
    #### NA check
    if(sum(is.na(TPM_data)) &gt; 0 ){      
      message(paste0(&quot;NA is not allowed in the data, there are &quot;,sum(is.na(TPM_data)),&quot; NAs, Exit!&quot;))
      quit(save = &quot;no&quot;, status = 1, runLast = FALSE)}
    #### make sure every expr column is in numeric type
    matrix_check &lt;- map(TPM_data, is.numeric) %&gt;% unlist
    if(sum(!matrix_check) &gt; 0){
      message(&quot;The following column(s) in expression matrix is/are NOT in numeric type. Plese check, Proceed by excluding those samples&quot;)
      message(paste(names(matrix_check)[!matrix_check], collapse = &quot;; &quot;))
      TPM_data = TPM_data[,matrix_check]
    }
    message(&quot;Gene expression profiles loaded successfully!&quot;)
    message(paste(nrow(TPM_data), &quot;genes and&quot;, ncol(TPM_data), &quot;samples are loaded from&quot;, $[_input:r], sep = &quot; &quot;))
    
    #### Filter out low exp genes
    keep_genes_idx &lt;- (rowMeans(TPM_data&gt;low_expr_TPM)&gt;low_expr_TPM_percent) 
    TPM_data = TPM_data[keep_genes_idx,]
    message(paste(sum(1 - keep_genes_idx), &quot;genes are filtered by filter: &gt;&quot;, low_expr_TPM_percent*100, &quot;% samples have expression values &lt;&quot;, low_expr_TPM))
    message(paste(sum(keep_genes_idx), &quot;genes left, saving output to $[_output].&quot;))
    TPM_data%&gt;%as_tibble(rownames = &quot;gene_ID&quot;)%&gt;%write_delim(&quot;$[_output]&quot;,&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
<p><strong>CAUTION: I (GW) am not sure what to do with the offset for log-transformation on TPM. GTEx suggests using offset = 1. Here I keep the recommendation from <a class="reference external" href="https://github.com/stormlovetao/eQTLQC/blob/master/Sample/src/report.Rmd">these authors</a> where both 0.0001 and 1 were used in different steps.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[qc_2 (remove outliers)]
parameter: RLEFilterPercent = 0.05
parameter: DSFilterPercent = 0.05
parameter: topk_genes = 100
parameter: cluster_percent = 0.6
parameter: pvalue_cutoff = 0.05
parameter: cluster_level = 5
output: f&#39;{cwd}/{_input:bnnn}.outlier_removed.tpm.gct.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
R: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output:nnn}.stderr&#39;, stdout = f&#39;{_output:nnn}.log&#39;, container = container 
    library(RColorBrewer)
    library(ape)
    library(reshape2)
    library(dplyr)
    library(readr)
    ## Setting parameters
    RLEFilterPercent &lt;-$[RLEFilterPercent]
    DSFilterPercent &lt;- $[DSFilterPercent]
    ### Hcluster parameter
    pvalues.cut &lt;- $[pvalue_cutoff]
    topk_genes &lt;- $[topk_genes]
    cluster_percent &lt;- $[cluster_percent]
    treesNum &lt;- $[cluster_level]
    ## Define functions to accomodate rank deficient covariance matrices in https://github.com/cumc/xqtl-pipeline/issues/307
    mahalanobis = function (x, center, cov, inverted = FALSE, ...) 
        {
            x &lt;- if (is.vector(x)) 
                matrix(x, ncol = length(x))
            else as.matrix(x)
            if (!isFALSE(center)) 
                x &lt;- sweep(x, 2L, center)
            if (!inverted) 
                cov &lt;- MASS::ginv(cov, ...)
            setNames(rowSums(x %*% cov * x), rownames(x))
        }
          
    ## laod Data
    TPM_data &lt;- read_tsv($[_input:ar], col_names = TRUE, comment = &quot;#&quot;)%&gt;%as.data.frame()
    rownames(TPM_data) &lt;- TPM_data$gene_ID
    TPM_data = TPM_data[,-1]
    RLEFilterLength &lt;- RLEFilterPercent*ncol(TPM_data)
    DSFilter &lt;- DSFilterPercent*ncol(TPM_data)
  
    ## Outlier detection
  
    ### RLE
    # https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L71
    log_offset &lt;- 0.0001
    logtpm = log10(TPM_data%&gt;%as.matrix + log_offset)
    rle=logtpm-apply(logtpm, 1, median) # change &quot;/&quot; to &quot;-&quot; so that we got log(fold-change) which centered on 0 on the RLE plot.
    iqr = apply(rle,2,IQR)
    rle=melt( rle , variable.name = &quot;Sample&quot;,value.name =&quot;TPM&quot;, id=&quot;feature&quot;)
    names(rle) &lt;- c(&quot;feature&quot;,&quot;Sample&quot;,&quot;TPM&quot;)
    rle_IQR &lt;- rle %&gt;% group_by(Sample) %&gt;% summarise(IQR = IQR(TPM))
    rle_IQR_range &lt;- rle_IQR$IQR %&gt;% range %&gt;% abs() %&gt;% max()
    rle_IQR_range &lt;- 2*rle_IQR_range %&gt;% ceiling()
    bymedian &lt;- with(rle, reorder(Sample, TPM, IQR))  # sort by IQR
    par(mar=c(3,3,3,3))
    pdf(file = &quot;$[_output:n].RLEplot.pdf&quot;)
    boxplot(TPM ~ bymedian, data=rle, outline=F, ylim = c(-rle_IQR_range, rle_IQR_range), las=2, boxwex=1, col=&#39;gray&#39;, cex.axis=0.3, main=&quot;RLE plot before QC&quot;, xlab=&quot;&quot;, ylab=&quot;Residual expression levels&quot;, frame=F)
    dev.off()
    ExpPerSample &lt;- nrow(TPM_data)
    RLEFilterList &lt;- unique(levels(bymedian)[((length(levels(bymedian))-(RLEFilterLength))+1):(length(levels(bymedian))+1)])
    RLEFilterList &lt;- as.character(RLEFilterList)
    message(paste0(&quot;The right most &quot;, RLEFilterPercent*100, &quot;% samples (N = &quot;, length(RLEFilterList), &quot;) are marked as candidate outliers in this step:&quot;) )
    for (x in 1:length(RLEFilterList)){
      message(RLEFilterList[x])
    }

    ### hcluster  
    sampleDists &lt;- 1 - cor(logtpm, method=&#39;spearman&#39;)
    hc &lt;- hclust(as.dist(sampleDists), method = &quot;complete&quot;)
    hcphy &lt;- as.phylo(hc)
  
    pdf(file = &quot;$[_output:n].preQC_cluster.pdf&quot;)
    plot(hcphy, type = &quot;unrooted&quot;, cex=.2, lab4ut=&#39;axial&#39;,underscore = T, main=&quot;Sample clustering before QC (Spearman - Cor.)&quot;)
    dev.off()
    # https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L102
    log_offset &lt;- 1
    logtpm = log10(TPM_data%&gt;%as.matrix + log_offset)
    ntop &lt;- topk_genes
    Pvars &lt;- apply(logtpm, 1, var)
    select &lt;- order(Pvars, decreasing =TRUE)[seq_len(min(ntop, length(Pvars)))]
    MD_matrix &lt;- logtpm[select, ]
    MahalanobisDistance = mahalanobis(t(MD_matrix), colMeans(t(MD_matrix)), cov(t(MD_matrix))) 
    # Note: t(MD_matrix)=sample_row*gene_column, Manhalanobis() returns one vector with length=row number
    pvalues = pchisq(MahalanobisDistance, df=nrow(MD_matrix), lower.tail=F)
    pvalues.adjust = p.adjust(pvalues, method =&quot;bonferroni&quot;) # adjusted pvalues for each sample
    pvalues.low &lt;- pvalues.adjust[pvalues.adjust&lt;pvalues.cut]
    
    HCoutliers &lt;- character()
    for(x in c(1:treesNum)){
      trees &lt;- cutree(hc,k=x)
      idx &lt;- c(1:x)#which tree is checking
      for(i in idx)
      {
        group &lt;- hc$labels[which(trees == i)]
        if(sum(group %in% names(pvalues.low))/length(group) &gt;= cluster_percent)
        {
          HCoutliers &lt;- union(HCoutliers,group)
        }
      }
    }
    
    message(paste(length(HCoutliers), &quot;samples are marked as candidate outlier(s) in this step.&quot;, sep = &quot; &quot;))
    if(length(HCoutliers)&gt;0){
      message(&quot;Sample outliers are marked in red as follows:&quot;)
      for (x in 1:length(HCoutliers)){
          message(HCoutliers[x])
      }
      co1 = hc$labels%in%HCoutliers
      co1[which(co1 == &quot;FALSE&quot;)]&lt;-&quot;gray0&quot;
      co1[which(co1 == &quot;TRUE&quot;)]&lt;-&quot;red&quot;
      par(mar=c(3,3,3,3))
  
    pdf(file = &quot;$[_output:n].cluster.pdf&quot;)
    plot(hcphy, tip.col = co1, type = &quot;unrooted&quot;, cex=.2, lab4ut=&#39;axial&#39;,underscore = T, main=&quot;Label Outliers in Red&quot;)
    Xcol = c(&quot;gray0&quot;, &quot;red&quot;)
    Xtext = c(&quot;Normal Sample&quot;, &quot;Outliers&quot;)
    legend(&#39;bottomleft&#39;,pch=21,Xtext, col=&#39;white&#39;,pt.bg=Xcol, cex=1)
    dev.off()
    }else{
      message(&quot;No outlier detected.&quot;)
    }
    
    ### D-s
    D = apply(1-sampleDists, 1, median)
    pdf(file = &quot;$[_output:n].D_stat_hist.pdf&quot;)
    hist(D, breaks=100, ylab=&quot;Number of samples&quot;, xlab=&quot;D-statistic&quot;, main=&quot;Histogram of Sample D-statistics before data QC&quot;)
    dev.off()
    
    DSFilter &lt;- sort(D)[DSFilter]
    D&lt;-as.data.frame(D)
    D&lt;-data.frame(Sample = rownames(D),D = D$D)
    D_filterList = D%&gt;%filter(D &lt;= DSFilter)
    D_filterList &lt;- D_filterList$Sample
    D_filterList&lt;-as.character(D_filterList)
    message(paste0(&quot;The right most &quot;, DSFilterPercent*100, &quot;% samples (N=&quot;, length(D_filterList), &quot;) are marked as candidate outliers in this step:&quot;) )
    for (x in 1:length(D_filterList)){
      message(D_filterList[x])
    }
    
    ## Outliers are the intersect of three candidates list
    outliersList &lt;- intersect(RLEFilterList,intersect(HCoutliers,D_filterList))
    message(&quot;Outliers:&quot;)
    for (x in 1:length(outliersList)){
      message(outliersList[x])
    }
    outliersIndex &lt;- which(colnames(logtpm) %in% outliersList)
    if(!length(outliersIndex) == 0){
        TPM_data &lt;- TPM_data[,-outliersIndex]
    }
    ## Add 2 header lines, https://github.com/getzlab/rnaseqc/blob/286f99dfd4164d33014241dd4f3149da0cddf5bf/src/RNASeQC.cpp#L426
    cat(paste(&quot;#1.2\n#&quot;, nrow(TPM_data), ncol(TPM_data), &quot;\n&quot;), file=$[_output:nr], append=FALSE)
    TPM_data%&gt;%as_tibble(rownames = &quot;gene_ID&quot;)%&gt;%write_delim($[_output:nr],delim = &quot;\t&quot;,col_names = T, append = T)
  
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output:nnn}.stderr&#39;, stdout = f&#39;{_output:nnn}.log&#39;, container = container
    gzip -f --best $[_output:n]
</pre></div>
</div>
</div>
</div>
<p>Additional formatting:</p>
<ol class="arabic simple">
<li><p>Filter out the geneCount table based on TPM table.</p></li>
<li><p>Adds two comment lines above the header of TPM and geneCount table to mimick the original output from RNASeQC.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[qc_3 (remove gene and samples from raw counts)]
stop_if(not counts_gct.is_file())
output: f&#39;{cwd}/{_input:bnnn}.geneCount.gct.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
R: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output:nn}.stderr&#39;, stdout = f&#39;{_output:nn}.stdout&#39;, container = container
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    # Reason to use read.table: 1.accomodate both &quot; &quot; and &quot;\t&quot;
    tpm = read_delim($[_input:ar], &quot;\t&quot;, col_names = T, comment = &quot;#&quot;)
    geneCount = read_delim($[counts_gct:ar], &quot;\t&quot; ,col_names = T, comment = &quot;#&quot;)
    ## Make geneCount consistant with tpm, remove gene by filter() and remove samples by select()
    geneCount = geneCount%&gt;%filter(gene_ID %in% tpm$gene_ID)%&gt;%select(colnames(tpm))
    ## Add 2 header lines, https://github.com/getzlab/rnaseqc/blob/286f99dfd4164d33014241dd4f3149da0cddf5bf/src/RNASeQC.cpp#L426
    cat(paste(&quot;#1.2\n#&quot;, nrow(geneCount), ncol(geneCount) - 1, &quot;\n&quot;), file=$[_output:nr], append=FALSE)
    ## Save each file with 3 header line
    geneCount%&gt;%write_delim($[_output:nr],delim = &quot;\t&quot;,col_names = T, append = T)

bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output:nnn}.stderr&#39;, stdout = f&#39;{_output:nnn}.log&#39;, container = container
    gzip -f --best $[_output:n]
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/molecular_phenotypes/QC"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../calling/RNA_calling.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Quantifying expression from RNA-seq data</p>
      </div>
    </a>
    <a class="right-next"
       href="bulk_expression_normalization.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Bulk RNA-seq counts normalization</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gtex-v8-qc-procedures">GTEx V8 QC procedures</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#our-implementation">Our implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#three-statistics-to-identify-outliers">Three statistics to identify outliers</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rle">RLE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hierarchical-clustering">Hierarchical clustering</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#d-statistic">D-statistic</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>