

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Fine-mapping with SuSiE RSS model &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/fine_mapping/SuSiE/SuSiE_RSS';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Fine-mapping with SuSiE model" href="SuSiE.html" />
    <link rel="prev" title="Univariate fine-mapping workflows" href="../univariate_fine_mapping.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/fine_mapping/SuSiE/SuSiE_RSS.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/fine_mapping/SuSiE/SuSiE_RSS.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/fine_mapping/SuSiE/SuSiE_RSS.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Fine-mapping with SuSiE RSS model</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mwe">MWE</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="fine-mapping-with-susie-rss-model">
<h1>Fine-mapping with SuSiE RSS model<a class="headerlink" href="#fine-mapping-with-susie-rss-model" title="Permalink to this headline">#</a></h1>
<p>This notebook take a list of LD file and a list of sumstat file and do salmon QC and susie RSS for each overlap LD block.</p>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p>A tab delimated table describing the path where LD per region stored, can be generated using the ld_per_region_plink step of the genotype processing module.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#id     dir</span>
<span class="n">chr17_60570445_65149278</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">LD</span><span class="o">/</span><span class="n">output_npz_2</span><span class="o">/</span><span class="mi">1300</span><span class="n">_hg38_EUR_LD_blocks_npz_files</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">chr17_60570445_65149278</span><span class="o">.</span><span class="n">flt16</span><span class="o">.</span><span class="n">npz</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>A tab delimated table describing  path where summary stat per chromosome stored, can be generated using the yml_generator module before the qced sumstat are generated.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>hs3163@csglogin:/mnt/vast/hpc/csg/xqtl_workflow_testing/susie_rss$ cat /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/qced_sumstat_list.txt
#chr    ADGWAS_Bellenguez_2022
1       /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/ADGWAS_Bellenguez_2022.1/ADGWAS2022.chr1.sumstat.tsv
2       /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/ADGWAS_Bellenguez_2022.2/ADGWAS2022.chr2.sumstat.tsv
3       /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/ADGWAS_Bellenguez_2022.3/ADGWAS2022.chr3.sumstat.tsv
4       /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/ADGWAS_Bellenguez_2022.4/ADGWAS2022.chr4.sumstat.tsv
5       /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/ADGWAS_Bellenguez_2022.5/ADGWAS2022.chr5.sumstat.tsv


</pre></div>
</div>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p>A RDS file containing the output susie object, the name of all variants that went through the analysis, the z score , and the LD used for the analysis.</p></li>
<li><p>A sumstat file with additional column containing the slalom results.</p></li>
</ol>
</section>
<section id="mwe">
<h2>MWE<a class="headerlink" href="#mwe" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE_RSS</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">SuSiE_RSS</span> \
    <span class="o">--</span><span class="n">LD_list</span> <span class="n">test</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">list</span> \
    <span class="o">--</span><span class="n">sumstat_list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">ADGWAS</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">ADGWAS2022</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">stephenslab</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">impute</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">output_impute_2</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">  File</span><span class="nn"> &quot;&lt;ipython-input-9-158593d50ce7&gt;&quot;</span><span class="gt">, line </span><span class="mi">1</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE_RSS</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">SuSiE_RSS</span> \
        <span class="o">^</span>
<span class="ne">SyntaxError</span>: invalid syntax
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="n">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="n">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">3</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># getting the overlapped input</span>
<span class="kn">parameter:</span> <span class="n">LD_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">sumstat_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">LD_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">LD_list</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">sumstat_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">LD_list</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="n">LD_list</span><span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span> <span class="p">]</span>
<span class="n">sumstat_list</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="n">sumstat_list</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">]</span> <span class="p">]</span>
<span class="n">input_inv</span> <span class="o">=</span> <span class="n">LD_list</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">)</span>
<span class="n">input_list</span> <span class="o">=</span> <span class="n">input_inv</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">lead_idx_choice</span> <span class="o">=</span> <span class="s2">&quot;pvalue&quot;</span>
<span class="kn">parameter:</span> <span class="n">abf_prior_variance</span> <span class="o">=</span> <span class="mf">0.4</span>
<span class="kn">parameter:</span> <span class="n">nlog10p_dentist_s_threshold</span> <span class="o">=</span> <span class="mi">4</span>
<span class="kn">parameter:</span> <span class="n">r2_threshold</span> <span class="o">=</span> <span class="mf">0.6</span>
<span class="kn">parameter:</span> <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
<span class="kn">parameter:</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="kn">parameter:</span> <span class="n">impute</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># Whether to impute the sumstat for all the snp in LD but not in sumstat.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>[SuSiE_RSS_1]
parameter: L = 10
parameter: max_L = 1000
input: input_list, group_by = 2
name = f&#39;{_input[0]:b}&#39;.split(&quot;.&quot;)[-3]
output: f&#39;{cwd:a}/{_input[1]:bn}.{name}.unisusie_rss.fit.rds&#39;,
        f&#39;{cwd:a}/{_input[1]:bn}.{name}.unisusie_rss.ss_qced.tsv&#39;

task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    import pandas as pd
    import numpy as np
    import rpy2
    from backports import zoneinfo
    import rpy2.robjects.numpy2ri as numpy2ri
    from rpy2.robjects import pandas2ri
    from rpy2.robjects.packages import SignatureTranslatedAnonymousPackage
    import rpy2.robjects as ro
    numpy2ri.activate()
    pandas2ri.activate()
    def load_npz_ld(path):
        np_ld_loaded = np.load(path,allow_pickle=True)
        # sort by start position
        snp_id = [x.replace(&quot;:&quot;,&quot;_&quot;) for x in np_ld_loaded.get(&quot;arr_1&quot;)]
        np_ld_loaded = np_ld_loaded.get(&quot;arr_0&quot;)
        new = np_ld_loaded + np_ld_loaded.T
        np.fill_diagonal(new, np.diag(new)/2)
        return new,snp_id

    def get_bcor_meta(bcor_obj):
        df_ld_snps = bcor_obj.getMeta()
        df_ld_snps.rename(columns={&#39;rsid&#39;:&#39;SNP&#39;, &#39;position&#39;:&#39;BP&#39;, &#39;chromosome&#39;:&#39;CHR&#39;, &#39;allele1&#39;:&#39;A1&#39;, &#39;allele2&#39;:&#39;A2&#39;}, inplace=True, errors=&#39;raise&#39;)
        ###df_ld_snps[&#39;CHR&#39;] = df_ld_snps[&#39;CHR&#39;].astype(np.int64)
        df_ld_snps[&#39;BP&#39;] = df_ld_snps[&#39;BP&#39;].astype(np.int64)
        return df_ld_snps

    def load_ld_bcor(ld_prefix):
        bcor_file = ld_prefix+&#39;.bcor&#39;
        import os
        import time
        from ldstore.bcor import bcor
        if not os.path.exists(bcor_file):
            raise IOError(&#39;%s not found&#39;%(bcor_file))
        t0 = time.time()
        bcor_obj = bcor(bcor_file)
        df_ld_snps = get_bcor_meta(bcor_obj)
        ld_arr = bcor_obj.readCorr([])
        assert np.all(~np.isnan(ld_arr))
        return ld_arr, df_ld_snps

    def abf(beta, se, W=0.04):
        from scipy import special 
        z = beta / se
        V = se ** 2
        r = W / (W + V)
        lbf = 0.5 * (np.log(1 - r) + (r * z ** 2))
        denom = special.logsumexp(lbf)
        prob = np.exp(lbf - denom)
        return lbf, prob
    
    def get_cs(variant, prob, coverage=0.95):
        ordering = np.argsort(prob)[::-1]
        idx = np.where(np.cumsum(prob[ordering]) &gt; coverage)[0][0]
        cs = variant[ordering][: (idx + 1)]
        return cs
    def slalom(df,LD,abf_prior_variance = 0.4 ,nlog10p_dentist_s_threshold = 4, r2_threshold = 0.6  ):
        from scipy import stats
        lbf, prob = abf(df.beta, df.se, W=abf_prior_variance)
        cs = get_cs(df.variant, prob, coverage=0.95)
        cs_99 = get_cs(df.variant, prob, coverage=0.99)
        df[&quot;lbf&quot;] = lbf
        df[&quot;prob&quot;] = prob
        df[&quot;cs&quot;] = df.variant.isin(cs)
        df[&quot;cs_99&quot;] = df.variant.isin(cs_99)
        
        if &quot;${lead_idx_choice}&quot; == &quot;pvalue&quot;:
            lead_idx_snp = df.pvalue.idxmin()
        else:
            lead_idx_snp = df.prob.idxmax()
            
        lead_variant = df.variant[lead_idx_snp]
        df[&quot;lead_variant&quot;] = False
        df[&quot;lead_variant&quot;].iloc[lead_idx_snp] = True
        # annotate LD     
        ## This is to identify the R for each snp vs the lead snp
        df[&quot;r&quot;] = [LD[np.where(np.in1d(df.variant,lead_variant))][:,np.where(np.in1d(df.variant,x))][0][0][0] for x in df.variant]
        lead_z = (df.beta / df.se).iloc[lead_idx_snp]
        df[&quot;t_dentist_s&quot;] = ((df.beta / df.se) - df.r * lead_z) ** 2 / (1 - df.r ** 2)
        df[&quot;t_dentist_s&quot;] = np.where(df[&quot;t_dentist_s&quot;] &lt; 0, np.inf, df[&quot;t_dentist_s&quot;])
        df[&quot;t_dentist_s&quot;].iloc[lead_idx_snp] = np.nan
        df[&quot;nlog10p_dentist_s&quot;] = stats.chi2.logsf(df[&quot;t_dentist_s&quot;], df=1) / -np.log(10)
        df[&quot;r2&quot;] = df.r ** 2
        df[&quot;outliers&quot;] = (df.r2 &gt; r2_threshold) &amp; (df.nlog10p_dentist_s &gt; nlog10p_dentist_s_threshold)
        df_output = df
        n_r2 = np.sum(df.r2 &gt; r2_threshold)
        n_dentist_s_outlier = np.sum(
            (df.r2 &gt; r2_threshold) &amp; (df.nlog10p_dentist_s &gt; nlog10p_dentist_s_threshold)
        )
        max_pip_idx = df.prob.idxmax()
        df_summary = pd.DataFrame(
            {
                &quot;lead_pip_variant&quot;: [df.variant.iloc[max_pip_idx]],
                &quot;n_total&quot;: [len(df.index)],
                &quot;n_r2&quot;: [n_r2],
                &quot;n_dentist_s_outlier&quot;: [n_dentist_s_outlier],
                &quot;fraction&quot;: [n_dentist_s_outlier / n_r2 if n_r2 &gt; 0 else 0],
                &quot;max_pip&quot;: [np.max(df.prob)]
            }
            )
        return df, df_summary
    
    ## Load LD
    if &quot;${_input[0]}&quot;.endswith(&quot;npz&quot;):
        LD,snp_id = load_npz_ld(${_input[0]:r}) 
    if &quot;${_input[0]}&quot;.endswith(&quot;bcor&quot;):
        LD,snp_id = load_ld_bcor(${_input[0]:nr}) 
        
    sumstat = pd.read_csv(${_input[1]:r}, sep = &quot;\t&quot;)
    ## Remove nan in LD
    where = np.where(np.isnan(LD))
    na_snp = np.array(snp_id)[[x for x in where[0]] + [x for x in where[1]]]
    ### Remove NA from LD
    LD = LD[np.ix_(~np.in1d(snp_id, na_snp),~np.in1d(snp_id, na_snp))]
    ### Remove NA from snp_list
    snp_id = np.array(snp_id)[~np.in1d(snp_id, na_snp)]
    
    ## Get only intersect snp
    intersct = np.intersect1d(sumstat.variant.to_numpy(),snp_id)
    sumstat = sumstat.query(&quot;variant in @intersct&quot;).reset_index()
    indice = np.where(np.in1d(snp_id, intersct))
    ## slalom
    ss_qc,ss_qc_sum = slalom(sumstat,LD[np.ix_(indice[0].tolist(), indice[0].tolist())],${abf_prior_variance},${nlog10p_dentist_s_threshold},${r2_threshold})
    print(ss_qc_sum)
    ss_qc.to_csv(${_output[1]:r},&quot;\t&quot;,index = 0 )
    ## Filter out outlier
    ### Get outliers
    outliers = ss_qc[ss_qc.outliers].variant
    ### Remove outliers from LD
    LD = LD[np.ix_(~np.in1d(snp_id, outliers),~np.in1d(snp_id, outliers))]
    ### Remove outliers from snp_list
    snp_id = np.array(snp_id)[~np.in1d(snp_id, outliers)]
    ss_qc = ss_qc.assign(z = ss_qc.beta/ss_qc.se)
    ss_qc = ss_qc[~ss_qc.outliers]

    if ${impute}:
        cur_LD = LD
        cur_miss = ~np.in1d(snp_id, ss_qc.variant)
        print(f&#39;{np.mean(cur_miss)} of snps in the LD panel does not have sumstat. Their sumstat are imputed&#39;)
        ## Sumstat imputation based on fusion code
        cur_wgt = cur_LD[np.ix_(cur_miss, np.logical_not(cur_miss))].dot(np.linalg.inv(cur_LD[np.logical_not(cur_miss), np.logical_not(cur_miss)] + 0.1 * np.eye(np.sum(np.logical_not(cur_miss)))))
        cur_impz  = cur_wgt @ ss_qc.z
        cur_r2pred = np.diag(cur_wgt.dot(cur_LD[np.ix_(np.logical_not(cur_miss),np.logical_not(cur_miss))]).dot(cur_wgt.T))
        imputed_z = cur_impz/np.sqrt(cur_r2pred)
        working_ss = pd.concat([ss_qc[[&quot;variant&quot;,&quot;z&quot;]], pd.DataFrame({&quot;z&quot; : imputed_z, &quot;variant&quot;: np.array(snp_id)[cur_miss]})]).set_index(&quot;variant&quot;).reindex(snp_id).reset_index()
        import scipy.stats as st
        working_ss[&quot;beta&quot;] = working_ss.z
        working_ss[&quot;se&quot;] = 1
        working_ss[&quot;pvalue&quot;] = (1 - st.norm.cdf(working_ss.z))*2
        
        
        working_ss_qc, working_ss_qc_sum = slalom(working_ss,cur_LD,0.4,4,0.6)
        print(working_ss_qc_sum)
        working_ss_qc.to_csv(&#39;${_output[1]:nn}.imputed_ss_qced.tsv&#39;,&quot;\t&quot;,index = 0 )
        imputed_outliers = working_ss_qc[working_ss_qc.outliers].variant
        ### Remove outliers from LD
        cur_LD = LD[np.ix_(~np.in1d(snp_id, imputed_outliers),~np.in1d(snp_id, imputed_outliers))]
        ### Remove outliers from snp_list
        snp_id = np.array(snp_id)[~np.in1d(snp_id, imputed_outliers)]
        working_ss_qc = working_ss_qc[~working_ss_qc.outliers]
        ss_qc = working_ss_qc
        LD = cur_LD 
    ### For non-imputed LD, retained only the overlap snps. For imputed LD, this should do nothing as snp_id, ss_qc.variant .
    indice = np.where(np.in1d(snp_id, ss_qc.variant))
    LD = LD[np.ix_(indice[0].tolist(), indice[0].tolist())]
    ## SuSiERSS
    string=&quot;&quot;&quot;
    susie_rss_analysis=function(ss_df, R, n, var_y, z_ld_weight = 0, estimate_residual_variance = FALSE, 
    prior_variance = 50, check_prior = TRUE, output_path,L = ${L} , max_iter  = ${max_iter} ){
    condz = susieR::kriging_rss(as.matrix(ss_df$z),R ,n = ${n})
    ggplot2::ggsave(plot = condz$plot,filename = &quot;${_output[0]}.kriging_rss.pdf&quot;, device = &quot;pdf&quot;)
    res = susieR:::susie_rss(as.matrix(ss_df$z),as.matrix(R), n = ${n},var_y, z_ld_weight = 0, estimate_residual_variance = FALSE, 
    prior_variance = 50, check_prior = TRUE, max_iter = ${max_iter},verbose = TRUE,L = ${L}   )
    res$variants = ss_df$variant
    res$z = ss_df$z
    res$corr = susieR:::get_cs_correlation(res, X = NULL, Xcorr = R, max = FALSE)
    res$conditional_dist = condz$conditional_dist 
    rownames(res$corr) &lt;- names(res$cs)
    colnames(res$corr) &lt;- names(res$cs)
    if (length(res$sets$cs) &gt; 1) {
        index_combos = expand.grid(1:length(res$sets$cs),1:length(res$sets$cs))
        in_common = apply(index_combos, 1, function(x) intersect(res$sets$cs[[x[1]]], res$sets$cs[[x[2]]]))
        counts = unlist(lapply(in_common, length))
        ovlp_mat = matrix(counts, ncol = length(res$sets$cs), byrow = T)
        ovlp_mat[lower.tri(ovlp_mat)] = NA
        rownames(ovlp_mat) = names(res$sets$cs)
        colnames(ovlp_mat) = names(res$sets$cs)
        print(ovlp_mat)
        res$sets[[&quot;ovlp_mat&quot;]] = ovlp_mat
    }
    saveRDS(res,output_path)
    return(res)
    }
    &quot;&quot;&quot;
    susie_analysis = SignatureTranslatedAnonymousPackage(string, &quot;susie_analysis&quot;)
    analysis_result = susie_analysis.susie_rss_analysis(ss_df = ss_qc,R = LD,output_path=${_output[0]:r} ${f&#39;, n ={n}&#39; if n &gt; 2 else &quot;&quot;})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>[SuSiE_RSS_2]
output: pip_plot = f&quot;{cwd}/{_input:bn}.png&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = &#39;12h&#39;, mem = &#39;20G&#39;, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;
    res = readRDS(${_input:r})
    png(${_output[0]:r}, width = 14, height=6, unit=&#39;in&#39;, res=300)
    par(mfrow=c(1,2))
    susieR::susie_plot(res, y= &quot;PIP&quot;, pos=list(attr=&#39;pos&#39;,start=res$pos[1],end=res$pos[length(res$pos)]), add_legend=T, xlab=&quot;position&quot;)
    susieR::susie_plot(res, y= &quot;z&quot;, pos=list(attr=&#39;pos&#39;,start=res$pos[1],end=res$pos[length(res$pos)]), add_legend=T, xlab=&quot;position&quot;, ylab=&quot;-log10(p)&quot;)
    dev.off()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>[SuSiE_RSS_3]
sep = &quot;&quot; #&#39;\n\n---\n&#39;
input: group_by = &#39;all&#39;
output: analysis_summary = f&#39;{cwd}/{sumstats_path:bnn}.analysis_summary.md&#39;, variants_csv = f&#39;{cwd}/{sumstats_path:bnn}.variants.csv&#39;
python: container=container, expand = &quot;${ }&quot;

    theme = &#39;&#39;&#39;---
    theme: base-theme
    style: |
     p {
       font-size: 24px;
       height: 900px;
       margin-top:1cm;
      }
      img {
        height: 70%;
        display: block;
        margin-left: auto;
        margin-right: auto;
      }
      body {
       margin-top: auto;
       margin-bottom: auto;
       font-family: verdana;
      }
    ---    
    &#39;&#39;&#39;
    import numpy as np
    import pandas as pd
    
    # will load the rds file outputted in a previous step
    def load_rds(filename, types=None):
        import os
        import pandas as pd, numpy as np
        import rpy2.robjects as RO
        import rpy2.robjects.vectors as RV
        import rpy2.rinterface as RI
        from rpy2.robjects import numpy2ri
        numpy2ri.activate()
        from rpy2.robjects import pandas2ri
        pandas2ri.activate()
        def load(data, types, rpy2_version=3):
            if types is not None and not isinstance(data, types):
                return np.array([])
            # FIXME: I&#39;m not sure if I should keep two versions here
            # rpy2_version 2.9.X is more tedious but it handles BoolVector better
            # rpy2 version 3.0.1 converts bool to integer directly without dealing with
            # NA properly. It gives something like (0,1,-234235).
            # Possibly the best thing to do is to open an issue for it to the developers.
            if rpy2_version == 2:
                # below works for rpy2 version 2.9.X
                if isinstance(data, RI.RNULLType):
                    res = None
                elif isinstance(data, RV.BoolVector):
                    data = RO.r[&#39;as.integer&#39;](data)
                    res = np.array(data, dtype=int)
                    # Handle c(NA, NA) situation
                    if np.sum(np.logical_and(res != 0, res != 1)):
                        res = res.astype(float)
                        res[res &lt; 0] = np.nan
                        res[res &gt; 1] = np.nan
                elif isinstance(data, RV.FactorVector):
                    data = RO.r[&#39;as.character&#39;](data)
                    res = np.array(data, dtype=str)
                elif isinstance(data, RV.IntVector):
                    res = np.array(data, dtype=int)
                elif isinstance(data, RV.FloatVector):
                    res = np.array(data, dtype=float)
                elif isinstance(data, RV.StrVector):
                    res = np.array(data, dtype=str)
                elif isinstance(data, RV.DataFrame):
                    res = pd.DataFrame(data)
                elif isinstance(data, RV.Matrix):
                    res = np.matrix(data)
                elif isinstance(data, RV.Array):
                    res = np.array(data)
                else:
                    # I do not know what to do for this
                    # But I do not want to throw an error either
                    res = str(data)
            else:
                if isinstance(data, RI.NULLType):
                    res = None
                else:
                    res = data
            if isinstance(res, np.ndarray) and res.shape == (1, ):
                res = res[0]
            return res
        def load_dict(res, data, types):
            &#39;&#39;&#39;load data to res&#39;&#39;&#39;
            names = data.names if not isinstance(data.names, RI.NULLType) else [
                i + 1 for i in range(len(data))
            ]
            for name, value in zip(names, list(data)):
                if isinstance(value, RV.ListVector):
                    res[name] = {}
                    res[name] = load_dict(res[name], value, types)
                else:
                    res[name] = load(value, types)
            return res
        #
        if not os.path.isfile(filename):
            raise IOError(&#39;Cannot find file ``{}``!&#39;.format(filename))
        rds = RO.r[&#39;readRDS&#39;](filename)
        if isinstance(rds, RV.ListVector):
            res = load_dict({}, rds, types)
        else:
            res = load(rds, types)
        return res
    
    def f7(seq):
        seen = set()
        seen_add = seen.add
        return [x for x in seq if not (x in seen or seen_add(x))]



    text = &quot;&quot;
    sep = &#39;\n\n---\n&#39;
    
    inp = &quot;${_input:r}&quot;.split(&quot; &quot;)
    for i, each in enumerate(inp):
        inp[i] = &quot;.&quot;.join(each.split(&quot;.&quot;)[:-1])

    r = f7(&quot;${_input:bn}&quot;.split(&quot; &quot;))
    
    num_csets = []
    region_info = []
    
    # this will be a 2d array that stores information about each variant of interest in the phenotype
    # this includes all the variants in a cs and all the variants past the cutoff
    variant_info = []

    for reg_i, each in enumerate(f7(inp)):
    
        rid = r[reg_i].split(&#39;.&#39;)[0]
        
        text_temp = &quot;&quot;
        text_temp += &quot;#\n\n SuSiE RSS {region} \n&quot;.format(region=r[reg_i])
        text_temp += &quot;![]({region}.png){sep} \n \n&quot;.format(region=r[reg_i], sep=sep)

        rd = load_rds(each[1:]+&quot;.rds&quot;)
        
        # find the number of cs in the current region
        if rd[&quot;sets&quot;][&quot;cs&quot;] == None:
            num_csets.append(0)
        else:
            num_csets.append(len(rd[&quot;sets&quot;][&quot;cs&quot;]))
        print(num_csets)
        
        # this will store the indicies of all variants that cross the threshold
        ind_p = []

        pval = ${pip_cutoff}

        for i, each in enumerate(rd[&quot;pip&quot;]):
            if each &gt;= pval:
                ind_p.append(i)
        sumvars = 0
        
        # if we have at least one cs in the current region
        if num_csets[reg_i] &gt; 0:
            tbl_header = &quot;| chr number | pos at highest pip | ref | alt | region id | cs | highest pip |  \n&quot;
            tbl_header += &quot;| --- | --- | --- | --- | --- | --- | --- |  \n&quot;

            table = &quot;&quot;
            
            sumpips = 0
            
            for cset in rd[&quot;sets&quot;][&quot;cs&quot;].keys():
                print(cset)
                
                # if we have many variants in the cs
                if isinstance(rd[&quot;sets&quot;][&quot;cs&quot;][cset], np.ndarray):
                    highestpip = 0
                    poswhighestpip = -1
                    for i in rd[&quot;sets&quot;][&quot;cs&quot;][cset]:
                        i = i.item() - 1
                        
                        # we make sure that ind_p only stores the variants that aren&#39;t in any cs
                        if i in ind_p: ind_p.remove(i) 
                        
                        # append variant info
                        variant_info.append( [rd[&quot;chr&quot;][i], rd[&quot;pos&quot;][i], rd[&quot;ref&quot;][i], rd[&quot;alt&quot;][i], rid, cset, rd[&quot;pip&quot;][i]] )
                        
                        if rd[&quot;pip&quot;][i] &gt; highestpip:
                            highestpip = rd[&quot;pip&quot;][i]
                            poswhighestpip = i
                            
                        sumpips += rd[&quot;pip&quot;][i]
                        sumvars += 1
                        
                    if poswhighestpip &gt; -1:
                        i = poswhighestpip
                        table += &quot;| {chr} | {pos} | {ref} | {alt} | {rid} | {cs} | {pip:.2f} |  \n&quot;.format(chr=rd[&quot;chr&quot;][i], pos=rd[&quot;pos&quot;][i], ref=rd[&quot;ref&quot;][i], alt=rd[&quot;alt&quot;][i], rid=rid, cs=cset, pip=rd[&quot;pip&quot;][i])
                
                else: # if we have only one variant in the cs
                    i =  rd[&quot;sets&quot;][&quot;cs&quot;][cset]
                    i = i.item() - 1
                    
                    # we make sure that ind_p only stores the variants that aren&#39;t in any cs
                    if i in ind_p: ind_p.remove(i)
                    
                    # append variant info
                    variant_info.append( [rd[&quot;chr&quot;][i], rd[&quot;pos&quot;][i], rd[&quot;ref&quot;][i], rd[&quot;alt&quot;][i], rid, cset, rd[&quot;pip&quot;][i]] )
                    
                    table += &quot;| {chr} | {pos} | {ref} | {alt} | {rid} | {cs} | {pip:.2f} |  \n&quot;.format(chr=rd[&quot;chr&quot;][i], pos=rd[&quot;pos&quot;][i], ref=rd[&quot;ref&quot;][i], alt=rd[&quot;alt&quot;][i], rid=rid, cs=cset, pip=rd[&quot;pip&quot;][i])
                    
                    sumpips += rd[&quot;pip&quot;][i]
                    sumvars += 1
            

            text_temp += &quot;- Total number of variants: {}\n&quot;.format(len(rd[&quot;pip&quot;]))
            text_temp += &quot;- Expected number of causal variants: {:.2f}\n&quot;.format(sumpips)
            text_temp += &quot;- Number of variants with PIP &gt; {} and not in any CS: {}\n\n&quot;.format(pval, len(ind_p))
            text_temp += tbl_header + table + sep
            
            if num_csets[reg_i] &gt; 1:
                text_temp += &quot;#### CORR: Correlation between CS | OLAP: Overlap between CS\n&quot;
                
                cs = list(rd[&quot;sets&quot;][&quot;cs&quot;].keys())

                corrheader = &quot;|  |&quot;
                corrbreak = &quot;| --- |&quot;

                for i in cs:
                    corrheader += &quot; CORR {} |&quot;.format(i)
                    corrbreak += &quot; --- |&quot;
                    
                corrheader += &quot;  |&quot;
                corrbreak += &quot; --- |&quot;
                    
                for i in cs:
                    corrheader += &quot; OLAP {} |&quot;.format(i)
                    corrbreak += &quot; --- |&quot;

                corrheader += &quot;\n&quot;
                corrbreak += &quot;\n&quot;

                body = &quot;&quot;

                for en, i in enumerate(cs):
                    body += &quot;| {} |&quot;.format(i)
                    for j in rd[&quot;cscorr&quot;][en]:
                        body += &quot; {:.2f} |&quot;.format(j)
                    body += &quot;  |&quot;
                    for j in rd[&quot;sets&quot;][&quot;cs&quot;]:
                        body += &quot; {} |&quot;.format(len(np.intersect1d(rd[&quot;sets&quot;][&quot;cs&quot;][i], rd[&quot;sets&quot;][&quot;cs&quot;][j])))
                    body += &quot;\n&quot;
                
                text_temp += corrheader + corrbreak + body + sep
            
        region_info.append(text_temp)
            
    f = open(${_output[&quot;analysis_summary&quot;]:r}, &quot;w&quot;)
    
    cset_order = np.argsort(num_csets)
    cset_order = cset_order.tolist()
    cset_order.reverse()
    for c in cset_order:
        text += region_info[c]
    
    f.write(theme + text)
    
    f.close()
    
    for i in ind_p:
        # append variant info
        variant_info.append( [rd[&quot;chr&quot;][i], rd[&quot;pos&quot;][i], rd[&quot;ref&quot;][i], rd[&quot;alt&quot;][i], rid, &quot;None&quot;, rd[&quot;pip&quot;][i]] )
        
    df = pd.DataFrame(variant_info, columns=[&quot;chr&quot;, &quot;pos&quot;, &quot;ref&quot;, &quot;alt&quot;, &quot;rid&quot;, &quot;cs&quot;, &quot;pip&quot;])
    df.to_csv(${_output[&quot;variants_csv&quot;]:r}, sep = &quot;\t&quot;, header = True, index = True)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Generate analysis report: HTML file, and optionally PPTX file
[SuSiE_RSS_4]
output: f&quot;{_input[&#39;analysis_summary&#39;]:n}.html&quot;
sh: container=container_marp, expand = &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;
    node /opt/marp/.cli/marp-cli.js ${_input[&#39;analysis_summary&#39;]} -o ${_output:a} \
        --title &#39;${region_file:bnn} fine mapping analysis&#39; \
        --allow-local-files
    node /opt/marp/.cli/marp-cli.js ${_input[&#39;analysis_summary&#39;]} -o ${_output:an}.pptx \
        --title &#39;${region_file:bnn} fine mapping analysis&#39; \
        --allow-local-files
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "rpy2_susie"
        },
        kernelOptions: {
            name: "rpy2_susie",
            path: "./code/fine_mapping/SuSiE"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'rpy2_susie'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../univariate_fine_mapping.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Univariate fine-mapping workflows</p>
      </div>
    </a>
    <a class="right-next"
       href="SuSiE.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Fine-mapping with SuSiE model</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mwe">MWE</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>