

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Fine-mapping with SuSiE model &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/fine_mapping/SuSiE/SuSiE';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Multivariate fine-mapping workflows" href="../multivariate_fine_mapping.html" />
    <link rel="prev" title="Fine-mapping with SuSiE RSS model" href="SuSiE_RSS.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/fine_mapping/SuSiE/SuSiE.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/fine_mapping/SuSiE/SuSiE.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/fine_mapping/SuSiE/SuSiE.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Fine-mapping with SuSiE model</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-list">Example genotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mwe">MWE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#fsusie">FSuSiE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-fsusie">Univariate fSuSiE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-f-susie">Multi F SuSiE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-susie">Univariate SuSiE</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="fine-mapping-with-susie-model">
<h1>Fine-mapping with SuSiE model<a class="headerlink" href="#fine-mapping-with-susie-model" title="Permalink to this headline">#</a></h1>
<p>This notebook conduct fine_mapping with complete data, unlike the susie_RSS module, this module perform analysis on 1 of the theme each time <strong>FIXME: why saying this “unlike the susie_RSS module, this module perform analysis on 1 of the theme each time”? we should unify the interfaces if we can (let me try to see why we cannot do that)</strong></p>
<p>sos run pipeline/SuSiE.ipynb uni_susie <br />
–genoFile &lt;(head GRCh38_plink_files_list.txt)  <br />
–region-list ALL.log2cpm.region_list <br />
–phenoFile ALL_pseudobulk_phenotype_list.txt <br />
–covFile ALL_pseudobulk_phenotype_list.txt -J 800 -c csg.yml -q csg  &amp;</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">uni_susie</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">cat</span> <span class="n">GRCh38_plink_files_list</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">grep</span> <span class="n">ENSG00000197943</span>  <span class="p">)</span>\
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">region_list</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">ALL_pseudobulk_phenotype_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">ALL_pseudobulk_covariate_list</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">ALL_PLCG2</span> <span class="o">--</span><span class="kp">name</span> <span class="n">ALL_pseudo_bulk</span>  <span class="o">&amp;</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">per_gene_data_merger</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">data_merger</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">cat</span> <span class="n">GRCh38_plink_files_list</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">grep</span> <span class="n">ENSG00000197943</span>  <span class="p">)</span>\
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">region_list</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">ALL_pseudobulk_phenotype_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">ALL_pseudobulk_covariate_list</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">ALL_PLCG2</span> <span class="o">--</span><span class="kp">name</span> <span class="n">ALL_pseudo_bulk</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">data_merger_test</span>  <span class="o">&amp;</span> 
</pre></div>
</div>
</div>
</div>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p>1 region list documenting the regions to be analysised</p></li>
<li><p>a list to the path where phenotype per gene data stored (the .tbi file should be with the pheno .gz file)</p></li>
<li><p>a list to the path where genotype per gene data stored</p></li>
</ol>
<p>By default the input 2 and 3 is the output from the data_preprocessing module.</p>
<section id="example-genotype-list">
<h3>Example genotype list<a class="headerlink" href="#example-genotype-list" title="Permalink to this headline">#</a></h3>
<p>region        dir
ENSG00000000457 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000457.bed
ENSG00000000460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000460.bed
ENSG00000000938 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000938.bed
ENSG00000000971 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000971.bed
ENSG00000001036 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001036.bed
ENSG00000001084 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001084.bed
ENSG00000001167 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001167.bed
ENSG00000001460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001460.bed</p>
</section>
<section id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Permalink to this headline">#</a></h3>
<p>region dir
ENSG00000000457 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000457.mol_phe.bed.gz
ENSG00000000460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000460.mol_phe.bed.gz
ENSG00000000938 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000938.mol_phe.bed.gz
ENSG00000000971 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000971.mol_phe.bed.gz
ENSG00000001036 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001036.mol_phe.bed.gz
ENSG00000001084 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001084.mol_phe.bed.gz
ENSG00000001167 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001167.mol_phe.bed.gz
ENSG00000001460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001460.mol_phe.bed.gz</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">#</a></h2>
<p>For each analysis unit we output:</p>
<ol class="arabic simple">
<li><p>Analysis results in RDS format: A mvsusie Model</p></li>
<li><p>A vcf file with selected snps
ES:PIP:CS</p></li>
</ol>
</section>
<section id="mwe">
<h2>MWE<a class="headerlink" href="#mwe" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">uni_susie_sqtl_1</span> \
<span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">Ast</span><span class="o">/</span><span class="n">genotype_per_gene</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">region_plink_files</span><span class="o">/</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
<span class="o">--</span><span class="n">cwd</span> <span class="n">MWE</span><span class="o">/</span><span class="n">susie_per_gene</span><span class="o">/</span> \
<span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">region</span><span class="o">.</span><span class="n">list</span> \
<span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">phenotype</span><span class="o">.</span><span class="n">list</span> \
<span class="o">--</span><span class="n">covFile</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">covar</span><span class="o">.</span><span class="n">list</span> <span class="o">-</span><span class="n">J</span> <span class="mi">8</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span>  <span class="o">&amp;</span>

<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mv_susie</span> \
<span class="o">--</span><span class="n">genoFile</span> <span class="o">./</span><span class="n">mwe</span><span class="o">.</span><span class="n">region_plink_files</span><span class="o">/</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">cwd</span> <span class="n">MWE</span><span class="o">/</span><span class="n">rds_per_gene</span><span class="o">/</span> \
<span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">MWE</span><span class="o">.</span><span class="n">region</span><span class="o">.</span><span class="n">list</span> \
<span class="o">--</span><span class="n">phenoFile</span> <span class="n">MWE</span><span class="o">.</span><span class="n">phenotype</span><span class="o">.</span><span class="n">list</span> \
<span class="o">--</span><span class="n">covFile</span> <span class="n">MWE</span><span class="o">.</span><span class="n">covar</span><span class="o">.</span><span class="n">list</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">uni_susie</span> \
<span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span><span class="n">dlpfc_tad_gene_genotype_list</span>  \
<span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">susie_per_gene_tad_2</span><span class="o">/</span> \
<span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="o">./</span><span class="n">dlpfc_region_list</span> \
<span class="o">--</span><span class="n">phenoFile</span> <span class="n">phenotype_list</span> \
<span class="o">--</span><span class="n">covFile</span> <span class="n">covar_list</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span>  <span class="o">--</span><span class="n">job_size</span> <span class="mi">1</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">80</span><span class="n">G</span> <span class="o">--</span><span class="n">numThreads</span> <span class="mi">8</span> <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">stephenslab</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">tad_list</span> <span class="n">dlpfc_tad_region_list</span>  <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="fsusie">
<h2>FSuSiE<a class="headerlink" href="#fsusie" title="Permalink to this headline">#</a></h2>
<p>The interface of fsusie is identical as the uni_susie. However, instead of extracting 1 row out of each bed.gz file. It will extract all the rows which tabix believe they are contained by the region specified in the region list</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">uni_fsusie</span> \
<span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">dlpfc_tad_region_list_plink_files</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
<span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">f_susie_tad_eQTL</span><span class="o">/</span> <span class="o">--</span><span class="kp">name</span> <span class="n">eQTL</span> \
<span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="o">./</span><span class="n">dlpfc_tad_list</span> \
<span class="o">--</span><span class="n">phenoFile</span> <span class="n">phenotype_list</span> \
<span class="o">--</span><span class="n">covFile</span> <span class="n">covar_list</span>  <span class="o">-</span><span class="n">s</span> <span class="n">build</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">uni_fsusie</span> \
<span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">dlpfc_tad_region_list_plink_files</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
<span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">f_susie_tad_haQTL_pos</span><span class="o">/</span> <span class="o">--</span><span class="kp">name</span> <span class="n">haQTL</span> \
<span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">dlpfc_tad_list</span> \
<span class="o">--</span><span class="n">phenoFile</span> <span class="n">phenotype_list</span> \
<span class="o">--</span><span class="n">covFile</span> <span class="n">covar_list</span> <span class="o">--</span><span class="n">prior</span> <span class="n">mixture_normal_per_scale</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">--</span><span class="n">job_size</span> <span class="mi">2</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg2</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span>  <span class="o">--</span><span class="n">job_size</span> <span class="mi">1</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">35</span><span class="n">G</span>  <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">uni_fsusie</span> \
<span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">dlpfc_tad_region_list_plink_files</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span>  \
<span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">f_susie_tad_haQTL_pos_check_pure_2</span><span class="o">/</span> <span class="o">--</span><span class="kp">name</span> <span class="n">haQTL</span> \
<span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">dlpfc_tad_list</span> \
<span class="o">--</span><span class="n">phenoFile</span> <span class="n">phenotype_list</span> \
<span class="o">--</span><span class="n">covFile</span> <span class="n">covar_list</span> <span class="o">--</span><span class="n">prior</span> <span class="n">mixture_normal_per_scale</span> <span class="o">-</span><span class="n">J</span> <span class="mi">20</span> <span class="o">--</span><span class="n">job_size</span> <span class="mi">2</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg2</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span>  <span class="o">--</span><span class="n">job_size</span> <span class="mi">1</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">20</span><span class="n">G</span> <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">stephenslab</span><span class="o">.</span><span class="n">sif</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1"># Input</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># A list of file path for phenotype and covariates, must be in the same order.</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">covFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;demo&quot;</span>
<span class="n">region_tbl</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">id_name</span> <span class="o">=</span> <span class="n">region_tbl</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">genoFile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genoFile</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="n">id_name</span><span class="p">,</span><span class="s2">&quot;path&quot;</span><span class="p">],</span><span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">region_tbl</span><span class="p">,</span><span class="n">on</span> <span class="o">=</span> <span class="n">id_name</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;records&quot;</span><span class="p">)</span>

<span class="c1">## Path to work directory where output locates</span>
<span class="c1">## Containers that contains the necessary packages</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">trait</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">tissue</span>
<span class="c1"># use this function to edit memory string for PLINK input</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="univariate-fsusie">
<h2>Univariate fSuSiE<a class="headerlink" href="#univariate-fsusie" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[uni_fsusie_1]
parameter: max_L = 20
# remove a variant if it has more than imiss missing individual data
parameter: imiss = 0.1
parameter: maf = 0.05
parameter: prior  = [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: pheno_ID = &quot;ID&quot;
parameter: max_SNP_EM = 1000
input: phenoFile,covFile, for_each = &quot;genoFile&quot;
output: [f&#39;{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.uni_Fsusie.{prior[0]}.rds&#39; for x in trait ] 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    ## Define function
    compute_maf &lt;- function(geno){
      f &lt;- mean(geno,na.rm = TRUE)/2
      return(min(f, 1-f))
    }

    compute_missing &lt;- function(geno){
      miss &lt;- sum(is.na(geno))/length(geno)
      return(miss)
    }
    
    mean_impute &lt;- function(geno){
      f &lt;- apply(geno, 2, function(x) mean(x,na.rm = TRUE))
      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] &lt;- f[i]
      return(geno)
    }

    is_zero_variance &lt;- function(x) {
      if (length(unique(x))==1) return(T)
      else return(F)
    }
  
    filter_X &lt;- function(X, missing_rate_thresh, maf_thresh) {
        rm_col &lt;- which(apply(X, 2, compute_missing) &gt; missing_rate_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, compute_maf) &lt; maf_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, is_zero_variance))
        if (length(rm_col)) X &lt;- X[, -rm_col]
        return(mean_impute(X))
    }
  
    compute_cov_flash &lt;- function(Y, error_cache = NULL){
        covar &lt;- diag(ncol(Y))
        tryCatch({
        fl &lt;- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)
        if(fl$n.factors==0){
          covar &lt;- diag(fl$residuals.sd^2)
        } else {
          fsd &lt;- sapply(fl$fitted.g[[1]], &#39;[[&#39;, &quot;sd&quot;)
          covar &lt;- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)
        }
        if (nrow(covar) == 0) {
          covar &lt;- diag(ncol(Y))
          stop(&quot;Computed covariance matrix has zero rows&quot;)
        }
        }, error = function(e) {
          if (!is.null(error_cache)) {
            saveRDS(list(data=Y, message=warning(e)), error_cache)
            warning(&quot;FLASH failed. Using Identity matrix instead.&quot;)
            warning(e)
          } else {
            stop(e)
          }
        })
        s &lt;- apply(Y, 2, sd, na.rm=T)
        if (length(s)&gt;1) s = diag(s)
        else s = matrix(s,1,1)
        covar &lt;- s%*%cov2cor(covar)%*%s
        return(covar)
    }

    read_multi_pheno = function(path){
      arg = paste0(c(&quot;tabix -h &quot;,path,&quot; ${_genoFile[&quot;#chr&quot;]}:${_genoFile[&quot;start&quot;]}-${_genoFile[&quot;end&quot;]}&quot;),collapse = &quot;&quot;)
      result = system(arg,intern = T)
        if (is.na(result[2]) == F){
          output = read.table(text = result[2:length(result)], sep = &quot;\t&quot;)
          colnames(output) = result[1]%&gt;%stringr::str_split(&quot;\t&quot;)%&gt;%unlist()
        }
        else {
            output = NA
            }
        return(output) # add an if so that for each traits, those region in geno_file contains nothing in the pheno_file, can be corrctly left out
    }       

    
    remove_covX = function(X,covar){
    for ( i in 1:ncol(X) ) {
        X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals
        }
     X = scale(X)
    }
    
    ## Load Library
    library(&quot;susiF.alpha&quot;)
    library(&quot;plink2R&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;purrr&quot;)
    ###
    # Core code
    ###
    # Input
    ### Genotype
    geno = read_plink(&quot;${path(_genoFile[&quot;path&quot;]):n}&quot;)
    X = filter_X(geno$bed,${imiss}, ${maf} )

    ### Phenotype
    phenotype_list = read_delim(&quot;${_input[0]}&quot;,&quot;\t&quot;)
    traits = phenotype_list$tissue
    covar_list = read_delim(&quot;${_input[1]}&quot;,&quot;\t&quot;)
    rownames(X) = read.table(text = rownames(X), sep= &quot;:&quot; )$V2
    covar_list = covar_list%&gt;%mutate(covar = map(path, ~read_delim(.x,&quot;\t&quot;)%&gt;%select(-`#id`)%&gt;%na.omit%&gt;%t()))
    phenotype_list = inner_join(phenotype_list,covar_list, by = &quot;tissue&quot;)
    phenotype_list = phenotype_list %&gt;% mutate(flag = map(path.x, read_multi_pheno)) %&gt;% dplyr::filter(lengths(flag) &gt;1 ) %&gt;% mutate(Y = map2(path.x,covar, ~read_multi_pheno(.x)%&gt;%select(start, rownames(.y))%&gt;%t%&gt;%as.matrix))%&gt;%
                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out
                                            pos = map(Y,~.x[&quot;start&quot;,]),
                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),
                                            covar = map2(covar, X_data , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            Y = map2(covar, Y ,~ .y[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            #### Get residue for each of tissue
                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale%&gt;%t%&gt;%as_tibble))
    
    Y_resid = phenotype_list%&gt;%pull(Y_resid)
    names(Y_resid) = phenotype_list$tissue
    X_list = phenotype_list%&gt;%mutate( X_resid = map2(X_data,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale))%&gt;%pull(X_resid)       # Remove covariate by tissues
    non_missing = list()
    fitted = list()
    output_list = c(${_output:r})
    rm_traits = traits[!traits %in% names(Y_resid)]
    output_list_valid = output_list[traits %in% names(Y_resid)]
    output_list_invalid = output_list[!traits %in% names(Y_resid)]
    
    # Fine-mapping with fSuSiE
    for (r in 1:length(Y_resid)) {
    Y_resid_mtx = (Y_resid[[r]])%&gt;%t%&gt;%as.matrix
    non_missing[[r]] = which(rowSums(is.na(Y_resid_mtx)) == 0)
    st = proc.time()
    X = X_list[[r]]
    print(paste(&quot;Dimension of X matrix:&quot;, nrow(X), ncol(X)))
    print(paste(&quot;Dimension of Y matrix:&quot;, nrow(Y_resid_mtx), ncol(Y_resid_mtx)))
    fitted[[r]] &lt;- susiF(X = X[names(non_missing[[r]]),], Y_resid_mtx[non_missing[[r]],],pos = phenotype_list$pos[[r]],
                       L=${max_L},prior = &quot;${prior[0]}&quot; ,  max_SNP_EM = ${max_SNP_EM} )
    fitted[[r]]$time = proc.time() - st
    fitted[[r]]$name = phenotype_list$tissue[r]
    fitted[[r]]$input_data = list(X =  X[names(non_missing[[r]]),] , Y_resid_mtx[non_missing[[r]],], pos =  phenotype_list$pos[[r]] )
    saveRDS(fitted[[r]], output_list_valid[r])
    }
    arg = paste0(c(&quot;touch &quot;,output_list_invalid),collapse = &quot; &quot;)
    system(arg)
</pre></div>
</div>
</div>
</div>
</section>
<section id="multi-f-susie">
<h2>Multi F SuSiE<a class="headerlink" href="#multi-f-susie" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[multi_f_susie_1]
parameter: max_L = 10
# remove a variant if it has more than imiss missing individual data
parameter: imiss = 0.1
parameter: maf = 0.05
parameter: prior  = [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: pheno_ID = &quot;ID&quot;
input: phenoFile,covFile, for_each = &quot;genoFile&quot;
output: f&#39;{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.multi_Fsusie.rds&#39; for x in trait
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    ## Define function
    compute_maf &lt;- function(geno){
      f &lt;- mean(geno,na.rm = TRUE)/2
      return(min(f, 1-f))
    }

    compute_missing &lt;- function(geno){
      miss &lt;- sum(is.na(geno))/length(geno)
      return(miss)
    }
    
    mean_impute &lt;- function(geno){
      f &lt;- apply(geno, 2, function(x) mean(x,na.rm = TRUE))
      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] &lt;- f[i]
      return(geno)
    }

    is_zero_variance &lt;- function(x) {
      if (length(unique(x))==1) return(T)
      else return(F)
    }
  
    filter_X &lt;- function(X, missing_rate_thresh, maf_thresh) {
        rm_col &lt;- which(apply(X, 2, compute_missing) &gt; missing_rate_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, compute_maf) &lt; maf_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, is_zero_variance))
        if (length(rm_col)) X &lt;- X[, -rm_col]
        return(mean_impute(X))
    }
  
    compute_cov_flash &lt;- function(Y, error_cache = NULL){
        covar &lt;- diag(ncol(Y))
        tryCatch({
        fl &lt;- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)
        if(fl$n.factors==0){
          covar &lt;- diag(fl$residuals.sd^2)
        } else {
          fsd &lt;- sapply(fl$fitted.g[[1]], &#39;[[&#39;, &quot;sd&quot;)
          covar &lt;- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)
        }
        if (nrow(covar) == 0) {
          covar &lt;- diag(ncol(Y))
          stop(&quot;Computed covariance matrix has zero rows&quot;)
        }
        }, error = function(e) {
          if (!is.null(error_cache)) {
            saveRDS(list(data=Y, message=warning(e)), error_cache)
            warning(&quot;FLASH failed. Using Identity matrix instead.&quot;)
            warning(e)
          } else {
            stop(e)
          }
        })
        s &lt;- apply(Y, 2, sd, na.rm=T)
        if (length(s)&gt;1) s = diag(s)
        else s = matrix(s,1,1)
        covar &lt;- s%*%cov2cor(covar)%*%s
        return(covar)
    }

    read_multi_pheno = function(path){
      arg = paste0(c(&quot;tabix -h &quot;,path,&quot; ${_genoFile[&quot;#chr&quot;]}:${_genoFile[&quot;start&quot;]}-${_genoFile[&quot;end&quot;]}&quot;),collapse = &quot;&quot;)
      result = system(arg,intern = T)
        if (is.na(result[2]) == F){
          output = read.table(text = result[2:length(result)], sep = &quot;\t&quot;)
          colnames(output) = result[1]%&gt;%stringr::str_split(&quot;\t&quot;)%&gt;%unlist()
        }
        else {
            output = NA
            }
        return(output) # add an if so that for each traits, those region in geno_file contains nothing in the pheno_file, can be corrctly left out
    }       

    
    remove_covX = function(X,covar){
    for ( i in 1:ncol(X) ) {
        X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals
        }
     X = scale(X)
    }
    
    ## Load Library
    library(&quot;mvf.susie.alpha&quot;)
    library(&quot;susiF.alpha&quot;)
    library(&quot;plink2R&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;purrr&quot;)
    ###
    # Core code
    ###
    # Input
    ### Genotype
    geno = read_plink(&quot;${path(_genoFile[&quot;path&quot;]):n}&quot;)
    X = filter_X(geno$bed,${imiss}, ${maf} )

    ### Phenotype
    phenotype_list = read_delim(&quot;${_input[0]}&quot;,&quot;\t&quot;)
    traits = phenotype_list$tissue
    covar_list = read_delim(&quot;${_input[1]}&quot;,&quot;\t&quot;)
    rownames(X) = read.table(text = rownames(X), sep= &quot;:&quot; )$V2
    covar_list = covar_list%&gt;%mutate(covar = map(path, ~read_delim(.x,&quot;\t&quot;)%&gt;%select(-`#id`)%&gt;%na.omit%&gt;%t()))
    phenotype_list = inner_join(phenotype_list,covar_list, by = &quot;tissue&quot;)
    phenotype_list = phenotype_list %&gt;% mutate(flag = map(path.x, read_multi_pheno)) %&gt;% dplyr::filter(lengths(flag) &gt;1 ) %&gt;% mutate(Y = map2(path.x,covar, ~read_multi_pheno(.x)%&gt;%select(start, rownames(.y))%&gt;%t%&gt;%as.matrix))%&gt;%
                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out
                                            pos = map(Y,~.x[&quot;start&quot;,]),
                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),
                                            covar = map2(covar, X_data , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            Y = map2(covar, Y ,~ .y[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            #### Get residue for each of tissue
                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale%&gt;%t%&gt;%as_tibble))
    
    Y_resid = phenotype_list%&gt;%pull(Y_resid)
    Y_resid_new = map(Y_resid, ~left_join(X[,1]%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;), .x%&gt;%t%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;) , by = &quot;rowname&quot;)%&gt;%select(-2)%&gt;%column_to_rownames(&quot;rowname&quot;)%&gt;%as.matrix )

  
  
  
    m1 &lt;- multfsusie(Y=list( Y_f = list(Y_resid_new[[1]],Y_resid_new[[3]]) , Y_u= Reduce(cbind, Y_resid_new[[2]]),pos = list(pos1 =  phenotype_list$pos[[1]],pos2 =  phenotype_list$pos[[3]]) ,
                 X=X ,
                 L=20 ,
                 data.format=&quot;list_df&quot;)
    m1%&gt;%saveRDS(&quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="univariate-susie">
<h2>Univariate SuSiE<a class="headerlink" href="#univariate-susie" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[uni_susie_1]
parameter: max_L = 10
# remove a variant if it has more than imiss missing individual data
parameter: imiss = 0.1
parameter: maf = 0.05
import pandas as pd
input: phenoFile,covFile, for_each = &quot;genoFile&quot;
output: [f&#39;{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.unisusie.fit.rds&#39; for x in trait  ] 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    ## Define function
    compute_maf &lt;- function(geno){
      f &lt;- mean(geno,na.rm = TRUE)/2
      return(min(f, 1-f))
    }

    compute_missing &lt;- function(geno){
      miss &lt;- sum(is.na(geno))/length(geno)
      return(miss)
    }
    
    mean_impute &lt;- function(geno){
      f &lt;- apply(geno, 2, function(x) mean(x,na.rm = TRUE))
      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] &lt;- f[i]
      return(geno)
    }

    is_zero_variance &lt;- function(x) {
      if (length(unique(x))==1) return(T)
      else return(F)
    }
  
    filter_X &lt;- function(X, missing_rate_thresh, maf_thresh) {
        rm_col &lt;- which(apply(X, 2, compute_missing) &gt; missing_rate_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, compute_maf) &lt; maf_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, is_zero_variance))
        if (length(rm_col)) X &lt;- X[, -rm_col]
        return(mean_impute(X))
    }
  
    compute_cov_flash &lt;- function(Y, error_cache = NULL){
        covar &lt;- diag(ncol(Y))
        tryCatch({
        fl &lt;- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)
        if(fl$n.factors==0){
          covar &lt;- diag(fl$residuals.sd^2)
        } else {
          fsd &lt;- sapply(fl$fitted.g[[1]], &#39;[[&#39;, &quot;sd&quot;)
          covar &lt;- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)
        }
        if (nrow(covar) == 0) {
          covar &lt;- diag(ncol(Y))
          stop(&quot;Computed covariance matrix has zero rows&quot;)
        }
        }, error = function(e) {
          if (!is.null(error_cache)) {
            saveRDS(list(data=Y, message=warning(e)), error_cache)
            warning(&quot;FLASH failed. Using Identity matrix instead.&quot;)
            warning(e)
          } else {
            stop(e)
          }
        })
        s &lt;- apply(Y, 2, sd, na.rm=T)
        if (length(s)&gt;1) s = diag(s)
        else s = matrix(s,1,1)
        covar &lt;- s%*%cov2cor(covar)%*%s
        return(covar)
    }

    read_gene_pheno = function(path){
      arg = paste0(c(&quot;tabix -h &quot;,path,&quot; ${_genoFile[&quot;#chr&quot;]}:${_genoFile[&quot;start&quot;]}-${_genoFile[&quot;start&quot;]+1}&quot;),collapse = &quot;&quot;)
      result = system(arg,intern = T)
        if (is.na(result[2]) == F){
          output = read.table(text = result[2], sep = &quot;\t&quot;)
          colnames(output) = result[1]%&gt;%stringr::str_split(&quot;\t&quot;)%&gt;%unlist()
        }
        else {
            output = NA
            }
        return(output) # add an if so that those genes in geno_file, but not presented in one specific tissue&#39;s pheno_file, can be corrctly left out
    }       

    
    remove_covX = function(X,covar){
    for ( i in 1:ncol(X) ) {
        X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals
        }
     X = scale(X)
    }
    
    ## Load Library
    library(&quot;susieR&quot;)
    library(&quot;plink2R&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;purrr&quot;)
    ###
    # Core code
    ###
    # Input
    ### Genotype
    geno = read_plink(&quot;${path(_genoFile[&quot;path&quot;]):n}&quot;)
    X = filter_X(geno$bed,${imiss}, ${maf} )

    ### Phenotype
    phenotype_list = read_delim(&quot;${_input[0]}&quot;,&quot;\t&quot;)
    traits = phenotype_list$tissue
    covar_list = read_delim(&quot;${_input[1]}&quot;,&quot;\t&quot;)
    rownames(X) = read.table(text = rownames(X), sep= &quot;:&quot; )$V2
    covar_list = covar_list%&gt;%mutate(covar = map(path, ~read_delim(.x,&quot;\t&quot;)%&gt;%select(-`#id`)%&gt;%na.omit%&gt;%t()))
    phenotype_list = inner_join(phenotype_list,covar_list, by = &quot;tissue&quot;)
    phenotype_list = phenotype_list %&gt;% mutate(flag = map(path.x, read_gene_pheno)) %&gt;% dplyr::filter(lengths(flag) &gt;1 ) %&gt;% mutate(Y = map2(path.x,covar, ~read_gene_pheno(.x)%&gt;%select(-c(`#chr`,start,end,${id_name}), rownames(.y))%&gt;%t%&gt;%as.matrix))%&gt;%
                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out
                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),
                                            covar = map2(covar, X_data , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            Y = map2(covar, Y ,~ .y[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            #### Get residue for each of tissue
                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale%&gt;%t%&gt;%as_tibble))
    
    Y_resid = phenotype_list%&gt;%select(Y_resid)%&gt;%tidyr::unnest(Y_resid)%&gt;%t%&gt;%as.matrix # in this part unnest will broad those matrix at smalled dimension (fewer rows/columns), so we should leave out those rows/ columns
    colnames(Y_resid) = phenotype_list$tissue
    
    X_list = phenotype_list%&gt;%mutate( X_resid = map2(X_data,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale))%&gt;%pull(X_resid)       # Remove covariate by tissues
    non_missing = list()
    fitted = list()
    output_list = c(${_output:r,})
    rm_traits = traits[!traits %in% colnames(Y_resid)]
    output_list_valid = output_list[traits %in% colnames(Y_resid)]
    output_list_invalid = output_list[!traits %in% colnames(Y_resid)]
    # Fine-mapping with SuSiE
       for (r in 1:ncol(Y_resid)) {
        non_missing[[r]] = which(!is.na(Y_resid[,r])) 
        st = proc.time()
        X = X_list[[r]]
    print(paste(&quot;Dimension of X matrix:&quot;, nrow(X), ncol(X)))
    print(paste(&quot;Dimension of Y matrix:&quot;, nrow(Y_resid), ncol(Y_resid)))
        fitted[[r]] &lt;- susieR::susie(X[names(non_missing[[r]]),], Y_resid[non_missing[[r]],r],
                           L=10,
                           max_iter=1000,
                           estimate_residual_variance=TRUE,
                           estimate_prior_variance=TRUE,
                           refine=TRUE,compute_univariate_zscore = TRUE, coverage = 0.95 )
        fitted[[r]]$time = proc.time() - st
        fitted[[r]]$cs_corr = susieR:::get_cs_correlation(fitted[[r]], X = X[names(non_missing[[r]]),])
        fitted[[r]]$cs_snps = names(fitted[[r]]$X_column_scale_factors[unlist(fitted[[r]]$sets$cs)])
        fitted[[r]]$variable_name =  names(fitted[[r]]$pip)
        fitted[[r]]$coef =  coef.susie(fitted[[r]])
        fitted[[r]]$input_data = list(&quot;X_resid&quot; = X[names(non_missing[[r]]),], &quot;Y_resid&quot; =Y_resid[non_missing[[r]],r] )
        fitted[[r]]$name = phenotype_list$tissue[r]
        saveRDS(fitted[[r]], output_list_valid[r])
    }
    arg = paste0(c(&quot;touch &quot;,output_list_invalid),collapse = &quot; &quot;)
    system(arg)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[uni_susie_sqtl_1]
parameter: max_L = 10
# remove a variant if it has more than imiss missing individual data
parameter: imiss = 0.1
parameter: maf = 0.05
import pandas as pd
input: phenoFile,covFile, for_each = &quot;genoFile&quot;
output: [f&#39;{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.unisusie.fit.rds&#39; for x in trait  ] 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    ## Define function
    compute_maf &lt;- function(geno){
      f &lt;- mean(geno,na.rm = TRUE)/2
      return(min(f, 1-f))
    }

    compute_missing &lt;- function(geno){
      miss &lt;- sum(is.na(geno))/length(geno)
      return(miss)
    }
    
    mean_impute &lt;- function(geno){
      f &lt;- apply(geno, 2, function(x) mean(x,na.rm = TRUE))
      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] &lt;- f[i]
      return(geno)
    }

    is_zero_variance &lt;- function(x) {
      if (length(unique(x))==1) return(T)
      else return(F)
    }
  
    filter_X &lt;- function(X, missing_rate_thresh, maf_thresh) {
        rm_col &lt;- which(apply(X, 2, compute_missing) &gt; missing_rate_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, compute_maf) &lt; maf_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, is_zero_variance))
        if (length(rm_col)) X &lt;- X[, -rm_col]
        return(mean_impute(X))
    }
  
    compute_cov_flash &lt;- function(Y, error_cache = NULL){
        covar &lt;- diag(ncol(Y))
        tryCatch({
        fl &lt;- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)
        if(fl$n.factors==0){
          covar &lt;- diag(fl$residuals.sd^2)
        } else {
          fsd &lt;- sapply(fl$fitted.g[[1]], &#39;[[&#39;, &quot;sd&quot;)
          covar &lt;- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)
        }
        if (nrow(covar) == 0) {
          covar &lt;- diag(ncol(Y))
          stop(&quot;Computed covariance matrix has zero rows&quot;)
        }
        }, error = function(e) {
          if (!is.null(error_cache)) {
            saveRDS(list(data=Y, message=warning(e)), error_cache)
            warning(&quot;FLASH failed. Using Identity matrix instead.&quot;)
            warning(e)
          } else {
            stop(e)
          }
        })
        s &lt;- apply(Y, 2, sd, na.rm=T)
        if (length(s)&gt;1) s = diag(s)
        else s = matrix(s,1,1)
        covar &lt;- s%*%cov2cor(covar)%*%s
        return(covar)
    }

    
        
      read_gene_pheno = function(path){
          arg = paste0(c(&quot;zgrep  ${_genoFile[&quot;ID&quot;]}  &quot;,path ),collapse = &quot;&quot;)
          result = system(arg,intern = T)
          arg2 = paste0(c(&quot;zcat &quot;, path, &quot; | head -n 1&quot;),collapse = &quot;&quot;)
          result2 = system(arg2,intern = T)
            if (is.na(result) == F){
              output = read.table(text = result, sep = &quot;\t&quot;)
              colnames(output) = result2%&gt;%stringr::str_split(&quot;\t&quot;)%&gt;%unlist()
            }
            else {
                output = NA
                }
            return(output) # add an if so that those genes in geno_file, but not presented in one specific tissue&#39;s pheno_file, can be corrctly left out
            }
  
  
        remove_covX = function(X,covar){
        for ( i in 1:ncol(X) ) {
            X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals
            }
         X = scale(X)
        }
        
    ## Load Library
    library(&quot;susieR&quot;)
    library(&quot;plink2R&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;purrr&quot;)
    ###
    # Core code
    ###
    # Input
    ### Genotype
    geno = read_plink(&quot;${path(_genoFile[&quot;path&quot;]):n}&quot;)
    X = filter_X(geno$bed,${imiss}, ${maf} )

    ### Phenotype
    phenotype_list = read_delim(&quot;${_input[0]}&quot;,&quot;\t&quot;)
    traits = phenotype_list$tissue
    covar_list = read_delim(&quot;${_input[1]}&quot;,&quot;\t&quot;)
    rownames(X) = read.table(text = rownames(X), sep= &quot;:&quot; )$V2
    covar_list = covar_list%&gt;%mutate(covar = map(path, ~read_delim(.x,&quot;\t&quot;)%&gt;%select(-`#id`)%&gt;%na.omit%&gt;%t()))
    phenotype_list = inner_join(phenotype_list,covar_list, by = &quot;tissue&quot;)
    phenotype_list = phenotype_list %&gt;% mutate(flag = map(path.x, read_gene_pheno)) %&gt;% dplyr::filter(lengths(flag) &gt;1 ) %&gt;% mutate(Y = map2(path.x,covar, ~read_gene_pheno(.x)%&gt;%select(-c(`#chr`,start,end,${id_name}), rownames(.y))%&gt;%t%&gt;%as.matrix))%&gt;%
                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out
                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),
                                            covar = map2(covar, X_data , ~.x[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            Y = map2(covar, Y ,~ .y[intersect(.x%&gt;%rownames,rownames(.y)),] ),
                                            #### Get residue for each of tissue
                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale%&gt;%t%&gt;%as_tibble))
    
    Y_resid = phenotype_list%&gt;%select(Y_resid)%&gt;%tidyr::unnest(Y_resid)%&gt;%t%&gt;%as.matrix # in this part unnest will broad those matrix at smalled dimension (fewer rows/columns), so we should leave out those rows/ columns
    colnames(Y_resid) = phenotype_list$tissue
    
    X_list = phenotype_list%&gt;%mutate( X_resid = map2(X_data,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%&gt;%scale))%&gt;%pull(X_resid)       # Remove covariate by tissues
    non_missing = list()
    fitted = list()
    output_list = c(${_output:r,})
    rm_traits = traits[!traits %in% colnames(Y_resid)]
    output_list_valid = output_list[traits %in% colnames(Y_resid)]
    output_list_invalid = output_list[!traits %in% colnames(Y_resid)]
    # Fine-mapping with SuSiE
       for (r in 1:ncol(Y_resid)) {
        non_missing[[r]] = which(!is.na(Y_resid[,r])) 
        st = proc.time()
        X = X_list[[r]]
    print(paste(&quot;Dimension of X matrix:&quot;, nrow(X), ncol(X)))
    print(paste(&quot;Dimension of Y matrix:&quot;, nrow(Y_resid), ncol(Y_resid)))
        fitted[[r]] &lt;- susieR::susie(X[names(non_missing[[r]]),], Y_resid[non_missing[[r]],r],
                           L=10,
                           max_iter=1000,
                           estimate_residual_variance=TRUE,
                           estimate_prior_variance=TRUE,
                           refine=TRUE,compute_univariate_zscore = TRUE, coverage = 0.95 )
        fitted[[r]]$time = proc.time() - st
        fitted[[r]]$cs_corr = susieR:::get_cs_correlation(fitted[[r]], X = X[names(non_missing[[r]]),])
        fitted[[r]]$cs_snps = names(fitted[[r]]$X_column_scale_factors[unlist(fitted[[r]]$sets$cs)])
        fitted[[r]]$variable_name =  names(fitted[[r]]$pip)
        fitted[[r]]$coef =  coef.susie(fitted[[r]])
        fitted[[r]]$input_data = list(&quot;X_resid&quot; = X[names(non_missing[[r]]),], &quot;Y_resid&quot; =Y_resid[non_missing[[r]],r] )
        fitted[[r]]$name = phenotype_list$tissue[r]
        saveRDS(fitted[[r]], output_list_valid[r])
    }
    arg = paste0(c(&quot;touch &quot;,output_list_invalid),collapse = &quot; &quot;)
    system(arg)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>#[uni_susie_2]
input: group_with = &quot;genoFile&quot;
output: f&quot;{_input:n}.vcf.bgz&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:nn}.stdout&quot;, stderr = f&quot;{_output:nn}.stderr&quot;, container = container
   ## Define create_vcf function
           create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, 
        se = NULL, pval = NULL, name = NULL,cs = NULL, pip = NULL) 
    {
        stopifnot(length(chrom) == length(pos))
        if (is.null(snp)) {
            snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        }
        snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        nsnp &lt;- length(chrom)
        gen &lt;- list()
        ## Setupt data content for each sample column
        if (!is.null(ea_af)) 
            gen[[&quot;AF&quot;]] &lt;- matrix(ea_af, nsnp)
        if (!is.null(effect)) 
            gen[[&quot;ES&quot;]] &lt;- matrix(effect, nsnp)
        if (!is.null(se)) 
            gen[[&quot;SE&quot;]] &lt;- matrix(se, nsnp)
        if (!is.null(pval)) 
            gen[[&quot;LP&quot;]] &lt;- matrix(-log10(pval), nsnp)
        if (!is.null(cs)) 
            gen[[&quot;CS&quot;]] &lt;- matrix(cs, nsnp)
        if (!is.null(pip)) 
            gen[[&quot;PIP&quot;]] &lt;- matrix(pip, nsnp)
        gen &lt;- S4Vectors::SimpleList(gen)
        
      ## Setup snps info for the fix columns
        gr &lt;- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, 
            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))
         coldata &lt;- S4Vectors::DataFrame(Studies = name, row.names = name)
    ## Setup header informations
        hdr &lt;- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = &quot;VCFv4.2&quot;, 
            row.names = &quot;fileformat&quot;)), sample = name)
        VariantAnnotation::geno(hdr) &lt;- S4Vectors::DataFrame(Number = c(&quot;A&quot;, 
            &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;), Type = c(&quot;Float&quot;, &quot;Float&quot;, 
            &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;), Description = c(&quot;Effect size estimate relative to the alternative allele&quot;, 
            &quot;Standard error of effect size estimate&quot;, &quot;-log10 p-value for effect estimate&quot;,  
            &quot;Alternate allele frequency in the association study&quot;,
            &quot;The CS this variate are captured, 0 indicates not in any cs&quot;, &quot;The posterior inclusion probability to a CS&quot;), 
            row.names = c(&quot;ES&quot;, &quot;SE&quot;, &quot;LP&quot;, &quot;AF&quot;, &quot;CS&quot;, &quot;PIP&quot;))
      ## Save only the meta information in the sample columns 
        VariantAnnotation::geno(hdr) &lt;- subset(VariantAnnotation::geno(hdr), 
            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))
      ## Save VCF 
        vcf &lt;- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, 
            exptData = list(header = hdr), geno = gen)
        VariantAnnotation::alt(vcf) &lt;- Biostrings::DNAStringSetList(as.list(ea))
        VariantAnnotation::ref(vcf) &lt;- Biostrings::DNAStringSet(nea)
      ## Add fixed values
        VariantAnnotation::fixed(vcf)$FILTER &lt;- &quot;PASS&quot;
          return(sort(vcf))
        }
    library(&quot;susieR&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;readr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;stringr&quot;)
    
    # Get list of cs snps
    susie_list = readRDS(${_input:r})
    susie_tb_ls = list()
    for (i in 1:length(susie_list)){
        susie_tb = tibble( snps =  names(susie_list[[1]]$pip)[which( susie_list[[i]]$pip &gt;= 0)], snps_index = which(( susie_list[[i]]$pip &gt;= 0))  )
        susie_tb_ls[[i]]= susie_tb%&gt;%mutate( cs = map(snps_index,~which( susie_list[[i]]$sets$cs %in% .x))%&gt;%as.numeric%&gt;%replace_na(0),
                                 pip = map_dbl(snps_index,~( susie_list[[i]]$pip[.x])),
                                 coef = map_dbl(snps_index,~(coef.susie( susie_list[[i]])[.x+1])))
        }
    if(length(susie_tb_ls) &gt;= 2){ 
      for(i in 2:length(susie_tb_ls)){
          susie_tb_ls[[i]] = full_join(susie_tb_ls[[i-1]],susie_tb_ls[[i]], by = &quot;snps&quot;) 
        }
    }
    m = c(&quot;cs&quot;,&quot;pip&quot;,&quot;coef&quot;)    
    output = list()
    for(i in m){
    output[[i]] = susie_tb_ls[[length(susie_tb_ls)]]%&gt;%select(contains(i))%&gt;%as.matrix
    }
    snps_tb = susie_tb_ls[[length(susie_tb_ls)]]%&gt;%mutate(
                         chr = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = T)$V1),
                         pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = TRUE)$V2),
                         pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE)$V1),
                         alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V2),
                         ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V3))
    
    snps_tb = snps_tb%&gt;%filter(str_detect(ref, &quot;[ACTG]&quot;) &amp; str_detect(alt, &quot;[ACTG]&quot;))
    output_vcf = create_vcf(
            chrom = snps_tb$chr,
             pos = snps_tb$pos,
             ea = snps_tb$alt,
             nea = snps_tb$ref,
             effect = snps_tb%&gt;%select(contains(&quot;coef&quot;))%&gt;%as.matrix ,
             pip = snps_tb%&gt;%select(contains(&quot;pip&quot;))%&gt;%as.matrix,
             cs = snps_tb%&gt;%select(contains(&quot;cs&quot;))%&gt;%as.matrix,
             name = names(susie_list))
    VariantAnnotation::writeVcf(output_vcf,${_output:nr},index = TRUE)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>#[*_susie_3]
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/{name}.susie.output_list.txt&#39;
python: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container
    import pandas as pd
    pd.DataFrame({&quot;output&quot; : [$[_input:ar,]]}).to_csv(&quot;$[_output]&quot;,index = False ,header = False, sep = &quot;t&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mv_susie_1]
parameter: max_L = 10
# remove a variant if it has more than imiss missing individual data
parameter: imiss = 0.1
parameter: maf = 0.05
# Only analyze `cis` variants -- cis = N means using N variants around the center column of X matrix  
parameter: cis = &#39;NULL&#39;
parameter: prior = path
input: phenoFile,covFile, for_each = &quot;genoFile&quot;
output: mvsusie = f&#39;{wd:a}/{_input:bn}{(&quot;_cis_%s&quot; % cis) if cis != &quot;NULL&quot; else &quot;&quot;}_{name}.mvsusie.rds&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    ###
    # Utility functions
    ###
    compute_maf &lt;- function(geno){
      f &lt;- mean(geno,na.rm = TRUE)/2
      return(min(f, 1-f))
    }

    compute_missing &lt;- function(geno){
      miss &lt;- sum(is.na(geno))/length(geno)
      return(miss)
    }
    
    mean_impute &lt;- function(geno){
      f &lt;- apply(geno, 2, function(x) mean(x,na.rm = TRUE))
      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] &lt;- f[i]
      return(geno)
    }

    is_zero_variance &lt;- function(x) {
      if (length(unique(x))==1) return(T)
      else return(F)
    }
  
    filter_X &lt;- function(X, missing_rate_thresh, maf_thresh) {
        rm_col &lt;- which(apply(X, 2, compute_missing) &gt; missing_rate_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, compute_maf) &lt; maf_thresh)
        if (length(rm_col)) X &lt;- X[, -rm_col]
        rm_col &lt;- which(apply(X, 2, is_zero_variance))
        if (length(rm_col)) X &lt;- X[, -rm_col]
        return(mean_impute(X))
    }

    compute_cov_flash &lt;- function(Y, error_cache = NULL){
        covar &lt;- diag(ncol(Y))
        tryCatch({
        fl &lt;- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)
        if(fl$n.factors==0){
          covar &lt;- diag(fl$residuals.sd^2)
        } else {
          fsd &lt;- sapply(fl$fitted.g[[1]], &#39;[[&#39;, &quot;sd&quot;)
          covar &lt;- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)
        }
        if (nrow(covar) == 0) {
          covar &lt;- diag(ncol(Y))
          stop(&quot;Computed covariance matrix has zero rows&quot;)
        }
        }, error = function(e) {
          if (!is.null(error_cache)) {
            saveRDS(list(data=Y, message=warning(e)), error_cache)
            warning(&quot;FLASH failed. Using Identity matrix instead.&quot;)
            warning(e)
          } else {
            stop(e)
          }
        })
        s &lt;- apply(Y, 2, sd, na.rm=T)
        if (length(s)&gt;1) s = diag(s)
        else s = matrix(s,1,1)
        covar &lt;- s%*%cov2cor(covar)%*%s
        return(covar)
    }
  
    compute_cov_diag &lt;- function(Y){
        covar &lt;- diag(apply(Y, 2, var, na.rm=T))
        return(covar)
    }

    get_center &lt;- function(k,n) {
      ## For given number k, get the range k surrounding n/2
      ## but have to make sure it does not go over the bounds
      if (is.null(k)) {
          return(1:n)
      }
      start = floor(n/2 - k/2)
      end = floor(n/2 + k/2)
      if (start&lt;1) start = 1
      if (end&gt;n) end = n
      return(start:end)
    }
    
    get_prior_indices &lt;- function(Y, U) {
      # make sure the prior col/rows match the colnames of the Y matrix
      y_names = colnames(Y)
      u_names = colnames(U)
      if (is.null(y_names) || is.null(u_names)) {
          return(NULL)
      } else if (identical(y_names, u_names)) {
          return(NULL)
      } else {
          return(match(y_names, u_names))
      }
    }
    
    ###
    # Core code
    ###
    # Input
    ### Genotype
    geno = read_plink(&quot;${path(_genoFile[&quot;path&quot;]):n}&quot;)
    X = filter_X(geno$bed,${imiss}, ${maf} )
    X = X[,get_center(${cis}, ncol(X))]
    ### Phenotype
    phenotype_list = read_delim(&quot;${_input[0]}&quot;,&quot;\t&quot;)
    covar_list = read_delim(&quot;${_input[1]}&quot;,&quot;\t&quot;)
    covar_list = covar_list%&gt;%mutate(covar = map(path, ~read_delim(.x,&quot;\t&quot;)%&gt;%select(-`#id`)%&gt;%na.omit%&gt;%t()))
    phenotype_list = inner_join(phenotype_list,covar_list, by = &quot;tissue&quot;)
    phenotype_list = phenotype_list%&gt;%mutate(Y = map(path.x, ~read_gene_pheno(.x)%&gt;%select(-c(`#chr`,start,end,${id_name}))%&gt;%t%&gt;%as.matrix))%&gt;%mutate(
                                            #### Get residue for each of tissue
                                              Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x[rownames(.y),])$residuals%&gt;%t%&gt;%as_tibble))
    
    Y_resid = phenotype_list%&gt;%select(Y_resid)%&gt;%tidyr::unnest(Y_resid)%&gt;%t%&gt;%as.matrix
    colnames(Y_resid) = phenotype_list$tissue
    # FIXME: handle it when prior does not exist
    prior = readRDS(${prior:r})
    print(paste(&quot;Number of components in the mixture prior:&quot;, length(prior$U)))
    prior = mvsusieR::create_mash_prior(mixture_prior=list(weights=prior$w, matrices=prior$U), include_indices = get_prior_indices(Y_resid, prior$U[[1]]), max_mixture_len=-1)
    print(paste(&quot;Dimension of X matrix:&quot;, nrow(X), ncol(X)))
    print(paste(&quot;Dimension of Y matrix:&quot;, nrow(Y_resid), ncol(Y_resid)))
      

    # GWAS Summary statistics
    univariate_res = lapply(1:ncol(Y_resid), function(r) susieR:::univariate_regression(X[non_missing[[r]], ], Y_resid[non_missing[[r]], r]))
    bhat = do.call(cbind, lapply(1:ncol(Y_resid), function(r) univariate_res[[r]]$betahat))
    sbhat = do.call(cbind, lapply(1:ncol(Y_resid), function(r) univariate_res[[r]]$sebetahat))
    saveRDS(list(bhat=bhat, sbhat=sbhat), &quot;${_output[0]:nn}.sumstat.rds&quot;)
    rm(bhat)
    rm(sbhat)
    # Multivariate fine-mapping
    st = proc.time()
    mv_res = mvsusieR::mvsusie(X, Y_resid, L=${max_L}, 
                              prior_variance=prior, residual_variance=resid_Y, 
                              precompute_covariances=F, compute_objective=T, 
                              estimate_residual_variance=F, estimate_prior_variance=T, estimate_prior_method=&#39;EM&#39;,
                              max_iter = 100, n_thread=1, approximate=F)
    mv_res$time = proc.time() - st
    mv_res$cs_corr = susieR:::get_cs_correlation(mv_res, X=X)
    saveRDS(mv_res, ${_output[0]:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mv_susie_2]
input: group_with = &quot;genoFile&quot;
output: f&quot;{_input:n}.vcf.bgz&quot;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;2h&#39;, mem = &#39;55G&#39;, cores = 1, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:nn}.stdout&quot;, stderr = f&quot;{_output:nn}.stderr&quot;, container = container
   ## Define create_vcf function
           create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, 
        se = NULL, pval = NULL, name = NULL,cs = NULL, pip = NULL) 
    {
        stopifnot(length(chrom) == length(pos))
        if (is.null(snp)) {
            snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        }
        snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        nsnp &lt;- length(chrom)
        gen &lt;- list()
        ## Setupt data content for each sample column
        if (!is.null(ea_af)) 
            gen[[&quot;AF&quot;]] &lt;- matrix(ea_af, nsnp)
        if (!is.null(effect)) 
            gen[[&quot;ES&quot;]] &lt;- matrix(effect, nsnp)
        if (!is.null(se)) 
            gen[[&quot;SE&quot;]] &lt;- matrix(se, nsnp)
        if (!is.null(pval)) 
            gen[[&quot;LP&quot;]] &lt;- matrix(-log10(pval), nsnp)
        if (!is.null(cs)) 
            gen[[&quot;CS&quot;]] &lt;- matrix(cs, nsnp)
        if (!is.null(pip)) 
            gen[[&quot;PIP&quot;]] &lt;- matrix(pip, nsnp)
        gen &lt;- S4Vectors::SimpleList(gen)
        
      ## Setup snps info for the fix columns
        gr &lt;- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, 
            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))
         coldata &lt;- S4Vectors::DataFrame(Studies = name, row.names = name)
    ## Setup header informations
        hdr &lt;- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = &quot;VCFv4.2&quot;, 
            row.names = &quot;fileformat&quot;)), sample = name)
        VariantAnnotation::geno(hdr) &lt;- S4Vectors::DataFrame(Number = c(&quot;A&quot;, 
            &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;), Type = c(&quot;Float&quot;, &quot;Float&quot;, 
            &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;), Description = c(&quot;Effect size estimate relative to the alternative allele&quot;, 
            &quot;Standard error of effect size estimate&quot;, &quot;-log10 p-value for effect estimate&quot;,  
            &quot;Alternate allele frequency in the association study&quot;,
            &quot;The CS this variate are captured, 0 indicates not in any cs&quot;, &quot;The posterior inclusion probability to a CS&quot;), 
            row.names = c(&quot;ES&quot;, &quot;SE&quot;, &quot;LP&quot;, &quot;AF&quot;, &quot;CS&quot;, &quot;PIP&quot;))
      ## Save only the meta information in the sample columns 
        VariantAnnotation::geno(hdr) &lt;- subset(VariantAnnotation::geno(hdr), 
            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))
      ## Save VCF 
        vcf &lt;- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, 
            exptData = list(header = hdr), geno = gen)
        VariantAnnotation::alt(vcf) &lt;- Biostrings::DNAStringSetList(as.list(ea))
        VariantAnnotation::ref(vcf) &lt;- Biostrings::DNAStringSet(nea)
      ## Add fixed values
        VariantAnnotation::fixed(vcf)$FILTER &lt;- &quot;PASS&quot;
          return(sort(vcf))
        }
    library(&quot;susieR&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;readr&quot;)
    library(&quot;tidyr&quot;)
    
    # Get list of cs snps
    res = readRDS(${_input:r})
    output_snps = tibble( snps = res$variable_name[which(res$pip &gt;= 0)], snps_index = which((res$pip &gt;= 0))  )
    output_snps = output_snps%&gt;%mutate( cs = map(snps_index,~which(res$sets$cs %in% .x))%&gt;%as.numeric%&gt;%replace_na(0),
                             pip = map_dbl(snps_index,~(res$pip[.x])),
                     chr = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = T)$V1),
                     pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = TRUE)$V2),
                     pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE)$V1),
                     alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V2),
                     ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V3))
    
    effect_mtr = res$coef[output_snps$snps_index+1]%&gt;%as.matrix
    colnames(effect_mtr) = &quot;${name}&quot;
    rownames(effect_mtr) = output_snps$snps
    cs_mtr = effect_mtr
    for(i in 1:nrow(cs_mtr)) cs_mtr[i,] =  output_snps$cs[[i]]  
    pip_mtr = effect_mtr
    for(i in 1:nrow(pip_mtr)) pip_mtr[i,] =  output_snps$pip[[i]]  
    
    output_vcf = create_vcf(
           chrom = output_snps$chr,
            pos = output_snps$pos,
            ea = output_snps$alt,
            nea = output_snps$ref,
            effect = effect_mtr ,
            pip = pip_mtr,
            cs = cs_mtr,
            name = colnames(effect_mtr)
              )
    VariantAnnotation::writeVcf(output_vcf,${_output:nr},index = TRUE)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/fine_mapping/SuSiE"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="SuSiE_RSS.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Fine-mapping with SuSiE RSS model</p>
      </div>
    </a>
    <a class="right-next"
       href="../multivariate_fine_mapping.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Multivariate fine-mapping workflows</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-list">Example genotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mwe">MWE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#fsusie">FSuSiE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-fsusie">Univariate fSuSiE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-f-susie">Multi F SuSiE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-susie">Univariate SuSiE</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>