

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Summary statistics standardization and export &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/misc/summary_stats_standardizer';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Summary statistics in VCF format" href="rds_to_vcf.html" />
    <link rel="prev" title="Fine-mapping with PolyFun" href="../integrative_analysis/SuSiE_Ann/polyfun.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/misc/summary_stats_standardizer.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/misc/summary_stats_standardizer.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/misc/summary_stats_standardizer.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Summary statistics standardization and export</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview-of-design">Overview of design</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#individual-xqtl-studies">Individual xQTL studies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-xqtl-integration">Multiple xQTL integration</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#column-name-standardization">Column name standardization</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#software-input-headers">Software (input) headers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#our-effect-level-summary">Our effect level summary</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trait-level-qtl-summary-multiple-testing-corrected">Trait level QTL summary (multiple-testing corrected)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#some-technical-notes">Some technical notes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pre-requisites">Pre-requisites</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-minimal-format-of-the-input-yaml-file">The minimal format of the input yaml file</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#memory-usage">Memory usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mwe-example-command">MWE Example command</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#target-generation">Target generation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sumstat-standardization">sumstat_standardization</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-codes">Workflow codes</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="summary-statistics-standardization-and-export">
<h1>Summary statistics standardization and export<a class="headerlink" href="#summary-statistics-standardization-and-export" title="Permalink to this headline">#</a></h1>
<p>This pipeline module contains codes to process summary statistics from conventional QTL association scan to standard formats for public distribution. It will also export multiple QTL studies to formats easily accessible for data integration methods to query and analyze the summary statistics.</p>
<section id="overview-of-design">
<h2>Overview of design<a class="headerlink" href="#overview-of-design" title="Permalink to this headline">#</a></h2>
<section id="individual-xqtl-studies">
<h3>Individual xQTL studies<a class="headerlink" href="#individual-xqtl-studies" title="Permalink to this headline">#</a></h3>
<ol class="arabic simple">
<li><p>Reorganize the QTL marginal statistics to standard formats (see section <code class="docutils literal notranslate"><span class="pre">Column</span> <span class="pre">name</span> <span class="pre">standardization</span></code> for details).</p></li>
<li><p>Report separately cis-QTL and trans-QTL. Additionally report filtered results, e.g. QTLs that survived multiple-testing correction.</p></li>
<li><p>Our column conventions are based on studies such as GTEx and eQTL category. Our design is “modular” in the sense that we do not provide information that could be trivially annotated after, such as <strong>rsID, gene symbol, gene biotype, gene start and end positions</strong>; or for information that can be inferred from other columns such as type of allele (SNP or INDEL).</p></li>
<li><p>We use GRCh38 reference allele and alternative allele, and the effect allele is adjusted, as necessary, to the alternative allele.</p></li>
<li><p>Summary statistics will be in conventional TSV format which can be converted to <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02248-0">GWAS-VCF format</a> as needed.</p></li>
</ol>
</section>
<section id="multiple-xqtl-integration">
<h3>Multiple xQTL integration<a class="headerlink" href="#multiple-xqtl-integration" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>We will <strong>NOT</strong> standardize integrative results. Instead we will distribute the original outputs from the data integration methods we chose.</p></li>
<li><p>For meta-analysis we’ll rely on output from <a class="reference external" href="http://genetics.cs.ucla.edu/meta/">METASOFT</a> using <code class="docutils literal notranslate"><span class="pre">effect_id</span></code> as SNP ID, created from <code class="docutils literal notranslate"><span class="pre">variant</span></code>, <code class="docutils literal notranslate"><span class="pre">molecular_trait_id</span></code> and <code class="docutils literal notranslate"><span class="pre">molecular_trait_object_id</span></code> as needed.</p></li>
</ul>
<p>To build internal database for multiple xQTL data-sets, we:</p>
<ol class="arabic simple">
<li><p>Include variants that presents in at least one xQTL.</p></li>
<li><p>Unify allele strand and frequency flips to GRCh38 reference.</p></li>
</ol>
</section>
</section>
<section id="column-name-standardization">
<h2>Column name standardization<a class="headerlink" href="#column-name-standardization" title="Permalink to this headline">#</a></h2>
<p>The header of actual output sumstat depends on how we configure it (see section <code class="docutils literal notranslate"><span class="pre">Input</span></code> for details). However, all of them will have the column of <code class="docutils literal notranslate"><span class="pre">chromosome,</span> <span class="pre">position,</span> <span class="pre">ref,</span> <span class="pre">alt,</span> <span class="pre">variant_id,</span> <span class="pre">beta,</span> <span class="pre">se,</span> <span class="pre">pvalue</span></code>.</p>
<section id="software-input-headers">
<h3>Software (input) headers<a class="headerlink" href="#software-input-headers" title="Permalink to this headline">#</a></h3>
<p>For example, when the input sumstat is from TensorQTL, the column specification is:</p>
<ul class="simple">
<li><p>GENE: Molecular trait identifier.(gene)</p></li>
<li><p>CHR: Variant chromosome.</p></li>
<li><p>POS: Variant chromosomal position (basepairs).</p></li>
<li><p>A0: Variant reference allele (A, C, T, or G).</p></li>
<li><p>A1: Variant alternate allele.</p></li>
<li><p>TSS_D: Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>AF: The allele frequency of this SNPs</p></li>
<li><p>MA_SAMPLES: Number of samples carrying the minor allele</p></li>
<li><p>MA_COUNT: Total number of minor alleles across individuals</p></li>
<li><p>P: Nominal P-value from linear regression</p></li>
<li><p>STAT: Slope of the linear regression</p></li>
<li><p>SE: Standard error of beta</p></li>
</ul>
<p>when the input sumstat is from APEX, the column specification is:</p>
<ul class="simple">
<li><p>GENE: Molecular trait identifier.(gene)</p></li>
<li><p>CHR: Variant chromosome.</p></li>
<li><p>POS: Variant chromosomal position (basepairs).</p></li>
<li><p>A0: Variant reference allele (A, C, T, or G).</p></li>
<li><p>A1: Variant alternate allele.</p></li>
<li><p>P: Nominal P-value from linear regression</p></li>
<li><p>STAT: Slope of the linear regression</p></li>
<li><p>SE: Standard error of beta</p></li>
</ul>
</section>
<section id="our-effect-level-summary">
<h3>Our effect level summary<a class="headerlink" href="#our-effect-level-summary" title="Permalink to this headline">#</a></h3>
<p>Our proposed xQTL summary statistics fields should include (cf. <a class="reference external" href="https://www.niagads.org/adsp/content/xqtl-fileformats-110921v3sharedxlsx">our xQTL format draft V3, Jan 2021</a>, <a class="reference external" href="https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources/blob/master/tabix/Columns.md">eQTL catalog</a>, <a class="reference external" href="https://www.gtexportal.org/home/datasets">GTEx</a>, <a class="reference external" href="https://www.metabrain.nl/">metaBrain</a>):</p>
<ul class="simple">
<li><p><strong>variant</strong> - The variant ID (chromosome_position_ref_alt) e.g. chr19_226776_C_T. Based on GRCh38 coordinates and reference genome, with ‘chr’ prefix added to the chromosome number.</p></li>
<li><p><strong>chromosome</strong> - GRCh38 chromosome name of the variant (e.g. 1,2,3 …,X).</p></li>
<li><p><strong>position</strong> - GRCh38 position of the variant.</p></li>
<li><p><strong>ref</strong> - GRCh38 reference allele.</p></li>
<li><p><strong>alt</strong> - GRCh38 alternative allele (also the effect allele).</p></li>
<li><p><strong>imputation_quality</strong> - Optional imputation quality score from the imputation software, can be replaced with NA if not available.</p></li>
<li><p><strong>molecular_trait_object_id</strong> - For phenotypes with multiple correlated alternatives (multiple alternative transcripts or exons within a gene, multple alternative promoters in txrevise, multiple alternative intons in Leafcutter), this defines the level at which the phenotypes were aggregated. Permutation p-values are calculated across this set of alternatives.</p></li>
<li><p><strong>molecular_trait_id</strong> - ID of the molecular trait used for QTL mapping. Depending on the quantification method used, this can be either a gene id, exon id, transcript id or a txrevise promoter, splicing or 3’end event id. Examples: ENST00000356937, ENSG00000008128.</p></li>
<li><p><strong>maf</strong> - Minor allele frequency within a QTL mapping context (e.g. cell type or tissues within a study).</p></li>
<li><p><strong>beta</strong> - Regression coefficient from the linear model <strong>with respect to alternative allele</strong>.</p></li>
<li><p><strong>se</strong> - Standard error of the beta.</p></li>
<li><p><strong>pvalue</strong> - Nominal p-value of association between the variant and the molecular trait.</p></li>
<li><p><strong>n</strong> - Total number of samples without missing data.</p></li>
<li><p><strong>ac</strong> - Count of the alternative allele.</p></li>
<li><p><strong>ma_samples</strong> - Number of samples carrying at least one copy of the minor allele.</p></li>
</ul>
</section>
<section id="trait-level-qtl-summary-multiple-testing-corrected">
<h3>Trait level QTL summary (multiple-testing corrected)<a class="headerlink" href="#trait-level-qtl-summary-multiple-testing-corrected" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><strong>molecular_trait_object_id</strong></p></li>
<li><p><strong>molecular_trait_id</strong></p></li>
<li><p><strong>n_traits</strong> - The number of molecular traits over which permutation p-values were calculated (e.g. the number of transcripts per gene). Note that the permutations are performed accross all molecular traits within the same molecular trait object (e.g. all transcripts of a gene) and the results are reported for the most significant variant and molecular trait pair.</p></li>
<li><p><strong>n_variants</strong> - number of genetic variants tested within the cis region of the molecular trait.</p></li>
<li><p><strong>variant</strong></p></li>
<li><p><strong>chromosome</strong> - GRCh38 chromosome name of the variant (e.g. 1,2,3 …,X).</p></li>
<li><p><strong>position</strong> - GRCh38 position of the variant.</p></li>
<li><p><strong>ref</strong> - GRCh38 reference allele.</p></li>
<li><p><strong>alt</strong> - GRCh38 alternative allele (also the effect allele).</p></li>
<li><p><strong>p_perm</strong> - Empirical p-value calculated from 1000 permutations.</p></li>
<li><p><strong>p_beta</strong> - Estimated empirical p-value based on the beta distribution. This is the column that you want to use for filtering the results. See the FastQTL <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btv722">paper</a> for more details.</p></li>
<li><p><strong>qvalue</strong> - FDR based on Storey’s q-value.</p></li>
</ul>
<p>Other summary:</p>
<ul class="simple">
<li><p>Quantiles of molecular phenotypes</p></li>
</ul>
</section>
</section>
<section id="some-technical-notes">
<h2>Some technical notes<a class="headerlink" href="#some-technical-notes" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p>If there are duplicated INDELs in the summary statistics, they will be removed. For example, two SNPs at 10000 on chr1. one’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> is <code class="docutils literal notranslate"><span class="pre">T</span></code>, and <code class="docutils literal notranslate"><span class="pre">A1</span></code> is <code class="docutils literal notranslate"><span class="pre">TC</span></code>. Whereas the other one’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> is <code class="docutils literal notranslate"><span class="pre">TC</span></code>, and <code class="docutils literal notranslate"><span class="pre">A1</span></code> is <code class="docutils literal notranslate"><span class="pre">T</span></code>. Both of them will be removed. More about INDEL issues(<a class="github reference external" href="https://github.com/statgenetics/UKBB_GWAS_dev/issues/81#issuecomment-1015556800">statgenetics/UKBB_GWAS_dev#81</a>). For SNPs, <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> can be easily standardized to ref/alt in GRCh38 reference genome.</p></li>
<li><p>If duplicated <code class="docutils literal notranslate"><span class="pre">chr:pos</span></code> (GWAS) or <code class="docutils literal notranslate"><span class="pre">gene:chr:pos</span></code> (TWAS) exist, run a recursive match for each pair of them between two summary statistic files (<code class="docutils literal notranslate"><span class="pre">query</span></code>(each of inputs) and <code class="docutils literal notranslate"><span class="pre">subject</span></code> (target file)).</p></li>
<li><p>under the same <code class="docutils literal notranslate"><span class="pre">chr:pos</span></code> or <code class="docutils literal notranslate"><span class="pre">gene:chr:pos</span></code>, The variants’ <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> are matched by exact, flip, reverse, or flip+reverse models. Only one of them is <code class="docutils literal notranslate"><span class="pre">True</span></code>, the variant in two files are matched. If they are matched by flip or flip+reverse, the sign of <code class="docutils literal notranslate"><span class="pre">query</span></code>’s <code class="docutils literal notranslate"><span class="pre">STAT</span></code> will be inversed. And the <code class="docutils literal notranslate"><span class="pre">query</span></code>’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> will be the same as the <code class="docutils literal notranslate"><span class="pre">subject</span></code>’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code>.  <strong>FIXME: should we standardize it to GRCh38 first?</strong></p></li>
</ol>
</section>
<section id="pre-requisites">
<h2>Pre-requisites<a class="headerlink" href="#pre-requisites" title="Permalink to this headline">#</a></h2>
<p>Make sure you install the pre-requisited before running this notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">cugg</span>
</pre></div>
</div>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--cwd</span></code>, the path of working directory</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--yml_list</span></code>, the path to a list of yaml file</p></li>
<li></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--keep-ambiguous</span></code>, boolean. default False. if add –keep-ambiguous parameter, keep ambiguous alleles which can not be decided from flip or reverse, such as A/T or C/G. Otherwise, remove them.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--intersect</span></code>, boolean. default False. if add –intersect parameter, output intersect SNPs in all input files.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--TARGET_list</span></code>a path to a list of reference file, with the column name CHR, POS, REF, ALT that represent the correct reference allele. If these reference is not availble, it can be generated by the TARGET_generation step of this workflow.</p></li>
<li><p>TARGET</p>
<ul>
<li><p>The target file is a reference summary statistic file or a file with at least variant ID relevant columns. When provided with standard <code class="docutils literal notranslate"><span class="pre">chr,</span> <span class="pre">pos,</span> <span class="pre">ref,</span> <span class="pre">alt</span></code> based on GRCh38, it can serve the purpose to standardize the REF/ALT alleles.</p></li>
</ul>
</li>
</ul>
<section id="the-minimal-format-of-the-input-yaml-file">
<h3>The minimal format of the input yaml file<a class="headerlink" href="#the-minimal-format-of-the-input-yaml-file" title="Permalink to this headline">#</a></h3>
<p>For GWAS summary statistics,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">INPUT:</span>
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/*.</span><span class="n">gz</span><span class="p">:</span>
        <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span><span class="n">position</span><span class="p">,</span><span class="n">ref</span><span class="p">,</span><span class="n">alt</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">position</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">ref</span><span class="p">:</span> <span class="n">A0</span>
        <span class="n">alt</span><span class="p">:</span> <span class="n">A1</span>
        <span class="n">beta</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">se</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">pvalue</span><span class="p">:</span> <span class="n">P</span>
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">flip</span><span class="o">/</span><span class="n">snps500_flip</span><span class="o">.</span><span class="n">regenie</span><span class="o">.</span><span class="n">snp_stats</span><span class="o">.</span><span class="n">gz</span><span class="p">:</span>
        <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span><span class="n">position</span><span class="p">,</span><span class="n">ref</span><span class="p">,</span><span class="n">alt</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">position</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">ref</span><span class="p">:</span> <span class="n">A0</span>
        <span class="n">alt</span><span class="p">:</span> <span class="n">A1</span>
        <span class="n">beta</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">se</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">pvalue</span><span class="p">:</span> <span class="n">P</span>
        
<span class="kn">OUTPUT:</span> <span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">output</span><span class="o">/</span>
</pre></div>
</div>
<p>For xQTL summary statistics, <code class="docutils literal notranslate"><span class="pre">molecular_trait_object_id</span></code> is required because a variant can be made association with multiple molecular traits.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">INPUT:</span>
  <span class="o">-</span> <span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/*.</span><span class="n">txt</span><span class="p">:</span>
        <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">molecular_trait_id</span><span class="p">,</span><span class="n">chromosome</span><span class="p">,</span><span class="n">position</span><span class="p">,</span><span class="n">ref</span><span class="p">,</span><span class="n">alt</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">position</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">ref</span><span class="p">:</span> <span class="n">A0</span>
        <span class="n">alt</span><span class="p">:</span> <span class="n">A1</span>
        <span class="n">molecular_trait_id</span><span class="p">:</span> <span class="n">GENE</span>
        <span class="n">beta</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">se</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">pvalue</span><span class="p">:</span> <span class="n">P</span> 

<span class="kn">OUTPUT:</span> <span class="o">../</span><span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/</span><span class="n">output</span><span class="o">/</span>
</pre></div>
</div>
<p><strong>Cautions: Be mindful that the ID component should not have any space or tabs between items</strong></p>
<p>There are three parts in the input yaml file.</p>
<ul class="simple">
<li><p>INPUT</p>
<ul>
<li><p>A list of yml file, as the output from yml_generator, each yml file documents a set of input</p>
<ul>
<li><p>the input summary statistic files with the column names in below.</p></li>
<li><p>the input files can be from multiple directory and from different format. The input paths must follow the rules related to Unix shell. the format is to pair the column names with required keys. If not provided, the column names of the input file will be considered as the default keys.</p></li>
<li><p>The input summary statistic file cannot have duplicated chr:pos</p></li>
<li><p>The input summary statstic file cannot have # in its header
-<code class="docutils literal notranslate"><span class="pre">variant</span></code> in yml is the rule to generate a unique identifier for each SNP, the content of variant ID shall be a combination of other columns such as chrom, position, ref, alt, build, but not taken from existing id columns in the original file.</p></li>
</ul>
</li>
</ul>
</li>
<li><p>OUTPUT</p>
<ul>
<li><p>the path of an output directory for new summary statistic files</p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">#</a></h2>
<p>New summary statistic files with common SNPs in all input files. the sign of statistics has been corrected to make it consistent in different data.</p>
<ul class="simple">
<li><p>for each input sumstat file, a standardized version of it will be generated.</p></li>
<li><p>The generated sumstat files will have header standardized header names. The minimal set of headers will be “chromosome, position, ref, alt, variant_id, beta, se, pvalue”</p></li>
<li><p>The generated sumstat files will be in gz format.</p></li>
</ul>
</section>
<section id="memory-usage">
<h2>Memory usage<a class="headerlink" href="#memory-usage" title="Permalink to this headline">#</a></h2>
<p>For merging two sumstat with ~85000 rows and of size of ~5MB, 1 GB of memory is needed</p>
<p>For merging two sumstat with ~2000000 rows and of size of ~1 GB, at least 50 GB of memory is needed.</p>
</section>
<section id="mwe-example-command">
<h2>MWE Example command<a class="headerlink" href="#mwe-example-command" title="Permalink to this headline">#</a></h2>
<section id="target-generation">
<h3>Target generation<a class="headerlink" href="#target-generation" title="Permalink to this headline">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">TARGET_generation</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">yml</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">yml_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">2</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="n">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
</section>
<section id="sumstat-standardization">
<h3>sumstat_standardization<a class="headerlink" href="#sumstat-standardization" title="Permalink to this headline">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">sumstat_standardization</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">yml</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">yml_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">TARGET_list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">TARGET</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">list</span> \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">2</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="n">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
<p>output/data_intergration/TensorQTL/MWE.3.yml.TARGET.ref.list</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">sumstat_to_vcf</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ROSMAP_methy_QTL</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>   \
      <span class="o">--</span><span class="n">cwd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ROSMAP_methy_QTL</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">23</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="kp">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">sumstat_to_vcf</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span>  <span class="n">output_rerun</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>  \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span> <span class="o">-</span><span class="n">n</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">23</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="kp">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">TARGET_generation</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration_new</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">yml</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration_new</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">yml_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">fasta</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">finalizing</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration_new</span><span class="o">/</span><span class="n">TensorQTL</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">22</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="kp">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 
<span class="c1"># Work directory where output will be saved to</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>

<span class="c1">#if add --keep-ambiguous parameter, keep ambiguous alleles which can not be decided from flip or reverse, such as A/T or C/G. Otherwise, remove them.</span>
<span class="kn">parameter:</span> <span class="n">keep_ambiguous</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># if add --intersect parameter, output intersect SNPs in all input files.</span>
<span class="kn">parameter:</span> <span class="n">intersect</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Containers that contains the necessary packages</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Walltime </span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;5h&#39;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;3G&#39;</span>
<span class="c1"># The directory of the output sumstat</span>
<span class="kn">parameter:</span> <span class="n">sumstat_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">sumstat_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;#chr&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="kp">name</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;#chr&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="c1">## Whether to rename the Chr name.</span>
<span class="kn">parameter:</span> <span class="n">remame</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;A&quot;</span> <span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span><span class="s2">&quot;Y&quot;</span><span class="p">,</span><span class="s2">&quot;MT&quot;</span><span class="p">],</span><span class="s2">&quot;X&quot;</span> <span class="p">:</span> <span class="p">[</span> <span class="sa">f</span><span class="s1">&#39;chr</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span><span class="s2">&quot;Y&quot;</span><span class="p">,</span><span class="s2">&quot;MT&quot;</span><span class="p">]]})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/chr_name&#39;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">None</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="workflow-codes">
<h2>Workflow codes<a class="headerlink" href="#workflow-codes" title="Permalink to this headline">#</a></h2>
<p>The first session is to generate a stand alone target file that can be changed into vcf and then standardized based on GTF. It include three step: Take the union of all snps without allele fliping, create a pseudo-vcf file, use bcftools to standardized the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[TARGET_generation_1]
## path to a list of yml file , with columns #chr and dir
parameter: yml_list = path
import pandas as pd
yml_path = pd.read_csv(yml_list,sep = &quot;\t&quot;).values.tolist()
chr_inv = [x[0] for x in yml_path]
file_inv = [x[1] for x in yml_path]
input: file_inv , group_by = 1, group_with = &quot;chr_inv&quot;
output: f&#39;{cwd}/{_input:bn}.{_chr_inv}.all_snp.vcf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand = &#39;${ }&#39;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;,container = container
    import os
    import pandas as pd
    from cugg.sumstat import read_sumstat
    from cugg.utils import *
    yml = load_yaml(&quot;${_input}&quot;)
    input_dict = parse_input(yml[&#39;INPUT&#39;])
    ## Remap the YML field name
    new_key = [&quot;molecular_trait_id&quot;,&quot;chromosome&quot;, &quot;position&quot;, &quot;ref&quot;, &quot;alt&quot;,&quot;beta&quot;,&quot;se&quot;,&quot;pvalue&quot;,&quot;variant&quot;]
    old_key = [&quot;GENE&quot; ,  &quot;CHR&quot; ,  &quot;POS&quot; ,   &quot;A0&quot; ,   &quot;A1&quot; ,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;,&quot;SNP&quot;]
    key_switch = pd.DataFrame({&quot;new_key&quot; :new_key ,&quot;old_key&quot; : old_key})
    for x in input_dict:
        input_dict[x] = pd.Series(input_dict[x]).rename(index = key_switch.set_index(&quot;new_key&quot;).old_key.to_dict()).to_dict()
        for i,j in zip(old_key,new_key):
            input_dict[x][&quot;ID&quot;] = input_dict[x][&quot;ID&quot;].replace(j,i)

    def ss_2_vcf(ss_df,name = &quot;name&quot;):
        ## Geno field
        df = pd.DataFrame()
        if &quot;SNP&quot; not in ss_df.columns:
            ss_df[&#39;SNP&#39;] = &#39;chr&#39;+ss_df.CHR.astype(str).str.strip(&quot;chr&quot;) + &#39;:&#39; + ss_df.POS.astype(str) + &#39;_&#39; + ss_df.A0.astype(str) + &#39;_&#39; + ss_df.A1.astype(str)
        df[[&#39;#CHROM&#39;, &#39;POS&#39;, &#39;ID&#39;, &#39;REF&#39;, &#39;ALT&#39;]] = ss_df[[&#39;CHR&#39;, &#39;POS&#39;, &#39;SNP&#39;, &#39;A0&#39;, &#39;A1&#39;]]
        ## Info field(Empty)
        df[&#39;QUAL&#39;] = &quot;.&quot;
        df[&#39;FILTER&#39;] = &quot;PASS&quot;
        df[&#39;INFO&#39;] = &quot;.&quot;
        fix_header = [&quot;SNP&quot;,&quot;A1&quot;,&quot;A0&quot;,&quot;POS&quot;,&quot;CHR&quot;,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]
        header_list = []
        if &quot;GENE&quot; in ss_df.columns:
            df[&#39;ID&#39;] = ss_df[&#39;GENE&#39;] + &quot;:&quot; + ss_df[&#39;SNP&#39;]
            df[&#39;INFO&#39;] = &quot;GENE=&quot; + ss_df[&quot;GENE&quot;]
            fix_header = [&quot;GENE&quot;,&quot;SNP&quot;,&quot;A1&quot;,&quot;A0&quot;,&quot;POS&quot;,&quot;CHR&quot;,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]
            header_list = [&#39;##INFO=&lt;ID=GENE,Number=1,Type=String,Description=&quot;The name of genes&quot;&gt;&#39;]
        ### Fix headers
        import time
        header = &#39;##fileformat=VCFv4.2\n&#39; + \
        &#39;##FILTER=&lt;ID=PASS,Description=&quot;All filters passed&quot;&gt;\n&#39; + \
        f&#39;##fileDate={time.strftime(&quot;%Y%m%d&quot;,time.localtime())}\n&#39;+ \
        &#39;##FORMAT=&lt;ID=STAT,Number=1,Type=Float,Description=&quot;Effect size estimate relative to the alternative allele&quot;&gt;\n&#39; + \
        &#39;##FORMAT=&lt;ID=SE,Number=1,Type=Float,Description=&quot;Standard error of effect size estimate&quot;&gt;\n&#39; + \
        &#39;##FORMAT=&lt;ID=P,Number=1,Type=Float,Description=&quot;The Pvalue corresponding to ES&quot;&gt;&#39; 
        ### Customized Field headers
        for x in ss_df.columns:
            if x not in fix_header:
                Prefix = f&#39;##FORMAT=&lt;ID={x},Number=1,Type=&#39;
                Type = str(type(ss_df[x][0])).replace(&quot;&lt;class \&#39;&quot;,&quot;&quot;).replace(&quot;&#39;&gt;&quot;,&quot;&quot;).replace(&quot;numpy.&quot;,&quot;&quot;).replace(&quot;64&quot;,&quot;&quot;).capitalize().replace(&quot;Int&quot;,&quot;Integer&quot;).replace(&quot;Str&quot;, &quot;String&quot;)
                Surfix = f&#39;,Description=&quot;Customized Field {x}&quot;&gt;&#39;
                header_list.append(Prefix+Type+Surfix)
        ## format and sample field
        df[&#39;FORMAT&#39;] = &quot;:&quot;.join([&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]  + ss_df.drop(fix_header,axis = 1).columns.values.tolist())
        df[f&#39;{name}&#39;] = ss_df.drop( [&quot;SNP&quot;,&quot;A1&quot;,&quot;A0&quot;,&quot;POS&quot;,&quot;CHR&quot;],axis = 1).astype(str).apply(&quot;:&quot;.join,axis = 1)
        ## Rearrangment
        df = df[[&#39;#CHROM&#39;, &#39;POS&#39;, &#39;ID&#39;, &#39;REF&#39;, &#39;ALT&#39;, &#39;QUAL&#39;, &#39;FILTER&#39;, &#39;INFO&#39;,&#39;FORMAT&#39;,f&#39;{name}&#39;]]
        df = df.sort_values([&#39;#CHROM&#39;, &#39;POS&#39;])
        # Add headers
        header = header + &quot;\n&quot;.join(header_list) + &quot;\n&quot;
        return df,header

    ## Verify data uniqueness
    lst_sumstats_file = [ os.path.basename(i) for i in input_dict.keys()]
    if len(set(lst_sumstats_file))&lt;len(lst_sumstats_file): 
        raise Exception(&quot;There are duplicated names in {}&quot;.format(lst_sumstats_file))
    #read all sumstats
    print(input_dict)
    lst_sumstats = {os.path.basename(i):read_sumstat(i,j,) for i,j in input_dict.items()}
    ## Retaining only chrom/pos/ref/alt, and dropping the duplicates (drop twice to reduce mem usage)
    union_snp = pd.concat([x[[&quot;CHR&quot; ,  &quot;POS&quot; ,   &quot;A0&quot; ,   &quot;A1&quot;,&quot;SNP&quot;]].drop_duplicates() for x in lst_sumstats.values() ]).drop_duplicates() 
    ## Create fake header
    union_snp[[&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]] = 1
    sumstats,header = ss_2_vcf(union_snp,&quot;PseudoVCF&quot;)
    with open(${_output:r}, &#39;w&#39;) as f:
        f.write(header)
    sumstats.to_csv(${_output:r}, sep = &quot;\t&quot;, header = True, index = False,mode = &quot;a&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[TARGET_generation_2]
## The reference fasta
parameter: fasta = path
output: f&#39;{_input:nn}.TARGET.ref&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &#39;${ }&#39;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;,container = container
    echo -e &quot;CHR\tPOS\tA0\tA1&quot; &gt; ${_output}
    bgzip -f ${_input}
    tabix -p vcf -f  ${_input}.gz
    ## our fasta required chr* as chromosome name format
    bcftools annotate --rename-chrs ${cwd}/chr_name ${_input}.gz -Oz | \
    bcftools norm  -N --check-ref ws -f ${fasta} | bcftools query -f &#39;%CHROM\t%POS\t%REF\t%ALT\n&#39; &gt;&gt; ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[TARGET_generation_3]
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/TARGET.ref.list&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand = &#39;${ }&#39;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;,container = container
    import pandas as pd
    target_path = [${_input:r,}]
    chrom = [x.split(&quot;.&quot;)[-3].replace(&quot;chr&quot;,&quot;&quot;) for x in target_path ]
    pd.DataFrame({&quot;#chr&quot;: chrom , &quot;TARGET&quot; : target_path }).to_csv(&quot;${_output}&quot;,&quot;\t&quot;,index = False)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sumstat_standardization]
## path to a list of yml file , with columns #chr and dir
parameter: yml_list = path
import pandas as pd
yml_path = pd.read_csv(yml_list,sep = &quot;\t&quot;).values.tolist()
parameter: TARGET_list = path
TARGET_path = pd.read_csv(TARGET_list,sep = &quot;\t&quot;)
yml_path = pd.read_csv(yml_list,sep = &quot;\t&quot;).merge(TARGET_path, on = &quot;#chr&quot;).values.tolist()
file_inv = [x[1] for x in yml_path]
TARGET_inv = [x[2] for x in yml_path]
input: file_inv , group_by = 1, group_with = &quot;TARGET_inv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}&#39;
python: expand = &#39;${ }&#39;, stderr = f&#39;{_input}.stderr&#39;, stdout = f&#39;{_input}.stdout&#39;,container = container
    import os
    import pandas as pd
    from cugg.sumstat import read_sumstat
    from cugg.utils import *
    def snps_match(query,subject,keep_ambiguous=True):
        print(&quot;Total rows of query: &quot;,query.shape[0],&quot;Total rows of subject: &quot;,subject.shape[0])
        if len(query.index[0].split(&#39;_&#39;)[-2].split(&#39;:&#39;))&gt;2:  # The rule of index is GENE:CHR:POS_REF:ALT. So using reverse order can accomodate strange Gene name.
            #gene:chr:pos case
            genes_query = query.index.to_series().apply(lambda x: &quot;:&quot;.join(x.split(&#39;:&#39;)[ : -3]))
            genes_query = query.index.to_series().apply(lambda x: &quot;:&quot;.join(x.split(&#39;:&#39;)[ : -3]))
            query = dict(tuple(query.groupby(genes_query)))
            subject = dict(tuple(subject.groupby(genes_subject)))
            new_query, new_subject = [],[]
            for g in genes_query.unique():
                if g in query.keys() and g in subject.keys():
                    new_q,new_s = snps_match_dup(query[g],subject[g],keep_ambiguous)
                    new_query.append(new_q)
                    new_subject.append(new_s)
            new_query, new_subject=pd.concat(new_query),pd.concat(new_query)
        else:
            #chr:pos case
            new_query, new_subject=snps_match_dup(query,subject,keep_ambiguous)
        return new_query, new_subject
    
    
    
    def snps_match_dup(query,subject,keep_ambiguous=True):
        pm = compare_snps(query,subject)
        if not keep_ambiguous:
            pm = pm[~pm.ambiguous]
        new_subject = subject.loc[pm.sidx]
        #update beta and snp info
        new_query = pd.concat([new_subject.iloc[:,:5].reset_index(),query.loc[pm.qidx].iloc[:,5:].reset_index()],axis=1)
        new_query.loc[list(pm.flip) , &quot;STAT&quot;] = -new_query.STAT[list(pm.flip)]
        return new_query, new_subject
    
    
    def compare_snps(query,subject,only_match=True):
        &#39;&#39;&#39;
        input: query and subject, two data frame with the first five column: chr, pos, snp, a0, a1
        output: data frame included six boolean columns (keep,exact,flip,reverse,both, complement) and two columns of the index query and subject.
        &#39;&#39;&#39;
        query = query.reset_index()
        subject = subject.reset_index()
        compare = query.merge(subject,left_on = [&quot;CHR&quot;,&quot;POS&quot;],right_on =  [&quot;CHR&quot;,&quot;POS&quot;])
        smry = compare.apply(lambda x: allele_match(x.A0_x, x.A1_x, x.A0_y,x.A1_y)+[x.index_x,x.index_y] , axis=1, result_type=&#39;expand&#39;)
        smry.columns = [&#39;keep&#39;,&#39;exact&#39;,&#39;flip&#39;,&#39;reverse&#39;,&#39;ambiguous&#39;,&#39;qidx&#39;,&#39;sidx&#39;]
        print(&quot;Total exact: &quot;,smry.exact.sum(),&quot;Total flip: &quot;,smry.flip.sum(),&quot;Total reverse: &quot;,smry.reverse.sum(),&quot;Total ambiguous: &quot;,smry.ambiguous.sum())
        if only_match:
            smry = smry[smry.keep==True]
        return smry
    yml = &quot;${_input}&quot;
    keep_ambiguous = ${keep_ambiguous}
    intersect = ${intersect}
    print(yml, keep_ambiguous,intersect)
    #parse yaml
    yml = load_yaml(yml)
    input_dict = parse_input(yml[&#39;INPUT&#39;])
    ## Remap the YML field name
    new_key = [&quot;molecular_trait_id&quot;,&quot;chromosome&quot;, &quot;position&quot;, &quot;ref&quot;, &quot;alt&quot;,&quot;beta&quot;,&quot;se&quot;,&quot;pvalue&quot;,&quot;variant&quot;]
    old_key = [&quot;GENE&quot; ,  &quot;CHR&quot; ,  &quot;POS&quot; ,   &quot;A0&quot; ,   &quot;A1&quot; ,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;,&quot;SNP&quot;]
    key_switch = pd.DataFrame({&quot;new_key&quot; :new_key ,&quot;old_key&quot; : old_key})
    for x in input_dict:
        input_dict[x] = pd.Series(input_dict[x]).rename(index = key_switch.set_index(&quot;new_key&quot;).old_key.to_dict()).to_dict()
        for i,j in zip(old_key,new_key):
            input_dict[x][&quot;ID&quot;] = input_dict[x][&quot;ID&quot;].replace(j,i)
    target_dict = &quot;${_TARGET_inv}&quot;
    output_path = yml[&#39;OUTPUT&#39;][0]
    lst_sumstats_file = [ os.path.basename(i) for i in input_dict.keys()]
    print(&#39;Total number of sumstats: &#39;,len(lst_sumstats_file))
    if len(set(lst_sumstats_file))&lt;len(lst_sumstats_file):
        raise Exception(&quot;There are duplicated names in {}&quot;.format(lst_sumstats_file))
    #read all sumstats
    print(input_dict)
    lst_sumstats = {os.path.basename(i):read_sumstat(i,j,) for i,j in input_dict.items()}
    nqs = []
    #Readin the reference target file
    subject = check_indels(read_sumstat(target_dict,None,True)[[&quot;CHR&quot;,&quot;POS&quot;,&quot;A0&quot;,&quot;A1&quot;,&quot;SNP&quot;]]).sort_index()
    for query in lst_sumstats.values():
        #check duplicated indels and remove them.
        query = check_indels(query)
        # Set the snp column to be the second column to satisify the requirement of compare_snps() function
        column =  query.pop(&quot;SNP&quot;)
        query.insert(4,&quot;SNP&quot;, column )
        #under the same chr:pos or gene:chr:pos. match A0 and A1 by exact, flip, reverse, or flip+reverse.
        #if duplicated chr_pos or gene_chr_pos exist, run a recursive match for each pair of them between query and subject.
        # If GENE info is in query but not subject, added it.
        if &quot;GENE&quot; in query.columns and &quot;GENE&quot; not  in subject.columns:
            subject = subject.merge(query[[&quot;GENE&quot;,&quot;CHR&quot;,&quot;POS&quot;]]).drop_duplicates()
            ## It is crucial that the index was built via this function, where the order of A0/A1 was removed. Otherwise will cause error is issue #306
            subject.index = namebyordA0_A1(subject[[&quot;GENE&quot;,&quot;CHR&quot;,&quot;POS&quot;,&quot;A0&quot;,&quot;A1&quot;]],cols=[&quot;GENE&quot;,&quot;CHR&quot;,&quot;POS&quot;,&quot;A0&quot;,&quot;A1&quot;])
            subject = subject.sort_index()
        nq,_ = snps_match(query,subject,keep_ambiguous)
        nq = nq.loc[:,~nq.columns.duplicated()] # Remove duplicated columns due to order of columns difference in subject and query
        nqs.append(nq)
    if intersect:
            #get common snps
            common_snps = set.intersection(*[set(nq.SNP) for nq in nqs])
            print(&#39;Total number of common SNPs: &#39;,len(common_snps))
            #write out new sumstats
            for output_sumstats,nq in zip(lst_sumstats_file,nqs):
                sumstats = nq[nq.SNP.isin(common_snps)]
                sumstats[&quot;SNP&quot;] = &quot;chr&quot; + nq.CHR.astype(str).str.replace(&quot;chr&quot;,&quot;&quot;)  + &quot;_&quot; + sumstats.POS.astype(str) + &quot;_&quot; + sumstats.A0 + &quot;_&quot; + sumstats.A1
                sumstats = sumstats.rename(columns = key_switch.set_index(&quot;old_key&quot;).new_key.to_dict() ).drop(&quot;index&quot;,axis = 1)
                sumstats.to_csv(os.path.join(output_path, output_sumstats), sep = &quot;\t&quot;, header = True, index = False)
    else:
            for output_sumstats,nq in zip(lst_sumstats_file,nqs):
                nq[&quot;SNP&quot;] = &quot;chr&quot; + nq.CHR.astype(str).str.replace(&quot;chr&quot;,&quot;&quot;)  + &quot;_&quot; + nq.POS.astype(str) + &quot;_&quot; + nq.A0 + &quot;_&quot; + nq.A1
                nq = nq.rename(columns = key_switch.set_index(&quot;old_key&quot;).new_key.to_dict() )
                #output match SNPs with target SNPs.
                nq = nq.drop(&quot;index&quot;,axis = 1)
                nq.to_csv(os.path.join(output_path, output_sumstats), sep = &quot;\t&quot;, header = True, index = False)
    print(&quot;Total remaining: &quot;,len(nq) )
    print(&#39;All are done&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sumstat_to_vcf_1 ]
input:  for_each = &quot;sumstat_path&quot;
output: [f&#39;{path(x):an}.vcf&#39; for x in _sumstat_path]
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand = &#39;${ }&#39;, stderr = f&#39;{cwd:a}/{path(_sumstat_path[0]):bn}.stderr&#39;, stdout = f&#39;{cwd:a}/{path(_sumstat_path[0]):bn}.stderr&#39;
    import pandas as pd
    from sos.targets import path
    def ss_2_vcf(ss_df,name = &quot;name&quot;):
        ## Geno field
        df = pd.DataFrame()
        if &quot;SNP&quot; not in ss_df.columns:
            ss_df[&#39;SNP&#39;] = &#39;chr&#39;+ss_df.CHR.astype(str).str.strip(&quot;chr&quot;) + &#39;:&#39; + ss_df.POS.astype(str) + &#39;_&#39; + ss_df.A0.astype(str) + &#39;_&#39; + ss_df.A1.astype(str)
        df[[&#39;#CHROM&#39;, &#39;POS&#39;, &#39;ID&#39;, &#39;REF&#39;, &#39;ALT&#39;]] = ss_df[[&#39;CHR&#39;, &#39;POS&#39;, &#39;SNP&#39;, &#39;A0&#39;, &#39;A1&#39;]].sort_values([&#39;CHR&#39;, &#39;POS&#39;])
        ## Info field(Empty)
        df[&#39;QUAL&#39;] = &quot;.&quot;
        df[&#39;FILTER&#39;] = &quot;PASS&quot;
        df[&#39;INFO&#39;] = &quot;.&quot;
        fix_header = [&quot;SNP&quot;,&quot;A1&quot;,&quot;A0&quot;,&quot;POS&quot;,&quot;CHR&quot;,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]
        header_list = []
        if &quot;GENE&quot; in ss_df.columns:
            df[&#39;ID&#39;] = ss_df[&#39;GENE&#39;] + &quot;:&quot; + ss_df[&#39;SNP&#39;]
            df[&#39;INFO&#39;] = &quot;GENE=&quot; + ss_df[&quot;GENE&quot;]
            fix_header = [&quot;GENE&quot;,&quot;SNP&quot;,&quot;A1&quot;,&quot;A0&quot;,&quot;POS&quot;,&quot;CHR&quot;,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]
            header_list = [&#39;##INFO=&lt;ID=GENE,Number=1,Type=String,Description=&quot;The name of genes&quot;&gt;&#39;]
        ### Fix headers
        import time
        header = &#39;##fileformat=VCFv4.2\n&#39; + \
        &#39;##FILTER=&lt;ID=PASS,Description=&quot;All filters passed&quot;&gt;\n&#39; + \
        f&#39;##fileDate={time.strftime(&quot;%Y%m%d&quot;,time.localtime())}\n&#39;+ \
        &#39;##FORMAT=&lt;ID=STAT,Number=1,Type=Float,Description=&quot;Effect size estimate relative to the alternative allele&quot;&gt;\n&#39; + \
        &#39;##FORMAT=&lt;ID=SE,Number=1,Type=Float,Description=&quot;Standard error of effect size estimate&quot;&gt;\n&#39; + \
        &#39;##FORMAT=&lt;ID=P,Number=1,Type=Float,Description=&quot;The Pvalue corresponding to ES&quot;&gt;&#39;
        ### Customized Field headers
        for x in ss_df.columns:
            if x not in fix_header:
                Prefix = f&#39;##FORMAT=&lt;ID={x},Number=1,Type=&#39;
                Type = str(type(ss_df[x][0])).replace(&quot;&lt;class \&#39;&quot;,&quot;&quot;).replace(&quot;&#39;&gt;&quot;,&quot;&quot;).replace(&quot;numpy.&quot;,&quot;&quot;).replace(&quot;64&quot;,&quot;&quot;).capitalize().replace(&quot;Int&quot;,&quot;Integer&quot;)
                Surfix = f&#39;,Description=&quot;Customized Field {x}&quot;&gt;&#39;
                header_list.append(Prefix+Type+Surfix)
        ## format and sample field
        df[&#39;FORMAT&#39;] = &quot;:&quot;.join([&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;]  + ss_df.drop(fix_header,axis = 1).columns.values.tolist())
        df[f&#39;{name}&#39;] = ss_df.drop( [&quot;SNP&quot;,&quot;A1&quot;,&quot;A0&quot;,&quot;POS&quot;,&quot;CHR&quot;],axis = 1).astype(str).apply(&quot;:&quot;.join,axis = 1)
        ## Rearrangment
        df = df[[&#39;#CHROM&#39;, &#39;POS&#39;, &#39;ID&#39;, &#39;REF&#39;, &#39;ALT&#39;, &#39;QUAL&#39;, &#39;FILTER&#39;, &#39;INFO&#39;,&#39;FORMAT&#39;,f&#39;{name}&#39;]]
        # Add headers
        header = header + &quot;\n&quot; + &quot;\n&quot;.join(header_list) + &quot;\n&quot;
        return df,header
    sumstat_path_list = ${_sumstat_path}
    name = ${name}
    ## Remap the YML field name
    new_key = [&quot;molecular_trait_id&quot;,&quot;chromosome&quot;, &quot;position&quot;, &quot;ref&quot;, &quot;alt&quot;,&quot;beta&quot;,&quot;se&quot;,&quot;pvalue&quot;,&quot;variant&quot;]
    old_key = [&quot;GENE&quot; ,  &quot;CHR&quot; ,  &quot;POS&quot; ,   &quot;A0&quot; ,   &quot;A1&quot; ,&quot;STAT&quot;,&quot;SE&quot;,&quot;P&quot;,&quot;SNP&quot;]
    name_map = {old_key[i]: new_key[i] for i in range(len(old_key))}
    name_map_rev = {new_key[i]: old_key[i] for i in range(len(old_key))}
    for x,y in zip(sumstat_path_list,name):
        sumstats = pd.read_csv(x,&quot;\t&quot;).rename(columns = name_map_rev )
        sumstats,header = ss_2_vcf(sumstats,y)
        with open(f&#39;{path(x):an}.vcf&#39;, &#39;w&#39;) as f:
            f.write(header)
        sumstats.to_csv(f&#39;{path(x):an}.vcf&#39;, sep = &quot;\t&quot;, header = True, index = False,mode = &quot;a&quot;)


bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{cwd:a}/{_output[0]:bn}.stderr&#39;, stdout = f&#39;{cwd:a}/{_output[0]:bn}.stderr&#39;, container = container
        stdout=$[_output[0]:bn].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -v &quot;##&quot;   | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sumstat_to_vcf_2]
skip_if(len(sumstat_path[0]) == 1)
output: f&#39;{cwd}/{_input[0]:bn}.merged.vcf.gz&#39;.replace(name[0],&quot;_&quot;.join(name))
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &#39;${ }&#39;, stderr = f&#39;{cwd:a}/{_output:bn}.stderr&#39;, stdout = f&#39;{cwd:a}/{_output:bn}.stdout&#39;,container = container
    for i in ${_input:r}; do
    bgzip -k -f $i 
    tabix -p vcf -f  $i.gz; done
    bcftools merge ${&quot; &quot;.join([f&#39;{str(x)}.gz&#39; for x in _input])} --force-samples -m id  -Oz -o ${_output:a}
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/misc"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../integrative_analysis/SuSiE_Ann/polyfun.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Fine-mapping with PolyFun</p>
      </div>
    </a>
    <a class="right-next"
       href="rds_to_vcf.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Summary statistics in VCF format</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview-of-design">Overview of design</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#individual-xqtl-studies">Individual xQTL studies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-xqtl-integration">Multiple xQTL integration</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#column-name-standardization">Column name standardization</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#software-input-headers">Software (input) headers</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#our-effect-level-summary">Our effect level summary</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trait-level-qtl-summary-multiple-testing-corrected">Trait level QTL summary (multiple-testing corrected)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#some-technical-notes">Some technical notes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pre-requisites">Pre-requisites</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-minimal-format-of-the-input-yaml-file">The minimal format of the input yaml file</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#memory-usage">Memory usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mwe-example-command">MWE Example command</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#target-generation">Target generation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sumstat-standardization">sumstat_standardization</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-codes">Workflow codes</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>