

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Gene coordinate annotation &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/phenotype/gene_annotation';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Phenotype data formatting" href="phenotype_formatting.html" />
    <link rel="prev" title="Phenotype data preprocessing" href="../phenotype_preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../phenotype_preprocessing.html">Phenotype data preprocessing</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/phenotype/gene_annotation.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/phenotype/gene_annotation.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/data_preprocessing/phenotype/gene_annotation.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Gene coordinate annotation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alternative-implementation">Alternative implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#region-list-generation">Region List generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-pyqtl">Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-biomart">Implementation using biomaRt</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-of-leafcutter-isoform">Annotation of leafcutter isoform</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-of-psichomics-output">Processing of psichomics output</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="gene-coordinate-annotation">
<h1>Gene coordinate annotation<a class="headerlink" href="#gene-coordinate-annotation" title="Permalink to this headline">#</a></h1>
<p>This workflow adds genomic coordinate annotation to gene-level molecular phenotype files generated in <code class="docutils literal notranslate"><span class="pre">gct</span></code> format and convert them to <code class="docutils literal notranslate"><span class="pre">bed</span></code> format for downstreams analysis.</p>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">#</a></h2>
<p>This pipeline is based on <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py"><code class="docutils literal notranslate"><span class="pre">pyqtl</span></code>, as demonstrated here</a>.</p>
<p><strong>FIXME: please explain here what we do with gene symbol vs gene ID</strong></p>
<section id="alternative-implementation">
<h3>Alternative implementation<a class="headerlink" href="#alternative-implementation" title="Permalink to this headline">#</a></h3>
<p>Previously we use <code class="docutils literal notranslate"><span class="pre">biomaRt</span></code> package in R instead of code from <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code>. The core function calls are:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span>    <span class="n">ensembl</span> <span class="o">=</span> <span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;$[ensembl_version]&quot;</span><span class="p">)</span>
    <span class="n">ensembl_df</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s">&quot;start_position&quot;</span><span class="p">,</span> <span class="s">&quot;end_position&quot;</span><span class="p">),</span><span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
</pre></div>
</div>
<p>We require ENSEMBL version to be specified explicitly in this pipeline. As of 2021 for the Brain xQTL project, we use ENSEMBL version 103.</p>
</section>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p>Molecular phenotype data in <code class="docutils literal notranslate"><span class="pre">gct</span></code> format, with the first column being ENSEMBL ID and other columns being sample names.</p></li>
<li><p>GTF for collapsed gene model</p>
<ul class="simple">
<li><p>the gene names must be consistent with the molecular phenotype data matrices (eg ENSG00000000003 vs. ENSG00000000003.1 will not work)</p></li>
</ul>
</li>
<li><p>(Optional) Meta-data to match between sample names in expression data and genotype files</p>
<ul class="simple">
<li><p>Tab delimited with header</p></li>
<li><p>Only 2 columns: first column is sample name in expression data, 2nd column is sample name in genotype data</p></li>
<li><p><strong>must contains all the sample name in expression matrices even if they don’t existing in genotype data</strong></p></li>
</ul>
</li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">#</a></h2>
<p>Molecular phenotype data in <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">#</a></h2>
<p>The MWE is uploaded to the <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk">Google Drive</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">annotate_coord_gene</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">pheno_log2cpm</span><span class="o">.</span><span class="n">tsv</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">annotation</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">sample</span><span class="o">-</span><span class="n">participant</span><span class="o">-</span><span class="n">lookup</span> <span class="n">data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="nb">type</span> <span class="n">gene_name</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">gene_annotation</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">annotate_coord_protein</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">hgrcgrid</span><span class="o">/</span><span class="n">homes</span><span class="o">/</span><span class="n">zq2209</span><span class="o">/</span><span class="n">WashU_BrainProt_442samples_matrix</span><span class="o">.</span><span class="n">csv</span> \
    <span class="o">--</span><span class="n">annotation</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">collapse_only</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">protein_name_index</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ftp_lisanwanglab_sync</span><span class="o">/</span><span class="n">ftp_fgc_xqtl</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">proteomics</span><span class="o">/</span><span class="n">knightadrc</span><span class="o">-</span><span class="n">washu</span><span class="o">/</span><span class="n">WashU_Brain_Somascan1</span><span class="mf">.3</span><span class="n">k_FEATUREinfo</span><span class="o">.</span><span class="n">csv</span> \
    <span class="o">--</span><span class="n">protein_ID_type</span>  <span class="s2">&quot;SOMAseqID&quot;</span>    \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="nb">type</span> <span class="n">gene_name</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run gene_annotation.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  annotate_coord
  annotate_coord_biomart

Global Workflow Options:
  --cwd output (as path)
                        Work directory &amp; output directory
  --annotation-gtf VAL (as path, required)
                        gene gtf annotation table
  --phenoFile VAL (as path, required)
                        Molecular phenotype matrix
  --phenotype-id-type &#39;gene_id&#39;
                        Whether the input data is named by gene_id or gene_name.
                        By default it is gene_id, if not, please change it to
                        gene_name
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 1 (as int)
                        Number of threads
  --container &#39;&#39;

Sections
  annotate_coord:
    Workflow Options:
      --sample-participant-lookup . (as path)
                        A file to map sample ID from expression to genotype,
                        must contain two columns, sample_id and participant_id,
                        mapping IDs in the expression files to IDs in the
                        genotype (these can be the same).
  annotate_coord_biomart:
    Workflow Options:
      --ensembl-version VAL (as int, required)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Work directory &amp; output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1">#  gene gtf annotation table</span>
<span class="kn">parameter:</span> <span class="n">annotation_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Molecular phenotype matrix</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Whether the input data is named by gene_id or gene_name. By default it is gene_id, if not, please change it to gene_name</span>
<span class="kn">parameter:</span> <span class="n">phenotype_id_type</span> <span class="o">=</span> <span class="s1">&#39;gene_id&#39;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="region-list-generation">
<h2>Region List generation<a class="headerlink" href="#region-list-generation" title="Permalink to this headline">#</a></h2>
<p>To partitioning the data by genes require a region list file which:</p>
<ol class="arabic simple">
<li><p>have 5 columns: chr,start,end,gene_id,gene_name</p></li>
<li><p>have the same gene as or less gene than that of the bed file</p></li>
</ol>
<p>Input:</p>
<ol class="arabic simple">
<li><p>A gtf file used to generated the bed</p></li>
<li><p>A phenotype bed file, must have a gene_id column indicating the name of genes.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[region_list_generation]
input: phenoFile, annotation_gtf
output: f&#39;{cwd:a}/{_input[0]:bnn}.region_list&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
    import pandas as pd
    import qtl.io
    # get the five column data
    bed_template_df_id = qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;transcript&#39;,phenotype_id = &quot;gene_id&quot; )
    bed_template_df_name = qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;transcript&#39;,phenotype_id = &quot;gene_name&quot; )
    bed_template_df = bed_template_df_id.merge(bed_template_df_name, on = [&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;])
    bed_template_df.columns = [&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;gene_id&quot;,&quot;gene_name&quot;]
    pheno = pd.read_csv(${_input[0]:r}, sep = &quot;\t&quot;)
    # Retaining only the genes in the data
    region_list = bed_template_df[bed_template_df.${phenotype_id_type}.isin(pheno.gene_id)]
    region_list.to_csv(&quot;${_output}&quot;, sep = &quot;\t&quot;,index = 0)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
        stdout=$[_output:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="implementation-using-pyqtl">
<h2>Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code><a class="headerlink" href="#implementation-using-pyqtl" title="Permalink to this headline">#</a></h2>
<p>Implementation based on <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py">GTEx pipeline</a>.</p>
<p>Following step serves to annotate cood for gene expression file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_coord_gene]
# A file to map sample ID from expression to genotype, must contain two columns, sample_id and participant_id, mapping IDs in the expression files to IDs in the genotype (these can be the same).
parameter: sample_participant_lookup = path()
input: phenoFile, annotation_gtf
output: f&#39;{cwd:a}/{_input[0]:bn}.bed.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container

    import pandas as pd
    import qtl.io
    from pathlib import Path
    def prepare_bed(df, bed_template_df, chr_subset=None):
        bed_df = pd.merge(bed_template_df, df, left_index=True, right_index=True)
        # sort by start position
        bed_df = bed_df.groupby(&#39;#chr&#39;, sort=False, group_keys=False).apply(lambda x: x.sort_values(&#39;start&#39;))
        if chr_subset is not None:
            # subset chrs from VCF
            bed_df = bed_df[bed_df.chr.isin(chr_subset)]
        return bed_df
    # Load data
    df = pd.read_csv(${_input[0]:ar}, sep=&#39;\t&#39;, skiprows=0)
    sample_participant_lookup = Path(&quot;${sample_participant_lookup:a}&quot;)
    if &quot;chr&quot; in df.columns and &quot;start&quot; in df.columns and  &quot;end&quot; in df.columns:
        df = df.drop([&quot;chr&quot;, &quot;start&quot;, &quot;end&quot; ])
    df.set_index( df.columns[0] , inplace=True)
    
    # change sample IDs to participant IDs
    if sample_participant_lookup.is_file():
        sample_participant_lookup_s = pd.read_csv(sample_participant_lookup, sep=&quot;\t&quot;, index_col=0, dtype={0:str,1:str}, squeeze=True)
        df.rename(columns=sample_participant_lookup_s.to_dict(), inplace=True)

    if sum(qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;gene&#39;,phenotype_id = &quot;gene_id&quot; ).index.duplicated()) &gt;0:
        raise ValueError(f&quot;GTF file ${_input[1]:ar} needs to be collapsed into gene model by reference data processing module&quot;)
         
    bed_template_df_id = qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;transcript&#39;,phenotype_id = &quot;gene_id&quot; )
    bed_template_df_name = qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;transcript&#39;,phenotype_id = &quot;gene_name&quot; )
    bed_template_df = bed_template_df_id.merge(bed_template_df_name, on = [&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;])
    bed_template_df.columns = [&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;gene_id&quot;,&quot;gene_name&quot;]
    bed_template_df = bed_template_df.set_index(&quot;${phenotype_id_type}&quot;, drop = False )
    ### Detect duplicated gene_id
    bed_df = prepare_bed(df, bed_template_df).drop(&quot;gene_name&quot;,axis = 1)
    bed_df = bed_df.drop_duplicates(&quot;gene_id&quot;,keep = False)
    qtl.io.write_bed(bed_df, ${_output:r})

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
        stdout=$[_output:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
<p>Following step serves to annotate cood for proteiomics data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_coord_protein]
# A file to map sample ID from expression to genotype, must contain two columns, sample_id and participant_id, mapping IDs in the expression files to IDs in the genotype (these can be the same).
parameter: sample_participant_lookup = path()
parameter: protein_name_index = path()
parameter: protein_ID_type = &quot;SOMAseqID&quot;
input: phenoFile, annotation_gtf
output: f&#39;{cwd:a}/{_input[0]:bn}.bed.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container

    import pandas as pd
    import qtl.io
    from pathlib import Path
    def prepare_bed(df, bed_template_df, chr_subset=None):
        bed_df = pd.merge(bed_template_df, df, left_index=True, right_index=True)
        # sort by start position
        bed_df = bed_df.groupby(&#39;#chr&#39;, sort=False, group_keys=False).apply(lambda x: x.sort_values(&#39;start&#39;))
        if chr_subset is not None:
            # subset chrs from VCF
            bed_df = bed_df[bed_df.chr.isin(chr_subset)]
        return bed_df
    # Load data
    df = pd.read_csv(${_input[0]:ar}, sep=None , skiprows=0)
    if &quot;chr&quot; in df.columns or &quot;start&quot; in df.columns or  &quot;end&quot; in df.columns:
        df = df.drop([&quot;chr&quot;, &quot;start&quot;, &quot;end&quot; ])
    protein_ID = df.columns.values[0]

    protein_name_index = Path(&quot;${protein_name_index:a}&quot;)
    # For other data please specify a protein ID index
    if protein_name_index.is_file():
        df_info = pd.read_csv(protein_name_index).rename(columns={&#39;${protein_ID_type}&#39;: protein_ID, &#39;EntrezGeneSymbol&#39;:&#39;gene_name&#39;})[[&#39;gene_name&#39;,protein_ID,&#39;UniProt&#39;]]
        df = df_info.merge(df, on  = protein_ID).drop(protein_ID,axis = 1)
    else: # For ROSMAP data
        df[[protein_ID, &#39;UniProt&#39;]] = df[protein_ID].str.split(&#39;|&#39;, 1, expand=True)


    sample_participant_lookup = Path(&quot;${sample_participant_lookup:a}&quot;)

    if &quot;chr&quot; in df.columns and &quot;start&quot; in df.columns and  &quot;end&quot; in df.columns:
        df = df.drop([&quot;chr&quot;, &quot;start&quot;, &quot;end&quot; ])
    df.set_index( df.columns[0] , inplace=True)

    # change sample IDs to participant IDs
    if sample_participant_lookup.is_file():
        sample_participant_lookup_s = pd.read_csv(sample_participant_lookup, sep=&quot;\t&quot;, index_col=0, dtype={0:str,1:str}, squeeze=True)
        df.rename(columns=sample_participant_lookup_s.to_dict(), inplace=True)

    if sum(qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;gene&#39;,phenotype_id = &quot;gene_id&quot; ).index.duplicated()) &gt;0:
        raise ValueError(f&quot;GTF file ${_input[1]:ar} needs to be collapsed into gene model by reference data processing module&quot;)
         
    bed_template_df_id = qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;transcript&#39;,phenotype_id = &quot;gene_id&quot; )
    bed_template_df_name = qtl.io.gtf_to_tss_bed(${_input[1]:ar}, feature=&#39;transcript&#39;,phenotype_id = &quot;gene_name&quot; )
    bed_template_df = bed_template_df_id.merge(bed_template_df_name, on = [&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;])
    bed_template_df.columns = [&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;gene_id&quot;,&quot;gene_name&quot;]
    bed_template_df = bed_template_df.set_index(&quot;${phenotype_id_type}&quot;, drop = False )

    ### Detect duplicated gene_id
    bed_df = prepare_bed(df, bed_template_df)
    bed_df[&quot;ID&quot;] = bed_df[&quot;#chr&quot;] + &quot;_&quot; + bed_df[&quot;gene_id&quot;] + &quot;_&quot; + bed_df[&quot;UniProt&quot;]
    bed_df = bed_df.drop_duplicates(&quot;ID&quot;,keep = False)[[&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;ID&quot;] + df.drop([&quot;UniProt&quot;],axis = 1 ).columns.values.tolist()]
    qtl.io.write_bed(bed_df, ${_output:r})

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
        stdout=$[_output:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="implementation-using-biomart">
<h2>Implementation using biomaRt<a class="headerlink" href="#implementation-using-biomart" title="Permalink to this headline">#</a></h2>
<p>This workflow adds the annotations of chr pos(TSS where start = end -1) and gene_ID to the <code class="docutils literal notranslate"><span class="pre">bed</span></code> file. <strong>This workflow is obsolete</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_coord_biomart]
parameter: ensembl_version=int

input: phenoFile
output: f&#39;{cwd:a}/{_input:bn}.bed.gz&#39;,
        f&#39;{cwd:a}/{_input:bn}.region_list&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;  
R:  expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39; ,container = container
    library(&quot;biomaRt&quot;)
    library(dplyr)
    library(readr)
    biomartCacheClear()
    gene_exp = readr::read_delim(&quot;$[_input[0]]&quot;,delim = &quot;\t&quot;)
    if(&quot;#chr&quot; %in% colnames(gene_exp) ){
      # need to re-annotate
      gene_exp = gene_exp[,4:ncol(gene_exp)]
    }
    ensembl = useEnsembl(biomart = &quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;, version = &quot;$[ensembl_version]&quot;)
    ensembl_df &lt;- getBM(attributes=c(&quot;ensembl_gene_id&quot;,&quot;chromosome_name&quot;, &quot;start_position&quot;, &quot;end_position&quot;),mart=ensembl)
    my_genes = gene_exp$gene_ID
    keep_genes =  my_genes
    my_genes_ann = ensembl_df[match(my_genes, ensembl_df$ensembl_gene_id),]%&gt;%filter(chromosome_name%in%1:23)%&gt;%dplyr::rename( &quot;#chr&quot; = chromosome_name, &quot;start&quot; = start_position, &quot;end&quot; = end_position,&quot;gene_ID&quot; = ensembl_gene_id)%&gt;%filter(gene_ID!=&quot;NA&quot;, gene_ID%in%keep_genes)
    my_genes_ann%&gt;%select(`#chr`,start,end,gene_ID)%&gt;%write_delim(path = &quot;$[_output[1]]&quot;,&quot;\t&quot;)
    my_gene_bed = inner_join(my_genes_ann %&gt;%mutate(end = start + 1) %&gt;%select(`#chr`,start,end,gene_ID),gene_exp,by = &quot;gene_ID&quot; )%&gt;%arrange(`#chr`,start) 
    my_gene_bed%&gt;%readr::write_tsv( path = &quot;$[_output[0]:n]&quot;, na = &quot;NA&quot;, append = FALSE, col_names = TRUE, quote_escape = &quot;double&quot;)

bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container = container
        bgzip -f $[_output[0]:n]
        tabix -p bed $[_output[0]] -f
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotation-of-leafcutter-isoform">
<h2>Annotation of leafcutter isoform<a class="headerlink" href="#annotation-of-leafcutter-isoform" title="Permalink to this headline">#</a></h2>
<p>The following steps processed the output files of leafcutter so that they are TensorQTL ready. Shown below are three intemediate files</p>
<p>Exon list</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>chr</p></th>
<th class="head"><p>start</p></th>
<th class="head"><p>end</p></th>
<th class="head"><p>strand</p></th>
<th class="head"><p>gene_id</p></th>
<th class="head"><p>gene_name</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>29554</p></td>
<td><p>30039</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>30564</p></td>
<td><p>30667</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>30976</p></td>
<td><p>31097</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>35721</p></td>
<td><p>36081</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>35277</p></td>
<td><p>35481</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>34554</p></td>
<td><p>35174</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>65419</p></td>
<td><p>65433</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>65520</p></td>
<td><p>65573</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>69037</p></td>
<td><p>71585</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
</tbody>
</table>
<p>clusters_to_genes</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>clu</p></th>
<th class="head"><p>genes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1:clu_1_+</p></td>
<td><p>ENSG00000116288</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_10_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_11_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_12_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_14_-</p></td>
<td><p>ENSG00000126709</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_15_-</p></td>
<td><p>ENSG00000121753</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_16_-</p></td>
<td><p>ENSG00000121753</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_17_-</p></td>
<td><p>ENSG00000116560</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_18_-</p></td>
<td><p>ENSG00000143549</p></td>
</tr>
</tbody>
</table>
<p>phenotype_group</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>X1</p></th>
<th class="head"><p>X2</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>7:102476270:102478811:clu_309_-:ENSG00000005075</p></td>
<td><p>ENSG00000005075</p></td>
</tr>
<tr class="row-odd"><td><p>7:102476270:102478808:clu_309_-:ENSG00000005075</p></td>
<td><p>ENSG00000005075</p></td>
</tr>
<tr class="row-even"><td><p>X:47572961:47574002:clu_349_-:ENSG00000008056</p></td>
<td><p>ENSG00000008056</p></td>
</tr>
<tr class="row-odd"><td><p>X:47572999:47574002:clu_349_-:ENSG00000008056</p></td>
<td><p>ENSG00000008056</p></td>
</tr>
<tr class="row-even"><td><p>8:27236905:27239971:clu_322_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>8:27239279:27239971:clu_322_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-even"><td><p>8:27241262:27241677:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>8:27241262:27242397:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-even"><td><p>8:27241757:27242397:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>1:35558223:35559107:clu_4_+:ENSG00000020129</p></td>
<td><p>ENSG00000020129</p></td>
</tr>
</tbody>
</table>
<p>The gtf used here should be the collapsed gtf, i.e. the final output of reference_data gtf processing and the one used to called rnaseq.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[map_leafcutter_cluster_to_gene]
## Extract the code in case psichromatic needs to be processed the same way
## PheoFile in this step is the intron_count file
parameter: intron_count = path
input: intron_count, annotation_gtf
output: f&#39;{_input[1]}.exon_list&#39;, f&#39;{cwd}/{_input[0]:b}.leafcutter.clusters_to_genes.txt&#39;
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
    import pandas as pd
    import qtl.annotation
    # Load data
    annot = qtl.annotation.Annotation(${_input[1]:r})
    exon_df = pd.DataFrame([[g.chr, e.start_pos, e.end_pos, g.strand, g.id, g.name]
                        for g in annot.genes for e in g.transcripts[0].exons],
                       columns=[&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;, &#39;strand&#39;, &#39;gene_id&#39;, &#39;gene_name&#39;])
    exon_df.to_csv(${_output[0]:r}, sep=&#39;\t&#39;, index=False)


R:expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
    suppressMessages(library(dplyr, quietly=TRUE))
    suppressMessages(library(stringr, quietly=TRUE))
    suppressMessages(library(foreach, quietly=TRUE))
    # leafcutter functions:
    
    #&#39; Make a data.frame of meta data about the introns
    #&#39; @param introns Names of the introns
    #&#39; @return Data.frame with chr, start, end, cluster id
    #&#39; @export
    get_intron_meta &lt;- function(introns) {
      intron_meta &lt;- do.call(rbind, strsplit(introns,&quot;:&quot;))
      colnames(intron_meta) &lt;- c(&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;clu&quot;)
      intron_meta &lt;- as.data.frame(intron_meta, stringsAsFactors=FALSE)
      intron_meta$start &lt;- as.numeric(intron_meta$start)
      intron_meta$end &lt;- as.numeric(intron_meta$end)
      intron_meta
    }
    
    #&#39; Work out which gene each cluster belongs to. Note the chromosome names used in the two inputs must match.
    #&#39; @param intron_meta Data frame describing the introns, usually from get_intron_meta
    #&#39; @param exons_table Table of exons, see e.g. /data/gencode19_exons.txt.gz
    #&#39; @return Data.frame with cluster ids and genes separated by commas
    #&#39; @import dplyr
    #&#39; @export
    map_clusters_to_genes &lt;- function(intron_meta, exons_table) {
      gene_df &lt;- foreach (chr=sort(unique(intron_meta$chr)), .combine=rbind) %dopar% {
    
        intron_chr &lt;- intron_meta[ intron_meta$chr==chr, ]
        exons_chr &lt;- exons_table[exons_table$chr==chr, ]
    
        exons_chr$temp &lt;- exons_chr$start
        intron_chr$temp &lt;- intron_chr$end
        three_prime_matches &lt;- inner_join( intron_chr, exons_chr, by=&quot;temp&quot;)
    
        exons_chr$temp &lt;- exons_chr$end
        intron_chr$temp &lt;- intron_chr$start
        five_prime_matches &lt;- inner_join( intron_chr, exons_chr, by=&quot;temp&quot;)
    
        all_matches &lt;- rbind(three_prime_matches, five_prime_matches)[ , c(&quot;clu&quot;, &quot;gene_name&quot;)]
    
        all_matches &lt;- all_matches[!duplicated(all_matches),]
    
        if (nrow(all_matches)==0) return(NULL)
        all_matches$clu &lt;- paste(chr,all_matches$clu,sep=&#39;:&#39;)
        all_matches
      }
    
      clu_df &lt;- gene_df %&gt;% group_by(clu) %&gt;% summarize(genes=paste(gene_name, collapse = &quot;,&quot;))
      class(clu_df) &lt;- &quot;data.frame&quot;
      clu_df
    }

    cat(&quot;LeafCutter: mapping clusters to genes\n&quot;)
    intron_counts &lt;- read.table(${_input[0]:r}, header=TRUE, check.names=FALSE, row.names=1)
    intron_meta &lt;- get_intron_meta(rownames(intron_counts))
    exon_table &lt;- read.table(${_output[0]:r}, header=TRUE, stringsAsFactors=FALSE)
    if(!str_detect(intron_meta,&quot;chr&quot;)) {
        exon_table = exon_table%&gt;%mutate(chr = str_remove_all(chr,&quot;chr&quot;))
    } else if (!any(str_detect(exon_table$chr[1],&quot;chr&quot;))) {
        exon_table = exon_table%&gt;%mutate(chr = paste0(&quot;chr&quot;,chr))
    } else (exon_table = exon_table)
    stopifnot(is.element(&#39;gene_id&#39;, colnames(exon_table)))
    exon_table[, &#39;gene_name&#39;] &lt;- exon_table[, &#39;gene_id&#39;]
    m &lt;- map_clusters_to_genes(intron_meta, exon_table)
    write.table(m, ${_output[1]:r}, sep = &quot;\t&quot;, quote=FALSE, row.names=FALSE)
  
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
        stdout=$[_output[0]:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_leafcutter_isoforms]
parameter: sample_participant_lookup = path()
input: phenoFile, annotation_gtf,output_from(&quot;map_leafcutter_cluster_to_gene&quot;)
output: f&#39;{cwd:a}/{_input[0]:bn}.formated.bed.gz&#39;, f&#39;{cwd:a}/{_input[0]:bn}.phenotype_group.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
    import pandas as pd
    import numpy as np
    import qtl.io
    from pathlib import Path
    # Load data
    tss_df = qtl.io.gtf_to_tss_bed(${_input[1]:r})
    bed_df = pd.read_csv(${_input[0]:ar}, sep=&#39;\t&#39;, skiprows=0)
    bed_df.columns.values[0] = &quot;#chr&quot; # Temporary
    sample_participant_lookup = Path(&quot;${sample_participant_lookup:a}&quot;)
    cluster2gene_dict = pd.read_csv(${_input[3]:r}, sep=&#39;\t&#39;, index_col=0, squeeze=True).to_dict()
    print(&#39;    ** assigning introns to gene mapping(s)&#39;)
    n = 0
    gene_bed_df = []
    group_s = {}
    for _,r in bed_df.iterrows():
        s = r[&#39;ID&#39;].split(&#39;:&#39;)
        cluster_id = s[0]+&#39;:&#39;+s[-1]
        if cluster_id in cluster2gene_dict:
            gene_ids = cluster2gene_dict[cluster_id].split(&#39;,&#39;)
            for g in gene_ids:
                gi = r[&#39;ID&#39;]+&#39;:&#39;+g
                gene_bed_df.append(tss_df.loc[g, [&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;]].tolist() + [gi] + r.iloc[4:].tolist())
                group_s[gi] = g
        else:
            n += 1
    if n &gt; 0:
        print(f&#39;    ** discarded {n} introns without a gene mapping&#39;)

    print(&#39;  * writing BED files for QTL mapping&#39;)
    gene_bed_df = pd.DataFrame(gene_bed_df, columns=bed_df.columns)
    # sort by TSS
    gene_bed_df = gene_bed_df.groupby(&#39;#chr&#39;, sort=False, group_keys=False).apply(lambda x: x.sort_values(&#39;start&#39;))
    # change sample IDs to participant IDs
    if sample_participant_lookup.is_file():
        sample_participant_lookup_s = pd.read_csv(sample_participant_lookup, sep=&quot;\t&quot;, index_col=0, dtype={0:str,1:str}, squeeze=True)
        gene_bed_df.rename(columns=sample_participant_lookup_s, inplace=True)
    gene_bed_df = gene_bed_df.drop_duplicates()
    qtl.io.write_bed(gene_bed_df, ${_output[0]:r})
    gene_bed_df[[&#39;start&#39;, &#39;end&#39;]] = gene_bed_df[[&#39;start&#39;, &#39;end&#39;]].astype(np.int32)
    gene_bed_df[gene_bed_df.columns[4:]] = gene_bed_df[gene_bed_df.columns[4:]].astype(np.float32)
    pd.Series(group_s).sort_values().to_csv(${_output[1]:r}, sep=&#39;\t&#39;, header=False)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
        stdout=$[_output[0]:n].stdout
        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="processing-of-psichomics-output">
<h2>Processing of psichomics output<a class="headerlink" href="#processing-of-psichomics-output" title="Permalink to this headline">#</a></h2>
<p>It occurs that the psichomatic by default grouped the isoforms by gene name, so only thing needs to be done is to extract this information and potentially renamed the gene symbol into ENSG ID</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_psichomics_isoforms]
parameter: sample_participant_lookup = path()
input: phenoFile, annotation_gtf
output: f&#39;{cwd:a}/{_input[0]:bn}.formated.bed.gz&#39;, f&#39;{cwd:a}/{_input[0]:bn}.phenotype_group.txt&#39;
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
    import pandas as pd
    import numpy as np
    import qtl.io
    from pathlib import Path
    tss_df = qtl.io.gtf_to_tss_bed(${_input[1]:r}, feature=&#39;gene&#39;,phenotype_id = &quot;gene_id&quot; )
    bed_df = pd.read_csv(${_input[0]:ar}, sep=&#39;\t&#39;, skiprows=0)
    bed_df[&quot;gene_id&quot;]  = [x[-1] for x in bed_df.ID.str.split(&quot;_&quot;)]
    sample_participant_lookup = Path(&quot;${sample_participant_lookup:a}&quot;)
    if &quot;start&quot; in  bed_df.columns:
        bed_df = bed_df.drop([&quot;#Chr&quot;,&quot;start&quot;,&quot;end&quot;],axis = 1)
    output = tss_df.merge(bed_df, left_on = [&quot;gene_id&quot;], right_on = [&quot;gene_id&quot;],how = &quot;right&quot;).sort_values([&quot;chr&quot;,&quot;start&quot;])
    # change sample IDs to participant IDs
    if sample_participant_lookup.is_file():
        sample_participant_lookup_s = pd.read_csv(sample_participant_lookup, sep=&quot;\t&quot;, index_col=0, dtype={0:str,1:str}, squeeze=True)
        output.rename(columns=sample_participant_lookup_s.to_dict(), inplace=True)

    # Old code grouping by each gene
    #bed_output = output.drop(&quot;gene_id&quot; , axis = 1)
    #phenotype_group = output[[&quot;ID&quot;,&quot;gene_id&quot;]]

    bed_output = output.drop(&quot;gene_id&quot; , axis = 1)

    # R version 
    #output = output%&gt;%mutate(gene_type_id = map_chr(str_split( ID ,&quot;_&quot;),~paste0(.x[length(.x)],&quot;_&quot;,.x[1] )) )
    
    # corrected in python
    output = output.assign(gene_type_id = output[&#39;ID&#39;].str.split(&#39;_&#39;).map(lambda x: x[-1] + &#39;_&#39; + x[0]))
    phenotype_group = output[[&quot;ID&quot;,&quot;gene_type_id&quot;]]

    bed_output.to_csv(${_output[0]:nr},&quot;\t&quot;,index = False)
    phenotype_group.to_csv(${_output[1]:r},&quot;\t&quot;,index = False,header=False)

bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
    bgzip -f ${_output[0]:n}
    tabix ${_output[0]}

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container
        stdout=$[_output[0]:n].stdout
        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing/phenotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../phenotype_preprocessing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Phenotype data preprocessing</p>
      </div>
    </a>
    <a class="right-next"
       href="phenotype_formatting.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Phenotype data formatting</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alternative-implementation">Alternative implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#region-list-generation">Region List generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-pyqtl">Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-biomart">Implementation using biomaRt</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-of-leafcutter-isoform">Annotation of leafcutter isoform</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-of-psichomics-output">Processing of psichomics output</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>