

<!DOCTYPE html>


<html >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>MASH analysis pipeline with data-driven prior matrices &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/multivariate/MASH/mashr';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Univariate fine-mapping workflows" href="../../fine_mapping/univariate_fine_mapping.html" />
    <link rel="prev" title="A multivariate EBNM approach for mixture multivariate distribution estimate" href="mixture_prior.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="None"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../MASH_pipeline.html">Multivariate association with MASH</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate/MASH/mashr.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate/MASH/mashr.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/multivariate/MASH/mashr.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>MASH analysis pipeline with data-driven prior matrices</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-overview">Data overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preparing-mash-input">Preparing MASH input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#some-data-integrity-check">Some data integrity check</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#data-job-summary">Data &amp; job summary</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data">Multivariate adaptive shrinkage (MASH) analysis of eQTL data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#global-parameter-settings">Global parameter settings</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#factor-analyses">Factor analyses</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#estimate-residual-variance">Estimate residual variance</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-mash-priors">Compute MASH priors</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mashr-mixture-model-fitting"><code class="docutils literal notranslate"><span class="pre">mashr</span></code> mixture model fitting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-posterior-computations">Optional posterior computations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-mash-posteriors">Compute MASH posteriors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-results">Posterior results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-test">Posterior test</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#slice-posterior">Slice Posterior</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="mash-analysis-pipeline-with-data-driven-prior-matrices">
<h1>MASH analysis pipeline with data-driven prior matrices<a class="headerlink" href="#mash-analysis-pipeline-with-data-driven-prior-matrices" title="Permalink to this headline">#</a></h1>
<p>This notebook is a pipeline written in SoS to run <code class="docutils literal notranslate"><span class="pre">flashr</span> <span class="pre">+</span> <span class="pre">mashr</span></code> for multivariate analysis described in Urbut et al (2019). This pipeline was last applied to analyze GTEx V8 eQTL data, although it can be used as is to perform similar multivariate analysis for other association studies.</p>
<p><em>Version: 2021.02.28 by Gao Wang and Yuxin Zou</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">revisions</span> <span class="o">-</span><span class="n">s</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
        <table class="revision_table">
        <tr>
        <th>Revision</th>
        <th>Author</th>
        <th>Date</th>
        <th>Message</th>
        <tr>
        <tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/2570aa68b8c8c1a543c8a98d1cf486575736cbe1/mashr_flashr_workflow.ipynb"><span class="revision_id">2570aa6<span></a></td>
<td>Gao Wang</td>
<td>2019-10-03</td>
<td>Update and improve data preprocessing pipelines</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/3a02d2d7635f52aa1e81687479a8b782cfdf19a4/mashr_flashr_workflow.ipynb"><span class="revision_id">3a02d2d<span></a></td>
<td>Gao Wang</td>
<td>2019-07-11</td>
<td>Update mashr version</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/3de66784e79d0b11a9f995ff11fdc7a4fbe12223/mashr_flashr_workflow.ipynb"><span class="revision_id">3de6678<span></a></td>
<td>Gao Wang</td>
<td>2019-02-05</td>
<td>Add corshrink pipeline implementation</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/80e89a86f9ce380b7f30dc4dd64ef4ec632ef2d0/mashr_flashr_workflow.ipynb"><span class="revision_id">80e89a8<span></a></td>
<td>Gao Wang</td>
<td>2019-01-28</td>
<td>Add a note on HPC job submission</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/982a1b4cc4d8a14a77587fd55a825bf9ef00391e/mashr_flashr_workflow.ipynb"><span class="revision_id">982a1b4<span></a></td>
<td>Gao Wang</td>
<td>2019-01-28</td>
<td>Implement new $\hat{V}$ estimation method (with Yuxin Zou)</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/515e86951d09a0f2c1d4d3efde43882bfb75427e/mashr_flashr_workflow.ipynb"><span class="revision_id">515e869<span></a></td>
<td>Gao Wang</td>
<td>2018-11-22</td>
<td>Add a prompt for empty input posterior</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/ba9b20e5d705b20cda80826989293022452c3101/mashr_flashr_workflow.ipynb"><span class="revision_id">ba9b20e<span></a></td>
<td>Gao Wang</td>
<td>2018-11-22</td>
<td>Add posterior calculation for input 'strong' set</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/49494706a413999201362e72c9e20cbb95351be2/mashr_flashr_workflow.ipynb"><span class="revision_id">4949470<span></a></td>
<td>Gao Wang</td>
<td>2018-11-22</td>
<td>Fix mashr null correlation estimate interface</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/6145b3366b7850eac2e10bdd69634f6482e3359f/mashr_flashr_workflow.ipynb"><span class="revision_id">6145b33<span></a></td>
<td>Gao Wang</td>
<td>2018-11-21</td>
<td>Add --optmethod to configure convex optimization method to use</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/485381c8483bc4b770f8648b325a6386cd1a68d9/mashr_flashr_workflow.ipynb"><span class="revision_id">485381c<span></a></td>
<td>Gao Wang</td>
<td>2018-11-20</td>
<td>Fix mixSQP package name #2</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/c587791ce58a5d393782466ba762a32544edcf0e/mashr_flashr_workflow.ipynb"><span class="revision_id">c587791<span></a></td>
<td>Gao Wang</td>
<td>2018-11-20</td>
<td>Use the first cran release of mixSQP in default flashr workflow</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/092e20680cce52f594dd7fb95ee2c6b8ea85f036/mashr_flashr_workflow.ipynb"><span class="revision_id">092e206<span></a></td>
<td>Gao Wang</td>
<td>2018-09-24</td>
<td>Minor edits to documentation</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/4a5f2b78f94619f5759448e1fd0a9ba8fb600cb0/mashr_flashr_workflow.ipynb"><span class="revision_id">4a5f2b7<span></a></td>
<td>Gao Wang</td>
<td>2018-09-24</td>
<td>Add notes to results</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/e653d39708c6ba5b96da27472aed2896232814a8/mashr_flashr_workflow.ipynb"><span class="revision_id">e653d39<span></a></td>
<td>Gao Wang</td>
<td>2018-09-24</td>
<td>Configure posterior computation resources</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/ca56229760bd274d193689dbe9bf3efea8103b65/mashr_flashr_workflow.ipynb"><span class="revision_id">ca56229<span></a></td>
<td>Gao Wang</td>
<td>2018-09-23</td>
<td>Add posterior computation for input data-set</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/7db5500fd4549d39a21360fbcbf689bf5576094f/mashr_flashr_workflow.ipynb"><span class="revision_id">7db5500<span></a></td>
<td>Gao Wang</td>
<td>2018-09-23</td>
<td>Notes on GTEx V8 data conversion steps</td></tr><tr><td><a target="_blank" href="git@github.com:stephenslab/gtexresults/blob/68d2571e7d6284d02ac42d53e901968546298251/mashr_flashr_workflow.ipynb"><span class="revision_id">68d2571<span></a></td>
<td>Gao Wang</td>
<td>2018-05-30</td>
<td>Add mashr_flashr workflow</td></tr></table></div></div>
</div>
<section id="data-overview">
<h2>Data overview<a class="headerlink" href="#data-overview" title="Permalink to this headline">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> summary statistics data were obtained from dbGaP (data on CRI at UChicago Genetic Medicine). It has 49 tissues. [more description to come]</p>
</section>
<section id="preparing-mash-input">
<h2>Preparing MASH input<a class="headerlink" href="#preparing-mash-input" title="Permalink to this headline">#</a></h2>
<p>Using an established workflow (which takes 33hrs to run on a cluster system as configured by <code class="docutils literal notranslate"><span class="pre">midway2.yml</span></code>; see inside <code class="docutils literal notranslate"><span class="pre">fastqtl_to_mash.ipynb</span></code> for a note on computing environment),</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INPUT_DIR=/project/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_all_associations
JOB_OPT=&quot;-c midway2.yml -q midway2&quot;
sos run workflows/fastqtl_to_mash.ipynb --data-list $INPUT_DIR/FastQTLSumStats.list --common-suffix &quot;.allpairs.txt&quot; $JOB_OPT
</pre></div>
</div>
<p>As a result of command above I obtained the “mashable” data-set in the same format <a class="reference external" href="https://stephenslab.github.io/gtexresults/gtexdata.html">as described here</a>.</p>
<section id="some-data-integrity-check">
<h3>Some data integrity check<a class="headerlink" href="#some-data-integrity-check" title="Permalink to this headline">#</a></h3>
<ol class="arabic simple">
<li><p>Check if I get the same number of groups (genes) at the end of HDF5 data conversion:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ zcat Whole_Blood.allpairs.txt.gz | cut -f1 | sort -u | wc -l
20316
$ h5ls Whole_Blood.allpairs.txt.h5 | wc -l
20315
</pre></div>
</div>
<p>The results agreed on Whole Blood sample (the original data has a header thus one line more than the H5 version). We should be good (since the pipeline reported success for all other files).</p>
</section>
<section id="data-job-summary">
<h3>Data &amp; job summary<a class="headerlink" href="#data-job-summary" title="Permalink to this headline">#</a></h3>
<p>The command above took 33 hours on UChicago RCC <code class="docutils literal notranslate"><span class="pre">midway2</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>[MW] cat FastQTLSumStats.log
39832 out of 39832 groups merged!
</pre></div>
</div>
<p>So we have a total of 39832 genes (union of 49 tissues).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>[MW] cat FastQTLSumStats.portable.log
15636 out of 39832 groups extracted!
</pre></div>
</div>
<p>We have 15636 groups without missing data in any tissue. This will be used to train the MASH model.</p>
<p>The “mashable” data file is <code class="docutils literal notranslate"><span class="pre">FastQTLSumStats.mash.rds</span></code>, 124Mb serialized R file.</p>
</section>
</section>
<section id="multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data">
<h2>Multivariate adaptive shrinkage (MASH) analysis of eQTL data<a class="headerlink" href="#multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data" title="Permalink to this headline">#</a></h2>
<p>Below is a “blackbox” implementation of the <code class="docutils literal notranslate"><span class="pre">mashr</span></code> eQTL workflow – blackbox in the sense that you can run this pipeline as an executable, without thinking too much about it, if you see your problem fits our GTEx analysis scheme. However when reading it as a notebook it is a good source of information to help developing your own <code class="docutils literal notranslate"><span class="pre">mashr</span></code> analysis procedures.</p>
<p>Since the submission to biorxiv of Urbut 2017 we have improved implementation of MASH algorithm and made a new R package, <a class="reference external" href="https://github.com/stephenslab/mashr"><code class="docutils literal notranslate"><span class="pre">mashr</span></code></a>. Major improvements compared to Urbut 2019 are:</p>
<ol class="arabic simple">
<li><p>Faster computation of likelihood and posterior quantities via matrix algebra tricks and a C++ implementation.</p></li>
<li><p>Faster computation of MASH mixture via convex optimization.</p></li>
<li><p>Replace <code class="docutils literal notranslate"><span class="pre">SFA</span></code> with <code class="docutils literal notranslate"><span class="pre">FLASH</span></code>, a new sparse factor analysis method to generate prior covariance candidates.</p></li>
<li><p>Improve estimate of residual variance <span class="math notranslate nohighlight">\(\hat{V}\)</span>.</p></li>
</ol>
<p>At this point, the input data have already been converted from the original eQTL summary statistics to a format convenient for analysis in MASH, as a result of running the data conversion pipeline in <code class="docutils literal notranslate"><span class="pre">fastqtl_to_mash.ipynb</span></code>.</p>
<p>Example command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">JOB_OPT</span><span class="o">=</span><span class="s2">&quot;-j 8&quot;</span>
<span class="c1">#JOB_OPT=&quot;-c midway2.yml -q midway2&quot;</span>
sos run workflows/mashr_flashr_workflow.ipynb mash <span class="nv">$JOB_OPT</span> <span class="c1"># --data ... --cwd ... --vhat ...</span>
</pre></div>
</div>
<p><strong>FIXME: add comments on submitting jobs to HPC. Here we use the UChicago RCC cluster but other users can similarly configure their computating system to run the pipeline on HPC.</strong></p>
<section id="global-parameter-settings">
<h3>Global parameter settings<a class="headerlink" href="#global-parameter-settings" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./mashr_flashr_workflow_output&#39;</span><span class="p">)</span>
<span class="c1"># Input summary statistics data</span>
<span class="kn">parameter:</span> <span class="n">data</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;fastqtl_to_mash_output/FastQTLSumStats.mash.rds&quot;</span><span class="p">)</span>
<span class="c1"># Prefix of output files. If not specified, it will derive it from data.</span>
<span class="c1"># If it is specified, for example, `--output-prefix AnalysisResults`</span>
<span class="c1"># It will save output files as `{cwd}/AnalysisResults*`.</span>
<span class="kn">parameter:</span> <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1"># Exchangable effect (EE) or exchangable z-scores (EZ)</span>
<span class="kn">parameter:</span> <span class="n">effect_model</span> <span class="o">=</span> <span class="s1">&#39;EZ&#39;</span>
<span class="c1"># Identifier of $\hat{V}$ estimate file</span>
<span class="c1"># Options are &quot;identity&quot;, &quot;simple&quot;, &quot;mle&quot;, &quot;vhat_corshrink_xcondition&quot;, &quot;vhat_simple_specific&quot;</span>
<span class="kn">parameter:</span> <span class="n">vhat</span> <span class="o">=</span> <span class="s1">&#39;simple&#39;</span>
<span class="kn">parameter:</span> <span class="n">mixture_components</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;flash&#39;</span><span class="p">,</span> <span class="s1">&#39;flash_nonneg&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span><span class="s2">&quot;canonical&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="nb">str</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">cwd</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">output_prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="n">prior_data</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.prior.rds&quot;</span><span class="p">)</span>
<span class="n">vhat_data</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.V_</span><span class="si">{</span><span class="n">vhat</span><span class="si">}</span><span class="s2">.rds&quot;</span><span class="p">)</span>
<span class="n">mash_model</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.V_</span><span class="si">{</span><span class="n">vhat</span><span class="si">}</span><span class="s2">.mash_model.rds&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">sort_uniq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq</span> <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="n">seen</span> <span class="ow">or</span> <span class="n">seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="command-interface">
<h3>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mashr_flashr_workflow</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run mashr_flashr_workflow.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  flash
  flash_nonneg
  pca
  vhat_identity
  vhat_simple
  vhat_mle
  vhat_corshrink_xcondition
  vhat_simple_specific
  prior
  mash
  posterior

Global Workflow Options:
  --cwd mashr_flashr_workflow_output (as path)
  --data fastqtl_to_mash_output/FastQTLSumStats.mash.rds (as path)
                        Input summary statistics data
  --output-prefix &#39;&#39;
                        Prefix of output files. If not specified, it will derive
                        it from data. If it is specified, for example,
                        `--output-prefix AnalysisResults` It will save output
                        files as `{cwd}/AnalysisResults*`.
  --effect-model EZ
                        Exchangable effect (EE) or exchangable z-scores (EZ)
  --vhat mle
                        Identifier of $\hat{V}$ estimate file Options are
                        &quot;identity&quot;, &quot;simple&quot;, &quot;mle&quot;,
                        &quot;vhat_corshrink_xcondition&quot;, &quot;vhat_simple_specific&quot;
  --mixture-components flash flash_nonneg pca (as list)

Sections
  flash:                Perform FLASH analysis with non-negative factor
                        constraint (time estimate: 20min)
  flash_nonneg:         Perform FLASH analysis with non-negative factor
                        constraint (time estimate: 20min)
  pca:
    Workflow Options:
      --npc 3 (as int)
                        Number of components in PCA analysis for prior set to 3
                        as in mash paper
  vhat_identity:        V estimate: &quot;identity&quot; method
  vhat_simple:          V estimate: &quot;simple&quot; method (using null z-scores)
  vhat_mle:             V estimate: &quot;mle&quot; method
    Workflow Options:
      --n-subset 6000 (as int)
                        number of samples to use
      --max-iter 6 (as int)
                        maximum number of iterations
  vhat_corshrink_xcondition_1: Estimate each V separately via corshrink
    Workflow Options:
      --util-script /project/mstephens/gtex/scripts/SumstatQuery.R (as path)
                        Utility script
      --gene-list . (as path)
                        List of genes to analyze
  vhat_simple_specific_1: Estimate each V separately via &quot;simple&quot; method
    Workflow Options:
      --util-script /project/mstephens/gtex/scripts/SumstatQuery.R (as path)
                        Utility script
      --gene-list . (as path)
                        List of genes to analyze
  vhat_corshrink_xcondition_2, vhat_simple_specific_2: Consolidate Vhat into one
                        file
    Workflow Options:
      --gene-list . (as path)
                        List of genes to analyze
  prior:                Compute data-driven / canonical prior matrices (time
                        estimate: 2h ~ 12h for ~30 49 by 49 matrix mixture)
  mash_1:               Fit MASH mixture model (time estimate: &lt;15min for 70K by
                        49 matrix)
  mash_2:               Compute posterior for the &quot;strong&quot; set of data as in
                        Urbut et al 2017. This is optional because most of the
                        time we want to apply the MASH model learned on much
                        larger data-set.
    Workflow Options:
      --[no-]compute-posterior (default to True)
                        default to True; use --no-compute-posterior to disable
                        this
  posterior:            Apply posterior calculations
    Workflow Options:
      --mash-model  path(f&quot;{vhat_data:n}.mash_model.rds&quot;)

      --posterior-input  paths()

      --posterior-vhat-files  paths()

      --data-table-name &#39;&#39;
                        eg, if data is saved in R list as data$strong, then when
                        you specify `--data-table-name strong` it will read the
                        data as readRDS(&#39;{_input:r}&#39;)$strong
      --bhat-table-name Bhat
      --shat-table-name Shat
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="factor-analyses">
<h2>Factor analyses<a class="headerlink" href="#factor-analyses" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Perform FLASH analysis with non-negative factor constraint (time estimate: 20min)
[flash]
input: data
output: f&quot;{cwd}/{output_prefix}.flash.rds&quot;
task: trunk_workers = 1, walltime = &#39;2h&#39;, trunk_size = 1, mem = &#39;8G&#39;, cores = 2, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
    dat = readRDS(${_input:r})
    dat = mashr::mash_set_data(dat$strong.b, Shat=dat$strong.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    res = mashr::cov_flash(dat, factors=&quot;default&quot;, remove_singleton=${&quot;TRUE&quot; if &quot;canonical&quot; in mixture_components else &quot;FALSE&quot;}, output_model=&quot;${_output:n}.model.rds&quot;)
    saveRDS(res, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Perform FLASH analysis with non-negative factor constraint (time estimate: 20min)
[flash_nonneg]
input: data
output: f&quot;{cwd}/{output_prefix}.flash_nonneg.rds&quot;
task: trunk_workers = 1, walltime = &#39;2h&#39;, trunk_size = 1, mem = &#39;8G&#39;, cores = 2, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
    dat = readRDS(${_input:r})
    dat = mashr::mash_set_data(dat$strong.b, Shat=dat$strong.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    res = mashr::cov_flash(dat, factors=&quot;nonneg&quot;, remove_singleton=${&quot;TRUE&quot; if &quot;canonical&quot; in mixture_components else &quot;FALSE&quot;}, output_model=&quot;${_output:n}.model.rds&quot;)
    saveRDS(res, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[pca]
# Number of components in PCA analysis for prior
# set to 3 as in mash paper
parameter: npc = 2
input: data
output: f&quot;{cwd}/{output_prefix}.pca.rds&quot;
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 2, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
    dat = readRDS(${_input:r})
    dat = mashr::mash_set_data(dat$strong.b, Shat=dat$strong.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    res = mashr::cov_pca(dat, ${npc})
    saveRDS(res, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[canonical]
input: data
output: f&quot;{cwd}/{output_prefix}.canonical.rds&quot;
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 2, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container
    library(&quot;mashr&quot;)
    dat = readRDS(${_input:r})
    dat = mashr::mash_set_data(dat$strong.b, Shat=dat$strong.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    res = mashr::cov_canonical(dat)
    saveRDS(res, ${_output:r})
</pre></div>
</div>
</div>
</div>
<section id="estimate-residual-variance">
<h3>Estimate residual variance<a class="headerlink" href="#estimate-residual-variance" title="Permalink to this headline">#</a></h3>
<p>FIXME: add some narratives here explaining what we do in each method.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># V estimate: &quot;identity&quot; method
[vhat_identity]
input: data
output: f&#39;{vhat_data:nn}.V_identity.rds&#39;
task: trunk_workers = 1, walltime = &#39;2h&#39;, trunk_size = 1, mem = &#39;8G&#39;, cores = 2, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    dat = readRDS(${_input:r})
    saveRDS(diag(ncol(dat$random.b)), ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># V estimate: &quot;simple&quot; method (using null z-scores)
[vhat_simple]

input: data
output: f&#39;{vhat_data:nn}.V_simple.rds&#39;
task: trunk_workers = 1, walltime = &#39;2h&#39;, trunk_size = 1, mem = &#39;8G&#39;, cores = 2, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    dat = readRDS(${_input:r})
    vhat = estimate_null_correlation_simple(mash_set_data(dat$random.b, Shat=dat$random.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3))
    saveRDS(vhat, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># V estimate: &quot;mle&quot; method
[vhat_mle]
# number of samples to use
parameter: n_subset = 6000
# maximum number of iterations
parameter: max_iter = 6

input: data, prior_data
output: f&#39;{vhat_data:nn}.V_mle.rds&#39;
task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    dat = readRDS(${_input[0]:r})
    # choose random subset
    set.seed(1)
    random.subset = sample(1:nrow(dat$random.b), min(${n_subset}, nrow(dat$random.b)))
    random.subset = mash_set_data(dat$random.b[random.subset,], dat$random.s[random.subset,], alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    # estimate V mle
    vhatprior = mash_estimate_corr_em(random.subset, readRDS(${_input[1]:r}), max_iter = ${max_iter})
    vhat = vhatprior$V
    saveRDS(vhat, ${_output:r})
    saveRDS(vhat, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Estimate each V separately via corshrink
[vhat_corshrink_xcondition_1]
# Utility script
parameter: util_script = path(&#39;/project/mstephens/gtex/scripts/SumstatQuery.R&#39;)
# List of genes to analyze
parameter: gene_list = path()

fail_if(not gene_list.is_file(), msg = &#39;Please specify valid path for --gene-list&#39;)
fail_if(not util_script.is_file() and len(str(util_script)), msg = &#39;Please specify valid path for --util-script&#39;)
genes = sort_uniq([x.strip().strip(&#39;&quot;&#39;) for x in open(f&#39;{gene_list:a}&#39;).readlines() if not x.strip().startswith(&#39;#&#39;)])


depends: R_library(&quot;CorShrink&quot;)
input: data, for_each = &#39;genes&#39;
output: f&#39;{vhat_data:nn}/{vhat_data:bnn}_V_corshrink_{_genes}.rds&#39;
task: trunk_workers = 1, walltime = &#39;3m&#39;, trunk_size = 500, mem = &#39;3G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    source(${util_script:r})
    CorShrink_sum = function(gene, database, z_thresh = 2){
      print(gene)
      dat &lt;- GetSS(gene, database)
      z = dat$&quot;z-score&quot;
      max_absz = apply(abs(z), 1, max)
      nullish = which(max_absz &lt; z_thresh)
      # if (length(nullish) &lt; ncol(z)) {
        # stop(&quot;not enough null data to estimate null correlation&quot;)
      # }
      if (length(nullish) &lt;= 1){
        mat = diag(ncol(z))
      } else {
        nullish_z = z[nullish, ]  
        mat = as.matrix(CorShrink::CorShrinkData(nullish_z, ash.control = list(mixcompdist = &quot;halfuniform&quot;))$cor)
      }
      return(mat)
    }
    V = Corshrink_sum(&quot;${_genes}&quot;, ${data:r})
    saveRDS(V, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Estimate each V separately via &quot;simple&quot; method
[vhat_simple_specific_1]
# Utility script
parameter: util_script = path(&#39;/project/mstephens/gtex/scripts/SumstatQuery.R&#39;)
# List of genes to analyze
parameter: gene_list = path()

fail_if(not gene_list.is_file(), msg = &#39;Please specify valid path for --gene-list&#39;)
fail_if(not util_script.is_file() and len(str(util_script)), msg = &#39;Please specify valid path for --util-script&#39;)
genes = sort_uniq([x.strip().strip(&#39;&quot;&#39;) for x in open(f&#39;{gene_list:a}&#39;).readlines() if not x.strip().startswith(&#39;#&#39;)])

depends: R_library(&quot;Matrix&quot;)
input: data, for_each = &#39;genes&#39;
output: f&#39;{vhat_data:nn}/{vhat_data:bnn}_V_simple_{_genes}.rds&#39;

task: trunk_workers = 1, walltime = &#39;1m&#39;, trunk_size = 500, mem = &#39;3G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    source(${util_script:r})
    simple_V = function(gene, database, z_thresh = 2){
      print(gene)
      dat &lt;- GetSS(gene, database)
      z = dat$&quot;z-score&quot;
      max_absz = apply(abs(z), 1, max)
      nullish = which(max_absz &lt; z_thresh)
      # if (length(nullish) &lt; ncol(z)) {
        # stop(&quot;not enough null data to estimate null correlation&quot;)
      # }
      if (length(nullish) &lt;= 1){
        mat = diag(ncol(z))
      } else {
        nullish_z = z[nullish, ]
        mat = as.matrix(Matrix::nearPD(as.matrix(cov(nullish_z)), conv.tol=1e-06, doSym = TRUE, corr=TRUE)$mat)
      }
      return(mat)
    }
    V = simple_V(&quot;${_genes}&quot;, ${data:r})
    saveRDS(V, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Consolidate Vhat into one file
[vhat_corshrink_xcondition_2, vhat_simple_specific_2]
depends: R_library(&quot;parallel&quot;)
# List of genes to analyze
parameter: gene_list = path()

fail_if(not gene_list.is_file(), msg = &#39;Please specify valid path for --gene-list&#39;)
genes = paths([x.strip().strip(&#39;&quot;&#39;) for x in open(f&#39;{gene_list:a}&#39;).readlines() if not x.strip().startswith(&#39;#&#39;)])


input: group_by = &#39;all&#39;
output: f&quot;{vhat_data:nn}.V_{step_name.rsplit(&#39;_&#39;,1)[0]}.rds&quot;

task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(parallel)
    files = sapply(c(${genes:r,}), function(g) paste0(c(${_input[0]:adr}), &#39;/&#39;, g, &#39;.rds&#39;), USE.NAMES=FALSE)
    V = mclapply(files, function(i){ readRDS(i) }, mc.cores = 1)
    R = dim(V[[1]])[1]
    L = length(V)
    V.array = array(as.numeric(unlist(V)), dim=c(R, R, L))
    saveRDS(V.array, ${_output:ar})
</pre></div>
</div>
</div>
</div>
</section>
<section id="compute-mash-priors">
<h3>Compute MASH priors<a class="headerlink" href="#compute-mash-priors" title="Permalink to this headline">#</a></h3>
<p>Main reference are our <code class="docutils literal notranslate"><span class="pre">mashr</span></code> vignettes <a class="reference external" href="https://stephenslab.github.io/mashr/articles/eQTL_outline.html">this for mashr eQTL outline</a> and <a class="reference external" href="https://github.com/stephenslab/mashr/blob/master/vignettes/flash_mash.Rmd">this for using FLASH prior</a>.</p>
<p>The outcome of this workflow should be found under <code class="docutils literal notranslate"><span class="pre">./mashr_flashr_workflow_output</span></code> folder (can be configured). File names have pattern <code class="docutils literal notranslate"><span class="pre">*.mash_model_*.rds</span></code>. They can be used to computer posterior for input list of gene-SNP pairs (see next section).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Compute data-driven / canonical prior matrices (time estimate: 2h ~ 12h for ~30 49 by 49 matrix mixture)
#[prior]

# if vhat method is `mle` it should use V_simple to analyze the data to provide a rough estimate, then later be refined via `mle`.
input: [data, vhat_data if vhat != &quot;mle&quot; else f&#39;{vhat_data:nn}.V_simple.rds&#39;] + [f&quot;{cwd}/{output_prefix}.{m}.rds&quot; for m in mixture_components]
output: prior_data

task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 4, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    rds_files = c(${_input:r,})
    dat = readRDS(rds_files[1])
    vhat = readRDS(rds_files[2])
    mash_data = mash_set_data(dat$strong.b, Shat=dat$strong.s, V=vhat, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    # setup prior
    U = list(XtX = t(mash_data$Bhat) %*% mash_data$Bhat / nrow(mash_data$Bhat))
    for (f in rds_files[3:length(rds_files)]) U = c(U, readRDS(f))
    U.ed = cov_ed(mash_data, U, logfile=${_output:nr})
    # Canonical matrices
    U.can = cov_canonical(mash_data)
    saveRDS(c(U.ed, U.can), ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Compute data-driven / canonical prior matrices (time estimate: 2h ~ 12h for ~30 49 by 49 matrix mixture)
[prior]
parameter: udr = 0

# if vhat method is `mle` it should use V_simple to analyze the data to provide a rough estimate, then later be refined via `mle`.
input: [data, vhat_data if vhat != &quot;mle&quot; else f&#39;{vhat_data:nn}.V_simple.rds&#39;] + [f&quot;{cwd}/{output_prefix}.{m}.rds&quot; for m in mixture_components]
output: prior_data

task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 4, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    library(remotes)
    library(mashr)
    #remotes::install_github(&quot;stephenslab/udr&quot;)
    library(udr)
    rds_files = c(${_input:r,})
    dat = readRDS(rds_files[1])
    vhat = readRDS(rds_files[2])
    mash_data = mash_set_data(dat$strong.b, Shat=dat$strong.s, V=vhat, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, zero_Bhat_Shat_reset = 1E3)
    # setup prior
    
    # Canonical matrices
    U.can = cov_canonical(mash_data)    
  
    if(${udr} == 0){
    U = list(XtX = t(mash_data$Bhat) %*% mash_data$Bhat / nrow(mash_data$Bhat))
    for (f in rds_files[3:length(rds_files)]) U = c(U, readRDS(f))
    U.ed = cov_ed(mash_data, U, logfile=${_output:nr})
    saveRDS(c(U.ed, U.can), ${_output:r})} else {
    nonmissing = apply(dat$strong.z, 1, function(x) sum(is.na(x)) == 0)
    strong.z = dat$strong.z[nonmissing, ]
    n_unconstrained = 50
    Vhat = cor(dat$null.z, use=&quot;complete.obs&quot;)
    # Penalty strength
    lambda = ncol(strong.z)
    # Initialize udr
    fit0 &lt;- ud_init(strong.z, U_scaled = NULL, n_rank1 = 0, n_unconstrained = n_unconstrained, V = Vhat)
    # Fit udr
    fit2 = ud_fit(fit0, control = list(unconstrained.update = &quot;ted&quot;, scaled.update = &quot;fa&quot;, resid.update = &#39;none&#39;, 
                                   lambda =lambda, penalty.type = &quot;iw&quot;, maxiter=2e3, tol = 1e-2, tol.lik = 1e-2), verbose=TRUE)

    # extract data-driven covariance from udr model. (A list of covariance matrices)
    U.ud &lt;- lapply(fit2$U,function (e) &quot;[[&quot;(e,&quot;mat&quot;)) 
    saveRDS(c(U.ud, U.can), ${_output:r})
    }                                                                           
    
    
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="mashr-mixture-model-fitting">
<h2><code class="docutils literal notranslate"><span class="pre">mashr</span></code> mixture model fitting<a class="headerlink" href="#mashr-mixture-model-fitting" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Fit MASH mixture model (time estimate: &lt;15min for 70K by 49 matrix)
[mash_1]

input: data, vhat_data, prior_data
output: mash_model

task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    dat = readRDS(${_input[0]:r})
    vhat = readRDS(${_input[1]:r})
    U = readRDS(${_input[2]:r})
    mash_data = mash_set_data(dat$random.b, Shat=dat$random.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)
    saveRDS(mash(mash_data, Ulist = U, outputlevel = 1), ${_output:r})
</pre></div>
</div>
</div>
</div>
<section id="optional-posterior-computations">
<h3>Optional posterior computations<a class="headerlink" href="#optional-posterior-computations" title="Permalink to this headline">#</a></h3>
<p>Additionally provide posterior for the “strong” set in MASH input data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Compute posterior for the &quot;strong&quot; set of data as in Urbut et al 2017.
# This is optional because most of the time we want to apply the 
# MASH model learned on much larger data-set.
[mash_2]
# default to True; use --no-compute-posterior to disable this
parameter: compute_posterior = True
# input Vhat file for the batch of posterior data
skip_if(not compute_posterior)

input: data, vhat_data, mash_model
output: f&quot;{cwd:a}/{output_prefix}.{effect_model}.posterior.rds&quot;

task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    dat = readRDS(${_input[0]:r})
    vhat = readRDS(${_input[1]:r})
    mash_data = mash_set_data(dat$strong.b, Shat=dat$strong.s, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)
    mash_model = readRDS(${_input[2]:ar})
    saveRDS(mash_compute_posterior_matrices(mash_model, mash_data), ${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="compute-mash-posteriors">
<h2>Compute MASH posteriors<a class="headerlink" href="#compute-mash-posteriors" title="Permalink to this headline">#</a></h2>
<p>In the GTEx V6 paper we assumed one eQTL per gene and applied the model learned above to those SNPs. Under that assumption, the input data for posterior calculation will be the <code class="docutils literal notranslate"><span class="pre">dat$strong.*</span></code> matrices.
It is a fairly straightforward procedure as shown in <a class="reference external" href="https://stephenslab.github.io/mashr/articles/eQTL_outline.html">this vignette</a>.</p>
<p>But it is often more interesting to apply MASH to given list of eQTLs, eg, from those from fine-mapping results. In GTEx V8 analysis we obtain such gene-SNP pairs from DAP-G fine-mapping analysis. See <a class="reference external" href="https://stephenslab.github.io/gtex-eqtls/analysis/Independent_eQTL_Results.html">this notebook</a> for how the input data is prepared. The workflow below takes a number of input chunks (each chunk is a list of matrices <code class="docutils literal notranslate"><span class="pre">dat$Bhat</span></code> and <code class="docutils literal notranslate"><span class="pre">dat$Shat</span></code>)
and computes posterior for each chunk. It is therefore suited for running in parallel posterior computation for all gene-SNP pairs, if input data chunks are provided.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>JOB_OPT=&quot;-c midway2.yml -q midway2&quot;
DATA_DIR=/project/compbio/GTEx_eQTL/independent_eQTL
sos run workflows/mashr_flashr_workflow.ipynb posterior \
    $JOB_OPT \
    --posterior-input $DATA_DIR/DAPG_pip_gt_0.01-AllTissues/DAPG_pip_gt_0.01-AllTissues.*.rds \
                      $DATA_DIR/ConditionalAnalysis_AllTissues/ConditionalAnalysis_AllTissues.*.rds
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Apply posterior calculations
[posterior_1]
parameter: analysis_units = path
regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: mash_model = path(f&quot;{cwd:a}/{output_prefix}.{effect_model}.V_{vhat}.mash_model.rds&quot;)
parameter: posterior_input = [path(x[0]) for x in regions]
parameter: posterior_vhat_files = paths()
# eg, if data is saved in R list as data$strong, then
# when you specify `--data-table-name strong` it will read the data as
# readRDS(&#39;{_input:r}&#39;)$strong
parameter: data_table_name = &#39;&#39;
parameter: bhat_table_name = &#39;bhat&#39;
parameter: shat_table_name = &#39;sbhat&#39;
mash_model = f&quot;{mash_model:a}&quot;
##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0
parameter: exclude_condition = [&quot;1&quot;,&quot;3&quot;]

skip_if(len(posterior_input) == 0, msg = &quot;No posterior input data to compute on. Please specify it using --posterior-input.&quot;)
fail_if(len(posterior_vhat_files) &gt; 1 and len(posterior_vhat_files) != len(posterior_input), msg = &quot;length of --posterior-input and --posterior-vhat-files do not agree.&quot;)
for p in posterior_input:
    fail_if(not p.is_file(), msg = f&#39;Cannot find posterior input file ``{p}``&#39;)

depends: mash_model
input: posterior_input, group_by = 1
output: f&quot;{cwd}/{_input:bn}.posterior.rds&quot;
task: trunk_workers = 1, walltime = &#39;20h&#39;, trunk_size = 1, mem = &#39;20G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    data = readRDS(&quot;${_input}&quot;)${(&#39;$&#39; + data_table_name) if data_table_name else &#39;&#39;}
    if(c(${&quot;,&quot;.join(exclude_condition)})[1] &gt; 0 ){
      message(paste(&quot;Excluding condition ${exclude_condition} from the analysis&quot;))
      data$bhat = data$bhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
      data$sbhat = data$sbhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
      data$Z = data$Z[,-c(${&quot;,&quot;.join(exclude_condition)})]
    }
  
    vhat = readRDS(&quot;${vhat_data if len(posterior_vhat_files) == 0 else posterior_vhat_files[_index]}&quot;)
    mash_data = mash_set_data(data$${bhat_table_name}, Shat=data$${shat_table_name}, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)
    mash_output = mash_compute_posterior_matrices(readRDS(&quot;${mash_model}&quot;), mash_data)
    mash_output$snps = data$snps
    saveRDS(mash_output, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[posterior_2]
input: group_by = &quot;all&quot;
output:f&quot;{cwd}/mash_output_list&quot;
python: expand = &quot;$[ ]&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container
    library(mashr)
    import pandas as pd
    pd.DataFrame({&quot;#mash_result&quot; :  [$[_input:ar,]] }).to_csv(&quot;$[_output]&quot;,index = False ,header = False, sep = &quot;t&quot;)
</pre></div>
</div>
</div>
</div>
<section id="posterior-results">
<h3>Posterior results<a class="headerlink" href="#posterior-results" title="Permalink to this headline">#</a></h3>
<ol class="arabic simple">
<li><p>The outcome of the <code class="docutils literal notranslate"><span class="pre">[posterior]</span></code> step should produce a number of serialized R objects <code class="docutils literal notranslate"><span class="pre">*.batch_*.posterior.rds</span></code> (can be loaded to R via <code class="docutils literal notranslate"><span class="pre">readRDS()</span></code>) – I chopped data to batches to take advantage of computing in multiple cluster nodes. It should be self-explanary but please let me know otherwise.</p></li>
<li><p>Other posterior related files are:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.batch_*.yaml</span></code>: gene-SNP pairs of interest, identified elsewhere (eg. fine-mapping analysis).</p></li>
<li><p>The corresponding univariate analysis summary statistics for gene-SNPs from <code class="docutils literal notranslate"><span class="pre">*.batch_*.yaml</span></code> are extracted and saved to <code class="docutils literal notranslate"><span class="pre">*.batch_*.rds</span></code>, creating input to the <code class="docutils literal notranslate"><span class="pre">[posterior]</span></code> step.</p></li>
<li><p>Note the <code class="docutils literal notranslate"><span class="pre">*.batch_*.stdout</span></code> file documents some SNPs found in fine-mapping results but not found in the original <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> output.</p></li>
</ol>
</li>
</ol>
</section>
</section>
<section id="posterior-test">
<h2>Posterior test<a class="headerlink" href="#posterior-test" title="Permalink to this headline">#</a></h2>
<p>skip the files with NA column. Containers are not used for now.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[posteriortest_1]
parameter: analysis_units = path
regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: mash_model = path(f&quot;{cwd:a}/{output_prefix}.{effect_model}.V_{vhat}.mash_model.rds&quot;)
parameter: posterior_input = [path(x[0]) for x in regions]
parameter: posterior_vhat_files = paths()
# eg, if data is saved in R list as data$strong, then
# when you specify `--data-table-name strong` it will read the data as
# readRDS(&#39;{_input:r}&#39;)$strong
parameter: data_table_name = &#39;&#39;
parameter: bhat_table_name = &#39;bhat&#39;
parameter: shat_table_name = &#39;sbhat&#39;
parameter: per_chunk = &#39;1000&#39;
mash_model = f&quot;{mash_model:a}&quot;
##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0
parameter: exclude_condition = [&quot;1&quot;,&quot;3&quot;]

skip_if(len(posterior_input) == 0, msg = &quot;No posterior input data to compute on. Please specify it using --posterior-input.&quot;)
fail_if(len(posterior_vhat_files) &gt; 1 and len(posterior_vhat_files) != len(posterior_input), msg = &quot;length of --posterior-input and --posterior-vhat-files do not agree.&quot;)
for p in posterior_input:
    fail_if(not p.is_file(), msg = f&#39;Cannot find posterior input file ``{p}``&#39;)

depends: mash_model
input: posterior_input, group_by = per_chunk
output: f&quot;{cwd}/mash_output_list_{_index+1}&quot;
task: trunk_workers = 1, walltime = &#39;20h&#39;, trunk_size = 1, mem = &#39;20G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;
    library(mashr)
    library(stringr)
    library(dplyr)
    outlist = data.frame()
    for (f in c(${_input:r,})) {
    data = readRDS(f)
    if(c(${&quot;,&quot;.join(exclude_condition)})[1] &gt; 0 ){
      message(paste(&quot;Excluding condition ${exclude_condition} from the analysis&quot;))
      data$bhat = data$bhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
      data$sbhat = data$sbhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
      data$Z = data$Z[,-c(${&quot;,&quot;.join(exclude_condition)})]
    }
  
    vhat = readRDS(&quot;${vhat_data if len(posterior_vhat_files) == 0 else posterior_vhat_files[_index]}&quot;)
   
    try({
      mash_data = mash_set_data(data$${bhat_table_name}, Shat=data$${shat_table_name}, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)
      mash_output = mash_compute_posterior_matrices(readRDS(&quot;${mash_model}&quot;), mash_data)
      mash_output$snps = data$snp
      
      samplename&lt;-str_split(f,&quot;/&quot;,simplify = T)%&gt;%.[length(.)]%&gt;%gsub(&#39;.rds&#39;,&#39;&#39;,.)
      saveRDS(mash_output, paste0(&quot;${_output:d}&quot;,&quot;/&quot;,samplename,&quot;.posterior.rds&quot;))
      outlist&lt;-rbind(outlist,paste0(&quot;${_output:d}&quot;,&quot;/&quot;,samplename,&quot;.posterior.rds&quot;))

    })
  
    }
    write.table(outlist,${_output:r},col.names=F, row.names=F)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[posteriortest_2]
input: group_by = &quot;all&quot;
output:f&quot;{cwd}/mash_output_list&quot;
bash: expand =&#39;${ }&#39;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;
     cd ${_input[0]:d}
     cat mash_output_list_*[0-9] &gt;&gt; ${_output:r}
</pre></div>
</div>
</div>
</div>
</section>
<section id="slice-posterior">
<h2>Slice Posterior<a class="headerlink" href="#slice-posterior" title="Permalink to this headline">#</a></h2>
<p>take all the 13K genes, and for those with missing conditions we just drop those corresponding rows and cols in the prior model</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Apply posterior calculations
[sliceposterior_1]
parameter: analysis_units = path
regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: mash_model = path(f&quot;{cwd:a}/{output_prefix}.{effect_model}.V_{vhat}.mash_model.rds&quot;)
parameter: posterior_input = [path(x[0]) for x in regions]
parameter: posterior_vhat_files = paths()
# eg, if data is saved in R list as data$strong, then
# when you specify `--data-table-name strong` it will read the data as
# readRDS(&#39;{_input:r}&#39;)$strong
parameter: data_table_name = &#39;&#39;
parameter: bhat_table_name = &#39;bhat&#39;
parameter: shat_table_name = &#39;sbhat&#39;
parameter: per_chunk = &#39;1000&#39;
mash_model = f&quot;{mash_model:a}&quot;
##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0
parameter: exclude_condition = [&quot;1&quot;,&quot;3&quot;]

skip_if(len(posterior_input) == 0, msg = &quot;No posterior input data to compute on. Please specify it using --posterior-input.&quot;)
fail_if(len(posterior_vhat_files) &gt; 1 and len(posterior_vhat_files) != len(posterior_input), msg = &quot;length of --posterior-input and --posterior-vhat-files do not agree.&quot;)
for p in posterior_input:
    fail_if(not p.is_file(), msg = f&#39;Cannot find posterior input file ``{p}``&#39;)

depends: mash_model
input: posterior_input, group_by = per_chunk
output: f&quot;{cwd}/cache/mash_output_list_{_index+1}&quot;
task: trunk_workers = 1, walltime = &#39;20h&#39;, trunk_size = 1, mem = &#39;20G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;
    library(mashr)
    library(dplyr)
    library(stringr)
    outlist = data.frame()
    for (f in c(${_input:r,})) try({
    data = readRDS(f)${(&#39;$&#39; + data_table_name) if data_table_name else &#39;&#39;}
    #data = readRDS(&quot;${_input}&quot;)${(&#39;$&#39; + data_table_name) if data_table_name else &#39;&#39;}

    if(c(${&quot;,&quot;.join(exclude_condition)})[1] &gt; 0 ){
      message(paste(&quot;Excluding condition ${exclude_condition} from the analysis&quot;))
      data$bhat = data$bhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
      data$sbhat = data$sbhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
      data$Z = data$Z[,-c(${&quot;,&quot;.join(exclude_condition)})]
    }
    
    all.samples&lt;-colnames(data$bhat)
    all.snps&lt;-rownames(data$bhat)
    
    vhat = readRDS(&quot;${vhat_data if len(posterior_vhat_files) == 0 else posterior_vhat_files[_index]}&quot;)
    mash_model &lt;- readRDS(&quot;${mash_model}&quot;)
    
    #remove the rows and cols containing NA
    na.test&lt;-data$bhat %&gt;% as.data.frame()%&gt;% select_if(~any(!is.na(.)))%&gt;% na.omit%&gt;%as.matrix
    #recording meaningful rows and cols
    samples&lt;-colnames(na.test)
    snps&lt;-rownames(na.test)
    
    if(length(all.snps)!=length(snps) | length(all.samples)!=length(samples)){
        #slice the data
        data$bhat&lt;-data$bhat[snps,samples]%&gt;%as.matrix
        colnames(data$bhat)&lt;-samples
        data$sbhat&lt;-data$sbhat[snps,samples]%&gt;%as.matrix
        colnames(data$sbhat)&lt;-samples
        data$Z&lt;-data$Z[snps,samples]%&gt;%as.matrix
        colnames(data$Z)&lt;-samples
        data$snp&lt;-data$snp[data$snp%in%snps]
        vhat&lt;-vhat[samples,samples]%&gt;%as.matrix
        colnames(vhat)&lt;-samples
  
        if(length(all.samples)!=length(samples)){
            ##slice the prior
            cov&lt;-mash_model$fitted_g$Ulist
            for (d in names(cov)) {
                if(d %in% setdiff(all.samples,samples)){
                    cov[[d]]&lt;-NULL
                }
                if(d %in% paste0(&quot;ED_&quot;,setdiff(all.samples,samples))){
                    cov[[d]]&lt;-NULL
                }
                if(d %in% samples){
                    cov[[d]]&lt;-matrix(0,length(samples),length(samples))
                    cov[[d]][which(samples==d),which(samples==d)]&lt;-1
                }else if(d == &quot;identity&quot;){
                    cov[[d]]&lt;-matrix(0,length(samples),length(samples))
                    cov[[d]][1,1]&lt;-1  
                }else if(is.null(colnames(cov[[d]]))){
                    cov[[d]] &lt;- cov[[d]][1:length(samples),1:length(samples)]
                } else {
                    cov[[d]] &lt;- cov[[d]][samples,samples]
                }
                    }
                for (d in names(cov)) {
                    cov[[d]] &lt;- cov[[d]]%&gt;%as.matrix
                }
                #slide the prior and related file
                mash_model$fitted_g$Ulist&lt;-cov
                diff.sam&lt;-setdiff(all.samples,samples)
                for(s in diff.sam){mash_model$fitted_g$pi&lt;-mash_model$fitted_g$pi[-grep(s,names(mash_model$fitted_g$pi))]}
            }
        }
    
  
  
    
    mash_data = mash_set_data(data$${bhat_table_name}, Shat=data$${shat_table_name}, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)
    mash_output = mash_compute_posterior_matrices(mash_model, mash_data)
    mash_output$snps = data$snps
    #saveRDS(mash_output, ${_output:r})
    samplename&lt;-str_split(f,&quot;/&quot;,simplify = T)%&gt;%.[length(.)]%&gt;%gsub(&#39;.rds&#39;,&#39;&#39;,.)
    saveRDS(mash_output, paste0(&quot;${_output:d}&quot;,&quot;/&quot;,samplename,&quot;.posterior.rds&quot;))
    outlist&lt;-rbind(outlist,paste0(&quot;${_output:d}&quot;,&quot;/&quot;,samplename,&quot;.posterior.rds&quot;))
    })
    write.table(outlist,${_output:r},col.names=F, row.names=F)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>#[sliceposterior_2]
input: group_by = &quot;all&quot;
output:f&quot;{cwd}/mash_output_list&quot;
bash: expand =&#39;${ }&#39;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;
     cd ${_input[0]:d}
     cat mash_output_list_*[0-9] &gt;&gt; ${_output:r}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sliceposterior_2]

input: group_by = 1
output:f&quot;{cwd}/cache/{cwd:b}_{_index+1}.posterior.rds&quot;
R: expand =&#39;${ }&#39;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;
    library(tidyverse)     
    ps&lt;-list(pm=NULL,lfsr=NULL)
    samples&lt;-read.table(c(${_input:r,}))$V1

    for (f in samples) {
    tmp&lt;-readRDS(f)
    #get gene name
    #g&lt;-f%&gt;%gsub(&quot;.posterior.rds&quot;,&quot;&quot;,.)%&gt;%stringr::str_split(.,&quot;[.]&quot;,simplify=T)
    #gene&lt;-ifelse(g[length(g)]%&gt;%grep(&quot;\\d&quot;,.)==g[length(g)], paste0(paste(g[length(g)-1],g[length(g)],sep=&quot;.&quot;)),g[length(g)])
    gene&lt;-stringr::str_split(string = f, pattern = &quot;norminal.cis_long_table.&quot;,simplify = T)[[2]]%&gt;%
               gsub(&quot;.posterior.rds&quot;,&quot;&quot;,.)
    #get the matrix of pm and lfsr
    tmp.pm&lt;-tmp$PosteriorMean%&gt;%as.data.frame()
    tmp.lf&lt;-tmp$lfsr%&gt;%as.data.frame()
    
    #change the rownames
    names &lt;- c( rownames(ps$pm) , paste(gene,str_split(rownames(tmp.pm),&quot;:&quot;,simplify=T)[,2],sep=&quot;_&quot;) )
    ps$pm&lt;-plyr::rbind.fill(ps$pm, tmp.pm)
    rownames(ps$pm) &lt;- make.names(names,unique=T)
    
    namess &lt;- c( rownames(ps$lfsr) , paste(gene,str_split(rownames(tmp.lf),&quot;:&quot;,simplify=T)[,2],sep=&quot;_&quot;) )
    ps$lfsr&lt;-plyr::rbind.fill(ps$lfsr, tmp.lf)
    rownames(ps$lfsr) &lt;- make.names(namess,unique=T)
    }
    
    saveRDS(ps,${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sliceposterior_3]
input: group_by = &quot;all&quot;
output: f&quot;{cwd}/{cwd:b}.posterior.rds&quot;
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = 1, mem = &#39;100G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, container = container,stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, volumes = [f&#39;{cwd:ad}:{cwd:ad}&#39;]
    merge_data = function(res, one_data) {
      if (length(res) == 0) {
          return(one_data)
      } else {
          for (d in names(one_data)) {
            res[[d]] = rbind(res[[d]], one_data[[d]])
          }
          return(res)
      }
    }
    dat = list()
    for (f in c(${_input:r,})) {
      dat = merge_data(dat, readRDS(f))
    }
    
    saveRDS(dat, ${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/multivariate/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="mixture_prior.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">A multivariate EBNM approach for mixture multivariate distribution estimate</p>
      </div>
    </a>
    <a class="right-next"
       href="../../fine_mapping/univariate_fine_mapping.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Univariate fine-mapping workflows</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-overview">Data overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preparing-mash-input">Preparing MASH input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#some-data-integrity-check">Some data integrity check</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#data-job-summary">Data &amp; job summary</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data">Multivariate adaptive shrinkage (MASH) analysis of eQTL data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#global-parameter-settings">Global parameter settings</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#factor-analyses">Factor analyses</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#estimate-residual-variance">Estimate residual variance</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-mash-priors">Compute MASH priors</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#mashr-mixture-model-fitting"><code class="docutils literal notranslate"><span class="pre">mashr</span></code> mixture model fitting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-posterior-computations">Optional posterior computations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-mash-posteriors">Compute MASH posteriors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-results">Posterior results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-test">Posterior test</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#slice-posterior">Slice Posterior</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>